--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -9,7 +_,7 @@
 public class TicketType<T> {
     private final String name;
     private final Comparator<T> comparator;
-    private final long timeout;
+    public final long timeout;
     public static final TicketType<Unit> START = create("start", (p_9471_, p_9472_) -> 0);
     public static final TicketType<Unit> DRAGON = create("dragon", (p_9460_, p_9461_) -> 0);
     public static final TicketType<ChunkPos> PLAYER = create("player", Comparator.comparingLong(ChunkPos::toLong));
