--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.ai.behavior;
 
 import com.mojang.datafixers.util.Pair;
+import io.papermc.paper.configuration.GlobalConfiguration;
 import it.unimi.dsi.fastutil.longs.Long2LongMap;
 import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2LongMap.Entry;
@@ -63,17 +_,20 @@
                                                 p_217372_ -> p_217372_.is(PoiTypes.HOME), predicate, p_258639_.blockPosition(), 48, PoiManager.Occupancy.ANY
                                             )
                                             .collect(Collectors.toSet());
-                                        Path path = AcquirePoi.findPathToPois(p_258639_, set);
-                                        if (path != null && path.canReach()) {
-                                            BlockPos blockpos = path.getTarget();
-                                            Optional<Holder<PoiType>> optional1 = poimanager.getType(blockpos);
-                                            if (optional1.isPresent()) {
-                                                p_258626_.set(new WalkTarget(blockpos, p_259960_, 1));
-                                                DebugPackets.sendPoiTicketCountPacket(p_258638_, blockpos);
-                                            }
-                                        } else if (mutableint.getValue() < 5) {
-                                            long2longmap.long2LongEntrySet().removeIf(p_258629_ -> p_258629_.getLongValue() < mutablelong.getValue());
-                                        }
+                                        // Kaiiju start - petal - Async path processing
+                                        if (GlobalConfiguration.get().asyncPathFinding.enabled) {
+                                            // await on path async
+                                            Path possiblePath = AcquirePoi.findPathToPois(p_258639_, set);
+
+                                            // wait on the path to be processed
+                                            org.dreeam.leaf.async.path.AsyncPathProcessor.awaitProcessing(possiblePath, path -> {
+                                                processPath(p_259960_, long2longmap, mutablelong, p_258626_, p_258638_, poimanager, mutableint, path);
+                                            });
+                                        } else {
+                                            // Kaiiju end
+                                            Path path = AcquirePoi.findPathToPois(p_258639_, set);
+                                            processPath(p_259960_, long2longmap, mutablelong, p_258626_, p_258638_, poimanager, mutableint, path);
+                                        } // Kaiiju - async path processing
 
                                         return true;
                                     } else {
@@ -84,4 +_,26 @@
                     )
         );
     }
+
+    // Leaf start - Kaiiju - petal - Async path processing
+    private static void processPath(float speedModifier,
+                                    Long2LongMap map,
+                                    MutableLong mutableLong,
+                                    net.minecraft.world.entity.ai.behavior.declarative.MemoryAccessor<com.mojang.datafixers.kinds.Const.Mu<com.mojang.datafixers.util.Unit>, WalkTarget> walkTarget,
+                                    net.minecraft.server.level.ServerLevel level,
+                                    PoiManager poiManager,
+                                    MutableInt mutableInt,
+                                    @org.jetbrains.annotations.Nullable Path path) {
+        if (path != null && path.canReach()) {
+            BlockPos target = path.getTarget();
+            Optional<Holder<PoiType>> type = poiManager.getType(target);
+            if (type.isPresent()) {
+                walkTarget.set(new WalkTarget(target, speedModifier, 1));
+                DebugPackets.sendPoiTicketCountPacket(level, target);
+            }
+        } else if (mutableInt.getValue() < 5) {
+            map.long2LongEntrySet().removeIf(entry -> entry.getLongValue() < mutableLong.getValue());
+        }
+    }
+    // Leaf end - Kaiiju - petal - Async path processing
 }
