--- a/net/minecraft/world/entity/ai/behavior/SleepInBed.java
+++ b/net/minecraft/world/entity/ai/behavior/SleepInBed.java
@@ -42,7 +_,8 @@
                     }
                 }
 
-                BlockState blockstate = p_24154_.getBlockState(globalpos.pos());
+                BlockState blockstate = p_24154_.getBlockStateIfLoaded(globalpos.pos()); // Paper - Prevent sync chunk loads when villagers try to find beds
+                if (blockstate == null) { return false; } // Paper - Prevent sync chunk loads when villagers try to find beds // Goldenforge: poor villager
                 return globalpos.pos().closerToCenterThan(p_24155_.position(), 2.0) && blockstate.is(BlockTags.BEDS) && !blockstate.getValue(BedBlock.OCCUPIED);
             }
         }
