--- a/net/minecraft/world/entity/ai/goal/AvoidEntityGoal.java
+++ b/net/minecraft/world/entity/ai/goal/AvoidEntityGoal.java
@@ -3,6 +_,8 @@
 import java.util.EnumSet;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.PathfinderMob;
@@ -58,6 +_,24 @@
 
     @Override
     public boolean canUse() {
+        // Leaf start - Async Avoid Entity Finding
+        if (GlobalConfiguration.get().asyncTargetFinding.searchEntity) {
+            if (!poll()) {
+                getNearestEntityAsync();
+                return false;
+            }
+        } else {
+            this.toAvoid = getServerLevel(this.mob)
+                    .getNearestEntity(
+                            this.mob.level().getEntitiesOfClass(this.avoidClass, this.mob.getBoundingBox().inflate(this.maxDist, 3.0, this.maxDist), livingEntity -> true),
+                            this.avoidEntityTargeting,
+                            this.mob,
+                            this.mob.getX(),
+                            this.mob.getEyeY(), // Leaf - Async Avoid Entity Finding
+                            this.mob.getZ()
+                    );
+        }
+        // Leaf end - Async Avoid Entity Finding
         this.toAvoid = this.mob
             .level()
             .getNearestEntity(
@@ -84,6 +_,36 @@
             }
         }
     }
+
+    // Leaf start - Async Avoid Entity Finding
+    private boolean poll() {
+        if (!(this.mob.getGoalCtx().result() instanceof LivingEntity target)) return false;
+        var serverLevel = getServerLevel(this.mob);
+        if (serverLevel == null || !target.isAlive() || !this.avoidEntityTargeting.test(this.mob, target)) return false;
+        this.toAvoid = (T) target;
+        return true;
+    }
+
+    private void getNearestEntityAsync() {
+        final var mob = this.mob;
+        final var ctx = mob.getGoalCtx();
+        if (!ctx.state) return;
+        final double x = mob.getX();
+        final double y = mob.getEyeY();
+        final double z = mob.getZ();
+        final var serverLevel = getServerLevel(mob);
+        final var avoidClass = this.avoidClass;
+        final var bound = mob.getBoundingBox().inflate(this.maxDist, 3.0, this.maxDist);
+        ctx.wake = () -> serverLevel.getNearestEntity(
+                serverLevel.getEntitiesOfClass(avoidClass, bound, livingEntity -> true),
+                avoidEntityTargeting,
+                mob,
+                x,
+                y,
+                z
+        );
+    }
+    // Leaf end - Async Avoid Entity Finding
 
     @Override
     public boolean canContinueToUse() {
