--- a/net/minecraft/world/entity/ai/goal/BegGoal.java
+++ b/net/minecraft/world/entity/ai/goal/BegGoal.java
@@ -2,6 +_,9 @@
 
 import java.util.EnumSet;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.animal.Wolf;
@@ -27,8 +_,43 @@
         this.setFlags(EnumSet.of(Goal.Flag.LOOK));
     }
 
+    // Leaf start - Async Target Finding
+    protected boolean poll() {
+        if (!(this.wolf.getGoalCtx().result() instanceof Player target)) return false;
+        if (target == null) return false;
+        ServerLevel serverLevel = getServerLevel(this.wolf);
+        if (serverLevel == null || !target.isAlive() || !playerHoldingInteresting(target)) return false;
+        this.player = target;
+        return true;
+    }
+
+    private void findTargetAsync() {
+        final Wolf wolf = this.wolf;
+        final var ctx = wolf.getGoalCtx();
+        if (!ctx.state) return;
+        final ServerLevel serverLevel = getServerLevel(wolf);
+        final TargetingConditions begTargeting = this.begTargeting;
+        ctx.wake = () -> {
+            var player = serverLevel.getNearestPlayer(begTargeting, wolf);
+            if (player != null && playerHoldingInteresting(player)) {
+                return player;
+            }
+            return null;
+        };
+    }
+    // Leaf end - Async Target Finding
+
     @Override
     public boolean canUse() {
+        // Leaf start - Async Target Finding
+        if (poll()) {
+            return true;
+        }
+        if (GlobalConfiguration.get().asyncTargetFinding.searchEntity) {
+            findTargetAsync();
+            return false;
+        }
+        // Leaf end - Async Target Finding
         this.player = this.level.getNearestPlayer(this.begTargeting, this.wolf);
         return this.player == null ? false : this.playerHoldingInteresting(this.player);
     }
