--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -27,6 +_,7 @@
     private final Set<WrappedGoal> availableGoals = new ObjectLinkedOpenHashSet<>();
     private final Supplier<ProfilerFiller> profiler;
     private final EnumSet<Goal.Flag> disabledFlags = EnumSet.noneOf(Goal.Flag.class);
+    private int curRate;
 
     public GoalSelector(Supplier<ProfilerFiller> p_25351_) {
         this.profiler = p_25351_;
@@ -40,6 +_,21 @@
     public void removeAllGoals(Predicate<Goal> p_262575_) {
         this.availableGoals.removeIf(p_262564_ -> p_262575_.test(p_262564_.getGoal()));
     }
+
+    // Paper start
+    public boolean inactiveTick() {
+        this.curRate++;
+        return this.curRate % 3 == 0; // TODO newGoalRate was already unused in 1.20.4, check if this is correct
+    }
+    public boolean hasTasks() {
+        for (WrappedGoal task : this.availableGoals) {
+            if (task.isRunning()) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // Paper end
 
     public void removeGoal(Goal p_25364_) {
         for (WrappedGoal wrappedgoal : this.availableGoals) {
