--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.ai.goal;
 
 import com.google.common.annotations.VisibleForTesting;
+import io.papermc.paper.configuration.GlobalConfiguration;
 import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
 import java.util.EnumMap;
 import java.util.EnumSet;
@@ -25,8 +_,19 @@
     };
     private final Map<Goal.Flag, WrappedGoal> lockedFlags = new EnumMap<>(Goal.Flag.class);
     private final Set<WrappedGoal> availableGoals = new ObjectLinkedOpenHashSet<>();
+    private static final Goal.Flag[] GOAL_FLAG_VALUES = Goal.Flag.values(); // Paper - remove streams from GoalSelector
+    private final ca.spottedleaf.moonrise.common.set.OptimizedSmallEnumSet<net.minecraft.world.entity.ai.goal.Goal.Flag> goalTypes = new ca.spottedleaf.moonrise.common.set.OptimizedSmallEnumSet<>(Goal.Flag.class); // Paper - remove streams from GoalSelector
     private final Supplier<ProfilerFiller> profiler;
-    private final EnumSet<Goal.Flag> disabledFlags = EnumSet.noneOf(Goal.Flag.class);
+    ///private final EnumSet<Goal.Flag> disabledFlags = EnumSet.noneOf(Goal.Flag.class);
+    private int curRate;
+
+    // Leaf start - Async target finding
+    private boolean availableGoalsDirty = true;
+    private WrappedGoal @org.jetbrains.annotations.Nullable[] ctxGoals = null;
+    private int ctxIndex = 0;
+    public int ctxState = -1;
+    public final org.dreeam.leaf.async.ai.Waker ctx = new org.dreeam.leaf.async.ai.Waker();
+    // Leaf end - Async target finding
 
     public GoalSelector(Supplier<ProfilerFiller> p_25351_) {
         this.profiler = p_25351_;
@@ -34,12 +_,34 @@
 
     public void addGoal(int p_25353_, Goal p_25354_) {
         this.availableGoals.add(new WrappedGoal(p_25353_, p_25354_));
+        availableGoalsDirty = true; // Leaf - Async target finding
     }
 
     @VisibleForTesting
     public void removeAllGoals(Predicate<Goal> p_262575_) {
-        this.availableGoals.removeIf(p_262564_ -> p_262575_.test(p_262564_.getGoal()));
-    }
+        this.availableGoals.removeIf(wrappedGoal -> p_262575_.test(wrappedGoal.getGoal()));
+        availableGoalsDirty = true; // Leaf - Async target finding
+    }
+
+    // Paper start - EAR 2
+    public boolean inactiveTick(int tickRate, boolean inactive) { // Pufferfish start
+        if (inactive && !GlobalConfiguration.get().dynamicActivationofBrain.enabled) tickRate = 4; // reset to Paper's
+        tickRate = Math.min(tickRate, 3); // Dreeam TODO - Waiting Paper
+        this.curRate++;
+        //return this.curRate % 3 == 0; // TODO newGoalRate was already unused in 1.20.4, check if this is correct
+        return this.curRate % tickRate == 0;
+        // Pufferfish end
+    }
+
+    public boolean hasTasks() {
+        for (WrappedGoal task : this.availableGoals) {
+            if (task.isRunning()) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // Paper end
 
     public void removeGoal(Goal p_25364_) {
         for (WrappedGoal wrappedgoal : this.availableGoals) {
@@ -49,21 +_,22 @@
         }
 
         this.availableGoals.removeIf(p_25378_ -> p_25378_.getGoal() == p_25364_);
-    }
-
-    private static boolean goalContainsAnyFlags(WrappedGoal p_186076_, EnumSet<Goal.Flag> p_186077_) {
-        for (Goal.Flag goal$flag : p_186076_.getFlags()) {
-            if (p_186077_.contains(goal$flag)) {
-                return true;
-            }
-        }
-
-        return false;
-    }
-
-    private static boolean goalCanBeReplacedForAllFlags(WrappedGoal p_186079_, Map<Goal.Flag, WrappedGoal> p_186080_) {
-        for (Goal.Flag goal$flag : p_186079_.getFlags()) {
-            if (!p_186080_.getOrDefault(goal$flag, NO_GOAL).canBeReplacedBy(p_186079_)) {
+        availableGoalsDirty = true; // Leaf - Async target finding
+    }
+
+    // Paper start - Perf: optimize goal types
+    private static boolean goalContainsAnyFlags(WrappedGoal goal, ca.spottedleaf.moonrise.common.set.OptimizedSmallEnumSet<Goal.Flag> flags) {
+        return goal.getFlags().hasCommonElements(flags);
+    }
+
+    private static boolean goalCanBeReplacedForAllFlags(WrappedGoal goal, Map<Goal.Flag, WrappedGoal> flag) {
+        long flagIterator = goal.getFlags().getBackingSet();
+        int wrappedGoalSize = goal.getFlags().size();
+        for (int i = 0; i < wrappedGoalSize; ++i) {
+            final Goal.Flag flag1 = GOAL_FLAG_VALUES[Long.numberOfTrailingZeros(flagIterator)];
+            flagIterator ^= ca.spottedleaf.concurrentutil.util.IntegerUtil.getTrailingBit(flagIterator);
+            // Paper end - Perf: optimize goal types
+            if (!flag.getOrDefault(flag1, NO_GOAL).canBeReplacedBy(goal)) {
                 return false;
             }
         }
@@ -71,12 +_,135 @@
         return true;
     }
 
+    // Leaf start - Async target finding
+    public final boolean poll() {
+        if (this.ctxGoals == null || ctx.wake != null) {
+            return false;
+        }
+        if (ctxState == 0) {
+            while (ctxIndex < this.ctxGoals.length) {
+                WrappedGoal goal = this.ctxGoals[ctxIndex];
+                // entity tempt
+                if (goal.isRunning() && (goalContainsAnyFlags(goal, this.goalTypes) || !goal.canContinueToUse())) {
+                    ctx.state = false;
+                    if (ctx.wake != null) {
+                        return true;
+                    }
+                    goal.stop();
+                }
+
+                ctxIndex++;
+                ctx.state = true;
+            }
+
+            this.lockedFlags.entrySet().removeIf(entry -> !entry.getValue().isRunning());
+
+            ctxIndex = 0;
+            ctx.state = true;
+            ctxState = 1;
+        }
+        if (ctxState == 1) {
+            while (ctxIndex < this.ctxGoals.length) {
+                WrappedGoal goal = this.ctxGoals[ctxIndex];
+                var flags = goal.getFlags();
+                // entity and block
+                if (!goal.isRunning() && !flags.hasCommonElements(this.goalTypes)) {
+                    // inline
+                    boolean result = true;
+                    long flagIterator1 = flags.getBackingSet();
+                    int wrappedGoalSize1 = flags.size();
+                    for (int i1 = 0; i1 < wrappedGoalSize1; ++i1) {
+                        final Goal.Flag flag1 = GOAL_FLAG_VALUES[Long.numberOfTrailingZeros(flagIterator1)];
+                        flagIterator1 ^= ca.spottedleaf.concurrentutil.util.IntegerUtil.getTrailingBit(flagIterator1);
+                        if (!this.lockedFlags.getOrDefault(flag1, NO_GOAL).canBeReplacedBy(goal)) {
+                            result = false;
+                            break;
+                        }
+                    }
+                    if (result) {
+                        if (goal.canUse()) {
+                            long flagIterator = flags.getBackingSet();
+                            int wrappedGoalSize = flags.size();
+                            for (int i = 0; i < wrappedGoalSize; ++i) {
+                                final Goal.Flag flag = GOAL_FLAG_VALUES[Long.numberOfTrailingZeros(flagIterator)];
+                                flagIterator ^= ca.spottedleaf.concurrentutil.util.IntegerUtil.getTrailingBit(flagIterator);
+                                WrappedGoal wrappedGoal1 = this.lockedFlags.getOrDefault(flag, NO_GOAL);
+                                wrappedGoal1.stop();
+                                this.lockedFlags.put(flag, goal);
+                            }
+
+                            goal.start();
+                        }
+                        ctx.state = false;
+                        if (ctx.wake != null) {
+                            return true;
+                        }
+                    }
+                }
+                // entity alert other
+                if (!ctx.state) {
+                    switch (goal.getGoal()) {
+                        case net.minecraft.world.entity.ai.goal.target.HurtByTargetGoal t -> t.poll();
+                        case net.minecraft.world.entity.ai.goal.target.ResetUniversalAngerTargetGoal t -> t.poll();
+                        default -> {}
+                    }
+                }
+                ctxIndex++;
+                ctx.state = true;
+            }
+
+            ctxIndex = 0;
+            ctx.state = true;
+            ctxState = 2;
+        }
+        if (ctxState == 2) {
+            while (ctxIndex < this.ctxGoals.length) {
+                WrappedGoal goal = this.ctxGoals[ctxIndex];
+                if (goal.isRunning()) {
+                    goal.tick();
+                }
+                ctxIndex++;
+            }
+
+            ctxState = -1;
+            ctxIndex = 0;
+            ctx.state = true;
+        }
+        if (ctxState == 3) {
+            while (ctxIndex < this.ctxGoals.length) {
+                WrappedGoal goal = this.ctxGoals[ctxIndex];
+                if (goal.isRunning() && goal.requiresUpdateEveryTick()) {
+                    goal.tick();
+                }
+                ctxIndex++;
+            }
+
+            ctxState = -1;
+            ctxIndex = 0;
+            ctx.state = true;
+        }
+        return false;
+    }
+    // Leaf end - Async target finding
+
     public void tick() {
+        // Leaf start - Async target finding
+        if (GlobalConfiguration.get().asyncTargetFinding.enabled) {
+            if (ctxState == -1) {
+                if (availableGoalsDirty || this.ctxGoals == null) {
+                    this.ctxGoals = this.availableGoals.toArray(new WrappedGoal[0]);
+                    availableGoalsDirty = false;
+                }
+                ctxState = 0;
+            }
+            return;
+        }
+        // Leaf end - Async target finding
         ProfilerFiller profilerfiller = this.profiler.get();
         profilerfiller.push("goalCleanup");
 
         for (WrappedGoal wrappedgoal : this.availableGoals) {
-            if (wrappedgoal.isRunning() && (goalContainsAnyFlags(wrappedgoal, this.disabledFlags) || !wrappedgoal.canContinueToUse())) {
+            if (wrappedgoal.isRunning() && (goalContainsAnyFlags(wrappedgoal, this.goalTypes) || !wrappedgoal.canContinueToUse())) { // Paper - Perf: optimize goal types by removing streams
                 wrappedgoal.stop();
             }
         }
@@ -85,18 +_,21 @@
         profilerfiller.pop();
         profilerfiller.push("goalUpdate");
 
-        for (WrappedGoal wrappedgoal2 : this.availableGoals) {
-            if (!wrappedgoal2.isRunning()
-                && !goalContainsAnyFlags(wrappedgoal2, this.disabledFlags)
-                && goalCanBeReplacedForAllFlags(wrappedgoal2, this.lockedFlags)
-                && wrappedgoal2.canUse()) {
-                for (Goal.Flag goal$flag : wrappedgoal2.getFlags()) {
-                    WrappedGoal wrappedgoal1 = this.lockedFlags.getOrDefault(goal$flag, NO_GOAL);
+        for (WrappedGoal wrappedGoalx : this.availableGoals) {
+            // Paper start
+            if (!wrappedGoalx.isRunning() && !goalContainsAnyFlags(wrappedGoalx, this.goalTypes) && goalCanBeReplacedForAllFlags(wrappedGoalx, this.lockedFlags) && wrappedGoalx.canUse()) {
+                long flagIterator = wrappedGoalx.getFlags().getBackingSet();
+                int wrappedGoalSize = wrappedGoalx.getFlags().size();
+                for (int i = 0; i < wrappedGoalSize; ++i) {
+                    final Goal.Flag flag = GOAL_FLAG_VALUES[Long.numberOfTrailingZeros(flagIterator)];
+                    flagIterator ^= ca.spottedleaf.concurrentutil.util.IntegerUtil.getTrailingBit(flagIterator);
+                    // Paper end
+                    WrappedGoal wrappedgoal1 = this.lockedFlags.getOrDefault(flag, NO_GOAL);
                     wrappedgoal1.stop();
-                    this.lockedFlags.put(goal$flag, wrappedgoal2);
+                    this.lockedFlags.put(flag, wrappedGoalx);
                 }
 
-                wrappedgoal2.start();
+                wrappedGoalx.start();
             }
         }
 
@@ -107,7 +_,18 @@
     public void tickRunningGoals(boolean p_186082_) {
         ProfilerFiller profilerfiller = this.profiler.get();
         profilerfiller.push("goalTick");
-
+        // Leaf start - Async target finding
+        if (GlobalConfiguration.get().asyncTargetFinding.enabled) {
+            if (ctxState == -1) {
+                if (availableGoalsDirty || this.ctxGoals == null) {
+                    this.ctxGoals = this.availableGoals.toArray(new WrappedGoal[0]);
+                    availableGoalsDirty = false;
+                }
+                ctxState = p_186082_ ? 2 : 3;
+            }
+            return;
+        }
+        // Leaf end - Async target finding
         for (WrappedGoal wrappedgoal : this.availableGoals) {
             if (wrappedgoal.isRunning() && (p_186082_ || wrappedgoal.requiresUpdateEveryTick())) {
                 wrappedgoal.tick();
@@ -122,11 +_,11 @@
     }
 
     public void disableControlFlag(Goal.Flag p_25356_) {
-        this.disabledFlags.add(p_25356_);
+        this.goalTypes.addUnchecked(p_25356_); // Paper - remove streams from GoalSelector
     }
 
     public void enableControlFlag(Goal.Flag p_25375_) {
-        this.disabledFlags.remove(p_25375_);
+        this.goalTypes.removeUnchecked(p_25375_); // Paper - remove streams from GoalSelector
     }
 
     public void setControlFlag(Goal.Flag p_25361_, boolean p_25362_) {
