--- a/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
@@ -2,6 +_,9 @@
 
 import java.util.EnumSet;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.LivingEntity;
@@ -47,37 +_,123 @@
 
     @Override
     public boolean canUse() {
+        // Leaf start - Async look finding
+        if (poll()) {
+            return true;
+        }
+
         if (this.mob.getRandom().nextFloat() >= this.probability) {
             return false;
-        } else {
-            if (this.mob.getTarget() != null) {
-                this.lookAt = this.mob.getTarget();
+        }
+        if (this.lookAtType == Player.class) {
+            if (GlobalConfiguration.get().asyncTargetFinding.searchPlayer) {
+                getLookAsync();
+                return false;
             }
-
-            if (this.lookAtType == Player.class) {
-                this.lookAt = this.mob.level().getNearestPlayer(this.lookAtContext, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
-            } else {
-                this.lookAt = this.mob
-                    .level()
-                    .getNearestEntity(
-                        this.mob
+        } else if (GlobalConfiguration.get().asyncTargetFinding.searchEntity) {
+            getLookAsync();
+            return false;
+        }
+
+        if (this.mob.getTarget() != null) {
+            this.lookAt = this.mob.getTarget();
+        }
+
+        ServerLevel serverLevel = getServerLevel(this.mob);
+        if (this.lookAtType == Player.class) {
+            this.lookAt = serverLevel.getNearestPlayer(this.lookAtContext, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
+        } else {
+            this.lookAt = serverLevel.getNearestEntity(
+                    this.mob
                             .level()
-                            .getEntitiesOfClass(
-                                this.lookAtType,
-                                this.mob.getBoundingBox().inflate((double)this.lookDistance, 3.0, (double)this.lookDistance),
-                                p_148124_ -> true
-                            ),
-                        this.lookAtContext,
-                        this.mob,
-                        this.mob.getX(),
-                        this.mob.getEyeY(),
-                        this.mob.getZ()
+                            .getEntitiesOfClass(this.lookAtType, this.mob.getBoundingBox().inflate(this.lookDistance, 3.0, this.lookDistance), livingEntity -> true),
+                    this.lookAtContext,
+                    this.mob,
+                    this.mob.getX(),
+                    this.mob.getEyeY(),
+                    this.mob.getZ()
+            );
+        }
+
+        return this.lookAt != null;
+        // Leaf end - Async look finding
+    }
+
+    // Leaf start - Async look finding
+    private final java.util.concurrent.atomic.AtomicBoolean isSearching = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private final java.util.concurrent.atomic.AtomicReference<LivingEntity> pendingTarget = new java.util.concurrent.atomic.AtomicReference<>(null);
+    private static final org.apache.logging.log4j.Logger LOGGER = org.dreeam.leaf.async.ai.AsyncGoalExecutor.LOGGER;
+    protected boolean poll() {
+        LivingEntity t = pendingTarget.getAndSet(null);
+        if (t == null) {
+            return false;
+        }
+        if (this.mob.getTarget() != null) {
+            this.lookAt = this.mob.getTarget();
+            return true;
+        }
+        ServerLevel serverLevel = getServerLevel(this.mob);
+        if (!t.isAlive() || !this.lookAtContext.test(this.mob, t)) {
+            return false;
+        }
+        this.lookAt = t;
+        return true;
+    }
+
+    protected void getLookAsync() {
+        if (!isSearching.compareAndSet(false, true)) {
+            return;
+        }
+
+        final var mob = this.mob;
+
+        if (mob == null || mob.isRemoved() || !mob.isAlive()) {
+            isSearching.setRelease(false);
+            return;
+        }
+
+        final double x = mob.getX();
+        final double y = mob.getEyeY();
+        final double z = mob.getZ();
+        final var serverLevel = getServerLevel(mob);
+        final var bound = mob.getBoundingBox().inflate(this.lookDistance, 3.0, this.lookDistance);
+        final var target = mob.getTarget();
+        final var lookAtContext = this.lookAtContext;
+        final var lookAtType = this.lookAtType;
+        final var pendingTarget = this.pendingTarget;
+        serverLevel.asyncAITasks.add(() -> {
+            try {
+                if (mob == null || !mob.isAlive() || mob.level() != serverLevel) {
+                    return;
+                }
+                if (target != null) {
+                    pendingTarget.setRelease(target);
+                    return;
+                }
+
+                if (lookAtType == Player.class) {
+                    var result = serverLevel.getNearestPlayer(lookAtContext, mob, x, y, z);
+                    pendingTarget.setRelease(result);
+                } else {
+                    var result = serverLevel.getNearestEntity(
+                            serverLevel
+                                    .getEntitiesOfClass(lookAtType, bound, livingEntity -> true),
+                            lookAtContext,
+                            mob,
+                            x,
+                            y,
+                            z
                     );
+                    pendingTarget.setRelease(result);
+                }
+            } catch (Exception e) {
+                LOGGER.error("Exception getting entities", e);
+            } finally {
+                isSearching.setRelease(false);
             }
-
-            return this.lookAt != null;
-        }
+        });
     }
+    // Leaf end - Async look finding
 
     @Override
     public boolean canContinueToUse() {
