--- a/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
@@ -2,6 +_,9 @@
 
 import java.util.EnumSet;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.LivingEntity;
@@ -45,39 +_,82 @@
         }
     }
 
+
     @Override
     public boolean canUse() {
+        // Leaf start - Async look finding
+        if (poll()) {
+            return true;
+        }
         if (this.mob.getRandom().nextFloat() >= this.probability) {
             return false;
+        }
+        if (this.mob.getTarget() != null) {
+            this.lookAt = this.mob.getTarget();
+        }
+        if (GlobalConfiguration.get().asyncTargetFinding.searchEntity) {
+            getLookAsync();
+            return false;
+        }
+        ServerLevel serverLevel = getServerLevel(this.mob);
+        if (this.lookAtType == Player.class) {
+            this.lookAt = serverLevel.getNearestPlayer(this.lookAtContext, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
         } else {
-            if (this.mob.getTarget() != null) {
-                this.lookAt = this.mob.getTarget();
-            }
-
-            if (this.lookAtType == Player.class) {
-                this.lookAt = this.mob.level().getNearestPlayer(this.lookAtContext, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
-            } else {
-                this.lookAt = this.mob
-                    .level()
-                    .getNearestEntity(
-                        this.mob
+            this.lookAt = serverLevel.getNearestEntity(
+                    this.mob
                             .level()
-                            .getEntitiesOfClass(
-                                this.lookAtType,
-                                this.mob.getBoundingBox().inflate((double)this.lookDistance, 3.0, (double)this.lookDistance),
-                                p_148124_ -> true
-                            ),
-                        this.lookAtContext,
-                        this.mob,
-                        this.mob.getX(),
-                        this.mob.getEyeY(),
-                        this.mob.getZ()
-                    );
+                            .getEntitiesOfClass(this.lookAtType, this.mob.getBoundingBox().inflate(this.lookDistance, 3.0, this.lookDistance), livingEntity -> true),
+                    this.lookAtContext,
+                    this.mob,
+                    this.mob.getX(),
+                    this.mob.getEyeY(),
+                    this.mob.getZ()
+            );
+        }
+
+        return this.lookAt != null;
+    }
+
+    protected boolean poll() {
+        if (!(this.mob.getGoalCtx().result() instanceof LivingEntity target)) return false;
+        if (this.mob.getTarget() != null) {
+            this.lookAt = this.mob.getTarget();
+            return true;
+        }
+        ServerLevel serverLevel = getServerLevel(this.mob);
+        if (!target.isAlive() || !this.lookAtContext.test(this.mob, target)) return false;
+        this.lookAt = target;
+        return true;
+    }
+
+    protected void getLookAsync() {
+        final var mob = this.mob;
+        final var ctx = mob.getGoalCtx();
+        if (!ctx.state) return;
+        final double x = mob.getX();
+        final double y = mob.getEyeY();
+        final double z = mob.getZ();
+        final var serverLevel = getServerLevel(mob);
+        final var bound = mob.getBoundingBox().inflate(this.lookDistance, 3.0, this.lookDistance);
+        final var lookAtContext = this.lookAtContext;
+        final var lookAtType = this.lookAtType;
+        ctx.wake = () -> {
+            if (lookAtType == Player.class) {
+                return serverLevel.getNearestPlayer(lookAtContext, mob, x, y, z);
+            } else {
+                return serverLevel.getNearestEntity(
+                        serverLevel
+                                .getEntitiesOfClass(lookAtType, bound, livingEntity -> true),
+                        lookAtContext,
+                        mob,
+                        x,
+                        y,
+                        z
+                );
             }
-
-            return this.lookAt != null;
-        }
+        };
     }
+    // Leaf end - Async look finding
 
     @Override
     public boolean canContinueToUse() {
