--- a/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
@@ -24,6 +_,15 @@
         this(p_25609_, p_25610_, p_25611_, 1);
     }
 
+    // Paper start - activation range improvements
+    @Override
+    public void stop() {
+        super.stop();
+        this.blockPos = BlockPos.ZERO;
+        this.mob.movingTarget = null;
+    }
+    // Paper end
+
     public MoveToBlockGoal(PathfinderMob p_25613_, double p_25614_, int p_25615_, int p_25616_) {
         this.mob = p_25613_;
         this.speedModifier = p_25614_;
@@ -113,8 +_,10 @@
                 for (int i1 = 0; i1 <= l; i1 = i1 > 0 ? -i1 : 1 - i1) {
                     for (int j1 = i1 < l && i1 > -l ? l : 0; j1 <= l; j1 = j1 > 0 ? -j1 : 1 - j1) {
                         blockpos$mutableblockpos.setWithOffset(blockpos, i1, k - 1, j1);
+                        if (!this.mob.level().hasChunkAt(blockpos$mutableblockpos)) continue; // Gale - Airplane - block goal does not load chunks - if this block isn't loaded, continue
                         if (this.mob.isWithinRestriction(blockpos$mutableblockpos) && this.isValidTarget(this.mob.level(), blockpos$mutableblockpos)) {
                             this.blockPos = blockpos$mutableblockpos;
+                            this.mob.movingTarget = blockpos$mutableblockpos == BlockPos.ZERO ? null : blockpos$mutableblockpos.immutable(); // Paper
                             return true;
                         }
                     }
