--- a/net/minecraft/world/entity/ai/goal/OfferFlowerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/OfferFlowerGoal.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.entity.ai.goal;
 
 import java.util.EnumSet;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.animal.IronGolem;
 import net.minecraft.world.entity.npc.Villager;
@@ -19,10 +_,20 @@
 
     @Override
     public boolean canUse() {
+        // Leaf start - Async offer flower finding
         if (!this.golem.level().isDay()) {
             return false;
-        } else if (this.golem.getRandom().nextInt(8000) != 0) {
-            return false;
+        }
+        if (poll()) {
+            return true;
+        }
+        if (this.golem.getRandom().nextInt(8000) != 0) {
+            return false;
+        }
+        if (GlobalConfiguration.get().asyncTargetFinding.searchEntity) {
+            getVillagerAsync();
+            return false;
+            // Leaf end - Async offer flower finding
         } else {
             this.villager = this.golem
                 .level()
@@ -38,6 +_,35 @@
             return this.villager != null;
         }
     }
+
+    // Leaf start - Async look finding
+    protected boolean poll() {
+        if (!(this.golem.getGoalCtx().result() instanceof Villager target)) return false;
+        var serverLevel = getServerLevel(this.golem);
+        if (!target.isAlive() || !OFFER_TARGER_CONTEXT.test(this.golem, target)) return false;
+        this.villager = target;
+        return true;
+    }
+
+    protected void getVillagerAsync() {
+        final var golem = this.golem;
+        final var ctx = golem.getGoalCtx();
+        if (!ctx.state) return;
+        final double x = golem.getX();
+        final double y = golem.getEyeY();
+        final double z = golem.getZ();
+        final var serverLevel = getServerLevel(golem);
+        final var bound = golem.getBoundingBox().inflate(6.0, 2.0, 6.0);
+        ctx.wake = () -> serverLevel.getNearestEntity(
+                serverLevel.getEntitiesOfClass(Villager.class, bound, livingEntity -> true),
+                OFFER_TARGER_CONTEXT,
+                golem,
+                x,
+                y,
+                z
+        );
+    }
+    // Leaf end - Async look finding
 
     @Override
     public boolean canContinueToUse() {
