--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -3,8 +_,11 @@
 import java.util.EnumSet;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.PathfinderMob;
+import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
@@ -35,13 +_,77 @@
         this.targetingConditions = TEMP_TARGETING.copy().selector(this::shouldFollow);
     }
 
+    // Leaf start - Async Tempt Finding
+    private final java.util.concurrent.atomic.AtomicBoolean isSearching = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private final java.util.concurrent.atomic.AtomicReference<Player> pendingTarget = new java.util.concurrent.atomic.AtomicReference<>(null);
+    private static final org.apache.logging.log4j.Logger LOGGER = org.dreeam.leaf.async.ai.AsyncGoalExecutor.LOGGER;
+
+    private boolean poll() {
+        Player t = pendingTarget.getAndSet(null);
+        if (t == null) {
+            t = this.player;
+        }
+        if (t == null) {
+            return false;
+        }
+        var serverLevel = getServerLevel(this.mob);
+        if (serverLevel == null || !t.isAlive() || !this.targetingConditions.test(this.mob, t)) {
+            return false;
+        }
+        this.player = t;
+        return true;
+    }
+
+    private void getNearestPlayerAsync() {
+        if (!isSearching.compareAndSet(false, true)) {
+            return;
+        }
+
+        final var mob = this.mob;
+
+        if (mob == null || mob.isRemoved() || !mob.isAlive()) {
+            isSearching.setRelease(false);
+            return;
+        }
+
+        final var serverLevel = getServerLevel(mob);
+        final var pendingTarget = this.pendingTarget;
+        final var conditions = this.targetingConditions
+                .copy();
+        serverLevel.asyncAITasks.add(() -> {
+            try {
+                if (mob == null || mob.isRemoved() || !mob.isAlive() || mob.level() != serverLevel) {
+                    return;
+                }
+                pendingTarget.setRelease(serverLevel.getNearestPlayer(conditions, mob));
+            } catch (Exception e) {
+                LOGGER.error("Exception getting player", e);
+            } finally {
+                isSearching.setRelease(false);
+            }
+        });
+    }
+    // Leaf end - Async Tempt Finding
+
     @Override
     public boolean canUse() {
         if (this.calmDown > 0) {
             this.calmDown--;
             return false;
         } else {
-            this.player = this.mob.level().getNearestPlayer(this.targetingConditions, this.mob);
+            // Leaf start - Async Tempt Finding
+            if (GlobalConfiguration.get().asyncTargetFinding.searchPlayerTempt) {
+                if (poll()) {
+                    return this.player != null;
+                } else {
+                    getNearestPlayerAsync();
+                    return false;
+                }
+            } else {
+                this.player = getServerLevel(this.mob)
+                        .getNearestPlayer(this.targetingConditions, this.mob);
+            }
+            // Leaf end - Async Tempt Finding
             return this.player != null;
         }
     }
