--- a/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -3,8 +_,11 @@
 import java.util.EnumSet;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.PathfinderMob;
+import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
@@ -35,12 +_,44 @@
         this.targetingConditions = TEMP_TARGETING.copy().selector(this::shouldFollow);
     }
 
+    // Leaf start - Async Tempt Finding
+    private boolean poll() {
+        if (!(this.mob.getGoalCtx().result() instanceof Player target)) return false;
+        var serverLevel = getServerLevel(this.mob);
+        if (!target.isAlive() || !this.targetingConditions.test(this.mob, target)) return false;
+        this.player = target;
+        return true;
+    }
+
+    private void getNearestPlayerAsync() {
+        final var mob = this.mob;
+        final var ctx = mob.getGoalCtx();
+        if (!ctx.state) return;
+        final var serverLevel = getServerLevel(mob);
+        final var conditions = this.targetingConditions
+                .copy();
+        ctx.wake = () -> serverLevel.getNearestPlayer(conditions, mob);
+    }
+    // Leaf end - Async Tempt Finding
+
     @Override
     public boolean canUse() {
         if (this.calmDown > 0) {
             this.calmDown--;
             return false;
         } else {
+            // Leaf start - Async Tempt Finding
+            if (GlobalConfiguration.get().asyncTargetFinding.searchEntity) {
+                if (poll()) {
+                    if (this.player != null) {
+                        return true;
+                    }
+                } else {
+                    getNearestPlayerAsync();
+                    return false;
+                }
+            }
+            // Leaf end - Async Tempt Finding
             this.player = this.mob.level().getNearestPlayer(this.targetingConditions, this.mob);
             return this.player != null;
         }
