--- a/net/minecraft/util/thread/BlockableEventLoop.java
+++ b/net/minecraft/util/thread/BlockableEventLoop.java
@@ -32,9 +_,11 @@
 
     protected abstract boolean shouldRun(R p_18703_);
 
+    // Paper start - rewrite chunk system
     public boolean isSameThread() {
-        return Thread.currentThread() == this.getRunningThread();
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThread();
     }
+    // Paper end - rewrite chunk system
 
     protected abstract Thread getRunningThread();
 
@@ -56,10 +_,15 @@
     }
 
     public CompletableFuture<Void> submitAsync(Runnable p_18690_) {
-        return CompletableFuture.supplyAsync(() -> {
+        return CompletableFuture.<Void>supplyAsync(() -> {
             p_18690_.run();
             return null;
-        }, this);
+        }, this)
+        .exceptionallyCompose(ex -> {
+            // Neo: Log since this is usually swallowed
+            LOGGER.error(LogUtils.FATAL_MARKER, "Error executing task on {}", name(), ex);
+            return CompletableFuture.failedStage(ex);
+        });
     }
 
     @CheckReturnValue
