--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -164,9 +_,9 @@
    }
 
    @Nullable
-   public Entity m_5489_(ServerLevel p_35295_) {
+   public Entity changeDimension(ServerLevel p_35295_, net.minecraftforge.common.util.ITeleporter teleporter) {
       this.m_7996_();
-      return super.m_5489_(p_35295_);
+      return super.changeDimension(p_35295_, teleporter);
    }
 
    protected void m_7996_() {
@@ -204,6 +_,8 @@
    protected abstract void m_7604_();
 
    protected void m_35277_(MerchantOffers p_35278_, VillagerTrades.ItemListing[] p_35279_, int p_35280_) {
+      if (!this.isValid)
+         return;
       Set<Integer> set = Sets.newHashSet();
       if (p_35279_.length > p_35280_) {
          while(set.size() < p_35280_) {
