--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -86,6 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.phys.AABB;
+import org.goldenforge.GoldenConfig;
 import org.slf4j.Logger;
 
 public class Villager extends AbstractVillager implements ReputationEventHandler, VillagerDataHolder {
@@ -118,6 +_,8 @@
    private int f_35364_;
    private long f_35365_;
    private boolean f_35366_;
+
+   public long nextGolemPanic = -1; // Pufferfish
    private static final ImmutableList<MemoryModuleType<?>> f_35367_ = ImmutableList.of(MemoryModuleType.f_26359_, MemoryModuleType.f_26360_, MemoryModuleType.f_26361_, MemoryModuleType.f_26362_, MemoryModuleType.f_148204_, MemoryModuleType.f_148205_, MemoryModuleType.f_26366_, MemoryModuleType.f_26367_, MemoryModuleType.f_26368_, MemoryModuleType.f_148206_, MemoryModuleType.f_26332_, MemoryModuleType.f_26370_, MemoryModuleType.f_26371_, MemoryModuleType.f_26374_, MemoryModuleType.f_26375_, MemoryModuleType.f_26377_, MemoryModuleType.f_26379_, MemoryModuleType.f_26380_, MemoryModuleType.f_26381_, MemoryModuleType.f_26382_, MemoryModuleType.f_26323_, MemoryModuleType.f_26363_, MemoryModuleType.f_26324_, MemoryModuleType.f_26325_, MemoryModuleType.f_26326_, MemoryModuleType.f_26328_, MemoryModuleType.f_26329_, MemoryModuleType.f_26330_, MemoryModuleType.f_26327_);
    private static final ImmutableList<SensorType<? extends Sensor<? super Villager>>> f_35368_ = ImmutableList.of(SensorType.f_26811_, SensorType.f_26812_, SensorType.f_26810_, SensorType.f_26813_, SensorType.f_26814_, SensorType.f_26815_, SensorType.f_26816_, SensorType.f_26817_, SensorType.f_26818_);
    public static final Map<MemoryModuleType<GlobalPos>, BiPredicate<Villager, Holder<PoiType>>> f_35370_ = ImmutableMap.of(MemoryModuleType.f_26359_, (p_219625_, p_219626_) -> {
@@ -203,9 +_,40 @@
       return this.f_35366_;
    }
 
+   // Spigot Start
+   @Override
+   public void inactiveTick() {
+      // SPIGOT-3874, SPIGOT-3894, SPIGOT-3846, SPIGOT-5286 :(
+      // Paper start
+      if (this.m_35303_() > 0) {
+         this.m_35319_(this.m_35303_() - 1);
+      }
+      if (this.m_6142_()) {
+         if (GoldenConfig.tickInactiveVillagers) {
+            this.m_8024_();
+         } else {
+            this.mobTick(true);
+         }
+      }
+      m_35530_();
+      // Paper end
+
+      super.inactiveTick();
+   }
+   // Spigot End
+
    protected void m_8024_() {
+      mobTick(false);
+   }
+
+   private int behaviorTick = 0; // Pufferfish
+   protected void mobTick(boolean inactive) {
       this.f_19853_.m_46473_().m_6180_("villagerBrain");
-      this.m_6274_().m_21865_((ServerLevel)this.f_19853_, this);
+      // Pufferfish start
+      if (!inactive && this.behaviorTick++ % this.activatedPriority == 0) {
+         this.m_6274_().m_21865_((ServerLevel) this.f_19853_, this); // Paper
+      }
+      // Pufferfish end
       this.f_19853_.m_46473_().m_7238_();
       if (this.f_35366_) {
          this.f_35366_ = false;
@@ -229,7 +_,7 @@
          this.f_35375_ = null;
       }
 
-      if (!this.m_21525_() && this.f_19796_.m_188503_(100) == 0) {
+      if (!inactive && !this.m_21525_() && this.f_19796_.m_188503_(100) == 0) { // Paper
          Raid raid = ((ServerLevel)this.f_19853_).m_8832_(this.m_20183_());
          if (raid != null && raid.m_37782_() && !raid.m_37706_()) {
             this.f_19853_.m_7605_(this, (byte)42);
@@ -240,6 +_,8 @@
          this.m_7996_();
       }
 
+      if (inactive) return; // Paper
+
       super.m_8024_();
    }
 
@@ -254,7 +_,7 @@
 
    public InteractionResult m_6071_(Player p_35472_, InteractionHand p_35473_) {
       ItemStack itemstack = p_35472_.m_21120_(p_35473_);
-      if (!itemstack.m_150930_(Items.f_42601_) && this.m_6084_() && !this.m_35306_() && !this.m_5803_()) {
+      if (itemstack.m_41720_() != Items.f_42601_ && this.m_6084_() && !this.m_35306_() && !this.m_5803_() && !p_35472_.m_36341_()) {
          if (this.m_6162_()) {
             this.m_35518_();
             return InteractionResult.m_19078_(this.f_19853_.f_46443_);
@@ -657,7 +_,8 @@
    }
 
    protected Component m_5677_() {
-      return Component.m_237115_(this.m_6095_().m_20675_() + "." + Registry.f_122869_.m_7981_(this.m_7141_().m_35571_()).m_135815_());
+      net.minecraft.resources.ResourceLocation profName = net.minecraftforge.registries.ForgeRegistries.VILLAGER_PROFESSIONS.getKey(this.m_7141_().m_35571_());
+      return Component.m_237115_(this.m_6095_().m_20675_() + '.' + (!"minecraft".equals(profName.m_135827_()) ? profName.m_135827_() + '.' : "") + profName.m_135815_());
    }
 
    public void m_7822_(byte p_35391_) {
@@ -709,7 +_,7 @@
    }
 
    public void m_8038_(ServerLevel p_35409_, LightningBolt p_35410_) {
-      if (p_35409_.m_46791_() != Difficulty.PEACEFUL) {
+      if (p_35409_.m_46791_() != Difficulty.PEACEFUL && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.f_20495_, (timer) -> {})) {
          f_201976_.info("Villager {} was struck by lightning {}.", this, p_35410_);
          Witch witch = EntityType.f_20495_.m_20615_(p_35409_);
          witch.m_7678_(this.m_20185_(), this.m_20186_(), this.m_20189_(), this.m_146908_(), this.m_146909_());
@@ -721,6 +_,7 @@
          }
 
          witch.m_21530_();
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, witch);
          p_35409_.m_47205_(witch);
          this.m_35524_();
          this.m_146870_();
