--- a/net/minecraft/world/level/Explosion.java
+++ b/net/minecraft/world/level/Explosion.java
@@ -65,92 +_,92 @@
     }
 
     public Explosion(
-        Level p_46051_,
-        @Nullable Entity p_46052_,
-        double p_46055_,
-        double p_46056_,
-        double p_46057_,
-        float p_46058_,
-        List<BlockPos> p_312600_,
-        Explosion.BlockInteraction p_46060_,
-        ParticleOptions p_312560_,
-        ParticleOptions p_312844_,
-        Holder<SoundEvent> p_320054_
+            Level p_46051_,
+            @Nullable Entity p_46052_,
+            double p_46055_,
+            double p_46056_,
+            double p_46057_,
+            float p_46058_,
+            List<BlockPos> p_312600_,
+            Explosion.BlockInteraction p_46060_,
+            ParticleOptions p_312560_,
+            ParticleOptions p_312844_,
+            Holder<SoundEvent> p_320054_
     ) {
         this(
-            p_46051_,
-            p_46052_,
-            getDefaultDamageSource(p_46051_, p_46052_),
-            null,
-            p_46055_,
-            p_46056_,
-            p_46057_,
-            p_46058_,
-            false,
-            p_46060_,
-            p_312560_,
-            p_312844_,
-            p_320054_
+                p_46051_,
+                p_46052_,
+                getDefaultDamageSource(p_46051_, p_46052_),
+                null,
+                p_46055_,
+                p_46056_,
+                p_46057_,
+                p_46058_,
+                false,
+                p_46060_,
+                p_312560_,
+                p_312844_,
+                p_320054_
         );
         this.toBlow.addAll(p_312600_);
     }
 
     public Explosion(
-        Level p_46041_,
-        @Nullable Entity p_46042_,
-        double p_46043_,
-        double p_46044_,
-        double p_46045_,
-        float p_46046_,
-        boolean p_46047_,
-        Explosion.BlockInteraction p_46048_,
-        List<BlockPos> p_46049_
+            Level p_46041_,
+            @Nullable Entity p_46042_,
+            double p_46043_,
+            double p_46044_,
+            double p_46045_,
+            float p_46046_,
+            boolean p_46047_,
+            Explosion.BlockInteraction p_46048_,
+            List<BlockPos> p_46049_
     ) {
         this(p_46041_, p_46042_, p_46043_, p_46044_, p_46045_, p_46046_, p_46047_, p_46048_);
         this.toBlow.addAll(p_46049_);
     }
 
     public Explosion(
-        Level p_46032_,
-        @Nullable Entity p_46033_,
-        double p_46034_,
-        double p_46035_,
-        double p_46036_,
-        float p_46037_,
-        boolean p_46038_,
-        Explosion.BlockInteraction p_46039_
+            Level p_46032_,
+            @Nullable Entity p_46033_,
+            double p_46034_,
+            double p_46035_,
+            double p_46036_,
+            float p_46037_,
+            boolean p_46038_,
+            Explosion.BlockInteraction p_46039_
     ) {
         this(
-            p_46032_,
-            p_46033_,
-            getDefaultDamageSource(p_46032_, p_46033_),
-            null,
-            p_46034_,
-            p_46035_,
-            p_46036_,
-            p_46037_,
-            p_46038_,
-            p_46039_,
-            ParticleTypes.EXPLOSION,
-            ParticleTypes.EXPLOSION_EMITTER,
-            SoundEvents.GENERIC_EXPLODE
+                p_46032_,
+                p_46033_,
+                getDefaultDamageSource(p_46032_, p_46033_),
+                null,
+                p_46034_,
+                p_46035_,
+                p_46036_,
+                p_46037_,
+                p_46038_,
+                p_46039_,
+                ParticleTypes.EXPLOSION,
+                ParticleTypes.EXPLOSION_EMITTER,
+                SoundEvents.GENERIC_EXPLODE
         );
     }
 
     public Explosion(
-        Level p_46024_,
-        @Nullable Entity p_46025_,
-        @Nullable DamageSource p_312268_,
-        @Nullable ExplosionDamageCalculator p_312205_,
-        double p_46026_,
-        double p_46027_,
-        double p_46028_,
-        float p_46029_,
-        boolean p_312333_,
-        Explosion.BlockInteraction p_312294_,
-        ParticleOptions p_312158_,
-        ParticleOptions p_311904_,
-        Holder<SoundEvent> p_320270_
+            Level p_46024_,
+            @Nullable Entity p_46025_,
+            @Nullable DamageSource p_312268_,
+            @Nullable ExplosionDamageCalculator p_312205_,
+            double p_46026_,
+            double p_46027_,
+            double p_46028_,
+            float p_46029_,
+            boolean p_312333_,
+            Explosion.BlockInteraction p_312294_,
+            ParticleOptions p_312158_,
+            ParticleOptions p_311904_,
+            Holder<SoundEvent> p_320270_
     ) {
         this.level = p_46024_;
         this.source = p_46025_;
@@ -190,7 +_,7 @@
                         double d10 = Mth.lerp(d7, aabb.minZ, aabb.maxZ);
                         Vec3 vec3 = new Vec3(d8 + d3, d9, d10 + d4);
                         if (p_46066_.level().clip(new ClipContext(vec3, p_46065_, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, p_46066_)).getType()
-                            == HitResult.Type.MISS) {
+                                == HitResult.Type.MISS) {
                             i++;
                         }
 
@@ -319,16 +_,16 @@
     public void finalizeExplosion(boolean p_46076_) {
         if (this.level.isClientSide) {
             this.level
-                .playLocalSound(
-                    this.x,
-                    this.y,
-                    this.z,
-                    this.explosionSound.value(),
-                    SoundSource.BLOCKS,
-                    4.0F,
-                    (1.0F + (this.level.random.nextFloat() - this.level.random.nextFloat()) * 0.2F) * 0.7F,
-                    false
-                );
+                    .playLocalSound(
+                            this.x,
+                            this.y,
+                            this.z,
+                            this.explosionSound.value(),
+                            SoundSource.BLOCKS,
+                            4.0F,
+                            (1.0F + (this.level.random.nextFloat() - this.level.random.nextFloat()) * 0.2F) * 0.7F,
+                            false
+                    );
         }
 
         boolean flag = this.interactsWithBlocks();
@@ -350,8 +_,8 @@
 
             for (BlockPos blockpos : this.toBlow) {
                 this.level
-                    .getBlockState(blockpos)
-                    .onExplosionHit(this.level, blockpos, this, (p_311741_, p_311742_) -> addOrAppendStack(list, p_311741_, p_311742_));
+                        .getBlockState(blockpos)
+                        .onExplosionHit(this.level, blockpos, this, (p_311741_, p_311742_) -> addOrAppendStack(list, p_311741_, p_311742_));
             }
 
             for (Pair<ItemStack, BlockPos> pair : list) {
@@ -364,8 +_,8 @@
         if (this.fire) {
             for (BlockPos blockpos1 : this.toBlow) {
                 if (this.random.nextInt(3) == 0
-                    && this.level.getBlockState(blockpos1).isAir()
-                    && this.level.getBlockState(blockpos1.below()).isSolidRender(this.level, blockpos1.below())) {
+                        && this.level.getBlockState(blockpos1).isAir()
+                        && this.level.getBlockState(blockpos1.below()).isSolidRender(this.level, blockpos1.below())) {
                     this.level.setBlockAndUpdate(blockpos1, BaseFireBlock.getState(this.level, blockpos1));
                 }
             }
@@ -452,8 +_,8 @@
     public boolean canTriggerBlocks() {
         if (this.blockInteraction == Explosion.BlockInteraction.TRIGGER_BLOCK && !this.level.isClientSide()) {
             return this.source != null && this.source.getType() == EntityType.BREEZE_WIND_CHARGE
-                ? this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
-                : true;
+                    ? this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+                    : true;
         } else {
             return false;
         }
