--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -52,7 +_,7 @@
     public static final int SPAWN_DISTANCE_CHUNK = 8;
     public static final int SPAWN_DISTANCE_BLOCK = 128;
     static final int MAGIC_NUMBER = (int)Math.pow(17.0, 2.0);
-    private static final MobCategory[] SPAWNING_CATEGORIES = Stream.of(MobCategory.values())
+    public static final MobCategory[] SPAWNING_CATEGORIES = Stream.of(MobCategory.values())
         .filter(p_47037_ -> p_47037_ != MobCategory.MISC)
         .toArray(MobCategory[]::new);
 
@@ -83,7 +_,7 @@
                             potentialcalculator.addCharge(entity.blockPosition(), mobspawnsettings$mobspawncost.charge());
                         }
 
-                        if (entity instanceof Mob) {
+                        if (p_186528_ != null && entity instanceof Mob) {
                             p_186528_.addMob(p_275163_.getPos(), mobcategory);
                         }
 
@@ -145,8 +_,8 @@
         StructureManager structuremanager = p_47040_.structureManager();
         ChunkGenerator chunkgenerator = p_47040_.getChunkSource().getGenerator();
         int i = p_47042_.getY();
-        BlockState blockstate = p_47041_.getBlockState(p_47042_);
-        if (!blockstate.isRedstoneConductor(p_47041_, p_47042_)) {
+        BlockState blockstate = p_47040_.getBlockStateIfLoadedAndInBounds(p_47042_); // Paper - don't load chunks for mob spawn
+        if (blockstate != null && !blockstate.isRedstoneConductor(p_47041_, p_47042_)) { // Paper - don't load chunks for mob spawn
             BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
             int j = 0;
 
@@ -168,7 +_,7 @@
                     Player player = p_47040_.getNearestPlayer(d0, (double)i, d1, -1.0, false);
                     if (player != null) {
                         double d2 = player.distanceToSqr(d0, (double)i, d1);
-                        if (isRightDistanceToPlayerAndSpawnPoint(p_47040_, p_47041_, blockpos$mutableblockpos, d2)) {
+                        if (p_47040_.isLoadedAndInBounds(blockpos$mutableblockpos) && isRightDistanceToPlayerAndSpawnPoint(p_47040_, p_47041_, blockpos$mutableblockpos, d2)) { // Paper - don't load chunks for mob spawn
                             if (mobspawnsettings$spawnerdata == null) {
                                 Optional<MobSpawnSettings.SpawnerData> optional = getRandomSpawnMobAt(
                                     p_47040_, structuremanager, chunkgenerator, p_47039_, p_47040_.random, blockpos$mutableblockpos
@@ -510,7 +_,7 @@
             this.spawnPotential.addCharge(blockpos, d0);
             MobCategory mobcategory = entitytype.getCategory();
             this.mobCategoryCounts.addTo(mobcategory, 1);
-            this.localMobCapCalculator.addMob(new ChunkPos(blockpos), mobcategory);
+            if (this.localMobCapCalculator != null) this.localMobCapCalculator.addMob(new ChunkPos(blockpos), mobcategory);
         }
 
         public int getSpawnableChunkCount() {
@@ -523,6 +_,7 @@
 
         boolean canSpawnForCategory(MobCategory p_186549_, ChunkPos p_186550_) {
             int i = p_186549_.getMaxInstancesPerChunk() * this.spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
+            if (this.localMobCapCalculator == null) return this.mobCategoryCounts.getInt(p_186549_) < i; // Paper - Optional per player mob spawns
             return this.mobCategoryCounts.getInt(p_186549_) >= i ? false : this.localMobCapCalculator.canSpawn(p_186549_, p_186550_);
         }
     }
