--- a/net/minecraft/world/level/PathNavigationRegion.java
+++ b/net/minecraft/world/level/PathNavigationRegion.java
@@ -32,6 +_,34 @@
     protected final Level level;
     private final Supplier<Holder<Biome>> plains;
 
+    // Paper start - if loaded util
+    @Nullable
+    private ChunkAccess getChunkIfLoaded(int x, int z) {
+        // Based on getChunk(int, int)
+        int xx = x - this.centerX;
+        int zz = z - this.centerZ;
+
+        if (xx >= 0 && xx < this.chunks.length && zz >= 0 && zz < this.chunks[xx].length) {
+            return this.chunks[xx][zz];
+        }
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public final FluidState getFluidIfLoaded(BlockPos pos) {
+        ChunkAccess chunk = getChunkIfLoaded(pos.getX() >> 4, pos.getZ() >> 4);
+        return chunk == null ? null : chunk.getFluidState(pos);
+    }
+
+    @Nullable
+    @Override
+    public final BlockState getBlockStateIfLoaded(BlockPos pos) {
+        ChunkAccess chunk = getChunkIfLoaded(pos.getX() >> 4, pos.getZ() >> 4);
+        return chunk == null ? null : chunk.getBlockState(pos);
+    }
+    // Paper end
+
     public PathNavigationRegion(Level p_47164_, BlockPos p_47165_, BlockPos p_47166_) {
         this.level = p_47164_;
         this.plains = Suppliers.memoize(() -> p_47164_.registryAccess().registryOrThrow(Registries.BIOME).getHolderOrThrow(Biomes.PLAINS));
