--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -7,6 +_,8 @@
 import java.util.Optional;
 import java.util.function.BiConsumer;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.particles.BlockParticleOption;
@@ -278,11 +_,13 @@
         }
     }
 
+    private int behaviorTick = 0; // Pufferfish
     @Override
     protected void customServerAiStep() {
         ServerLevel serverlevel = (ServerLevel)this.level();
         serverlevel.getProfiler().push("wardenBrain");
-        this.getBrain().tick(serverlevel, this);
+        if (this.behaviorTick++ % this.activatedPriority == 0) // Pufferfish // Leaf - Plazma - Add missing purpur configuration options
+            this.getBrain().tick(serverlevel, this);
         this.level().getProfiler().pop();
         super.customServerAiStep();
         if ((this.tickCount + this.getId()) % 120 == 0) {
@@ -567,6 +_,16 @@
             protected PathFinder createPathFinder(int p_219479_) {
                 this.nodeEvaluator = new WalkNodeEvaluator();
                 this.nodeEvaluator.setCanPassDoors(true);
+                // Kaiiju start - petal - async path processing
+                if (GlobalConfiguration.get().asyncPathFinding.enabled) {
+                    return new PathFinder(this.nodeEvaluator, p_219479_, GroundPathNavigation.nodeEvaluatorGenerator) {
+                        @Override
+                        protected float distance(Node first, Node second) {
+                            return first.distanceToXZ(second);
+                        }
+                    };
+                }
+                // Kaiiju end
                 return new PathFinder(this.nodeEvaluator, p_219479_) {
                     @Override
                     protected float distance(Node p_219486_, Node p_219487_) {
