--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -268,10 +_,33 @@
         return this.assignProfessionWhenSpawned;
     }
 
+    // Paper start - EAR 2
+    @Override
+    public void inactiveTick() {
+        // SPIGOT-3874, SPIGOT-3894, SPIGOT-3846, SPIGOT-5286 :(
+        if (this.getUnhappyCounter() > 0) {
+            this.setUnhappyCounter(this.getUnhappyCounter() - 1);
+        }
+        if (this.isEffectiveAi()) {
+            if (this.level().paperConfig().spigotConfigs.tickInactiveVillagers) {
+                this.customServerAiStep();
+            } else {
+                this.customServerAiStep(true);
+            }
+        }
+        this.maybeDecayGossip();
+        super.inactiveTick();
+    }
+    // Paper end - EAR 2
     @Override
     protected void customServerAiStep() {
+        // Paper start - EAR 2
+        this.customServerAiStep(false);
+    }
+
+    protected void customServerAiStep(final boolean inactive) {
         this.level().getProfiler().push("villagerBrain");
-        this.getBrain().tick((ServerLevel)this.level(), this);
+        if (!inactive) this.getBrain().tick((ServerLevel)this.level(), this); // Paper - EAR 2
         this.level().getProfiler().pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
@@ -295,7 +_,7 @@
             this.lastTradedPlayer = null;
         }
 
-        if (!this.isNoAi() && this.random.nextInt(100) == 0) {
+        if (!inactive && !this.isNoAi() && this.random.nextInt(100) == 0) { // Paper - EAR 2
             Raid raid = ((ServerLevel)this.level()).getRaidAt(this.blockPosition());
             if (raid != null && raid.isActive() && !raid.isOver()) {
                 this.level().broadcastEntityEvent(this, (byte)42);
@@ -306,6 +_,7 @@
             this.stopTrading();
         }
 
+        if (inactive) return; // Paper - EAR 2
         super.customServerAiStep();
     }
 
