--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level;
 
 import com.mojang.logging.LogUtils;
+import io.papermc.paper.configuration.GlobalConfiguration;
 import it.unimi.dsi.fastutil.objects.Object2IntMap;
 import it.unimi.dsi.fastutil.objects.Object2IntMaps;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
@@ -43,23 +_,25 @@
 import net.minecraft.world.level.levelgen.structure.structures.NetherFortressStructure;
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.phys.Vec3;
+import net.neoforged.neoforge.event.EventHooks;
 import org.slf4j.Logger;
 
+// TODO: ForgeHooks.canEntitySpawn
 public final class NaturalSpawner {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int MIN_SPAWN_DISTANCE = 24;
     public static final int SPAWN_DISTANCE_CHUNK = 8;
     public static final int SPAWN_DISTANCE_BLOCK = 128;
     static final int MAGIC_NUMBER = (int)Math.pow(17.0, 2.0);
-    private static final MobCategory[] SPAWNING_CATEGORIES = Stream.of(MobCategory.values())
+    public static final MobCategory[] SPAWNING_CATEGORIES = Stream.of(MobCategory.values())
         .filter(p_47037_ -> p_47037_ != MobCategory.MISC)
         .toArray(MobCategory[]::new);
 
     private NaturalSpawner() {
     }
 
-    public static NaturalSpawner.SpawnState createState(
-        int p_186525_, Iterable<Entity> p_186526_, NaturalSpawner.ChunkGetter p_186527_, LocalMobCapCalculator p_186528_
+    public static SpawnState createState(
+        int p_186525_, Iterable<Entity> p_186526_, ChunkGetter p_186527_, LocalMobCapCalculator p_186528_
     ) {
         PotentialCalculator potentialcalculator = new PotentialCalculator();
         Object2IntOpenHashMap<MobCategory> object2intopenhashmap = new Object2IntOpenHashMap<>();
@@ -69,8 +_,17 @@
                 continue;
             }
 
-            MobCategory mobcategory = entity.getType().getCategory();
+            MobCategory mobcategory = entity.getClassification(true);
             if (mobcategory != MobCategory.MISC) {
+                if (entity instanceof Mob mob) {
+                    // Paper start - Only count natural spawns
+                    if (!entity.level().paperConfig().entities.spawning.countAllMobsForSpawning &&
+                            !(mob.getSpawnType() == MobSpawnType.NATURAL ||
+                                    mob.getSpawnType() == MobSpawnType.CHUNK_GENERATION)) {
+                        continue;
+                    }
+                    // Paper end - Only count natural spawns
+                }
                 BlockPos blockpos = entity.blockPosition();
                 p_186527_.query(
                     ChunkPos.asLong(blockpos),
@@ -82,17 +_,22 @@
                             potentialcalculator.addCharge(entity.blockPosition(), mobspawnsettings$mobspawncost.charge());
                         }
 
-                        if (entity instanceof Mob) {
+                        if (p_186528_ != null &&  entity instanceof Mob) {// Paper - Optional per player mob spawns
                             p_186528_.addMob(p_275163_.getPos(), mobcategory);
                         }
 
                         object2intopenhashmap.addTo(mobcategory, 1);
+                        // Paper start - Optional per player mob spawns
+                        if (true) {
+                            ((ServerLevel) p_275163_.level).getChunkSource().chunkMap.updatePlayerMobTypeMap(entity);
+                        }
+                        // Paper end - Optional per player mob spawns
                     }
                 );
             }
         }
 
-        return new NaturalSpawner.SpawnState(p_186525_, object2intopenhashmap, potentialcalculator, p_186528_);
+        return new SpawnState(p_186525_, object2intopenhashmap, potentialcalculator, p_186528_);
     }
 
     static Biome getRoughBiome(BlockPos p_47096_, ChunkAccess p_47097_) {
@@ -100,7 +_,7 @@
     }
 
     public static void spawnForChunk(
-        ServerLevel p_47030_, LevelChunk p_47031_, NaturalSpawner.SpawnState p_47032_, boolean p_47033_, boolean p_47034_, boolean p_47035_
+        ServerLevel p_47030_, LevelChunk p_47031_, SpawnState p_47032_, boolean p_47033_, boolean p_47034_, boolean p_47035_
     ) {
         p_47030_.getProfiler().push("spawner");
 
@@ -117,7 +_,7 @@
     }
 
     public static void spawnCategoryForChunk(
-        MobCategory p_47046_, ServerLevel p_47047_, LevelChunk p_47048_, NaturalSpawner.SpawnPredicate p_47049_, NaturalSpawner.AfterSpawnCallback p_47050_
+        MobCategory p_47046_, ServerLevel p_47047_, LevelChunk p_47048_, SpawnPredicate p_47049_, AfterSpawnCallback p_47050_
     ) {
         BlockPos blockpos = getRandomPosWithin(p_47047_, p_47048_);
         if (blockpos.getY() >= p_47047_.getMinBuildHeight() + 1) {
@@ -138,14 +_,14 @@
         ServerLevel p_47040_,
         ChunkAccess p_47041_,
         BlockPos p_47042_,
-        NaturalSpawner.SpawnPredicate p_47043_,
-        NaturalSpawner.AfterSpawnCallback p_47044_
+        SpawnPredicate p_47043_,
+        AfterSpawnCallback p_47044_
     ) {
         StructureManager structuremanager = p_47040_.structureManager();
         ChunkGenerator chunkgenerator = p_47040_.getChunkSource().getGenerator();
         int i = p_47042_.getY();
-        BlockState blockstate = p_47041_.getBlockState(p_47042_);
-        if (!blockstate.isRedstoneConductor(p_47041_, p_47042_)) {
+        BlockState blockstate = p_47040_.getBlockStateIfLoadedAndInBounds(p_47042_); // Paper - don't load chunks for mob spawn
+        if (blockstate != null && !blockstate.isRedstoneConductor(p_47041_, p_47042_)) { // Paper - don't load chunks for mob spawn
             BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
             int j = 0;
 
@@ -167,7 +_,7 @@
                     Player player = p_47040_.getNearestPlayer(d0, (double)i, d1, -1.0, false);
                     if (player != null) {
                         double d2 = player.distanceToSqr(d0, (double)i, d1);
-                        if (isRightDistanceToPlayerAndSpawnPoint(p_47040_, p_47041_, blockpos$mutableblockpos, d2)) {
+                        if (p_47040_.isLoadedAndInBounds(blockpos$mutableblockpos) && (p_47041_.getPos().longKey == ChunkPos.asLong(blockpos$mutableblockpos) || p_47040_.getChunkIfLoadedImmediately(blockpos$mutableblockpos.getX() >> 4, blockpos$mutableblockpos.getZ() >> 4) != null) && isRightDistanceToPlayerAndSpawnPoint(p_47040_, p_47041_, blockpos$mutableblockpos, d2)) { // Paper - don't load chunks for mob spawn
                             if (mobspawnsettings$spawnerdata == null) {
                                 Optional<MobSpawnSettings.SpawnerData> optional = getRandomSpawnMobAt(
                                     p_47040_, structuremanager, chunkgenerator, p_47039_, p_47040_.random, blockpos$mutableblockpos
@@ -181,10 +_,14 @@
                                     + p_47040_.random.nextInt(1 + mobspawnsettings$spawnerdata.maxCount - mobspawnsettings$spawnerdata.minCount);
                             }
 
-                            if (isValidSpawnPostitionForType(
-                                    p_47040_, p_47039_, structuremanager, chunkgenerator, mobspawnsettings$spawnerdata, blockpos$mutableblockpos, d2
-                                )
-                                && p_47043_.test(mobspawnsettings$spawnerdata.type, blockpos$mutableblockpos, p_47041_)) {
+                            boolean doSpawning = isValidSpawnPostitionForType(p_47040_, p_47039_, structuremanager, chunkgenerator, mobspawnsettings$spawnerdata, blockpos$mutableblockpos, d2);
+                            // Paper start - per player mob count backoff
+                            if (!doSpawning) {
+                                p_47040_.getChunkSource().chunkMap.updateFailurePlayerMobTypeMap(blockpos$mutableblockpos.getX() >> 4, blockpos$mutableblockpos.getZ() >> 4, p_47039_);
+                            }
+                            // Paper end - per player mob count backoff
+
+                            if (doSpawning && p_47043_.test(mobspawnsettings$spawnerdata.type, blockpos$mutableblockpos, p_47041_)) {
                                 Mob mob = getMobForSpawn(p_47040_, mobspawnsettings$spawnerdata.type);
                                 if (mob == null) {
                                     return;
@@ -199,7 +_,12 @@
                                     l1++;
                                     p_47040_.addFreshEntityWithPassengers(mob);
                                     p_47044_.run(mob, p_47041_);
-                                    if (j >= mob.getMaxSpawnClusterSize()) {
+                                    // Goldenforge start - Optional per player mob spawns
+                                    if (p_47040_.paperConfig().entities.spawning.perPlayerMobSpawns) {
+                                        p_47040_.getChunkSource().chunkMap.updatePlayerMobTypeMap(mob);
+                                    }
+                                    // Goldenforge end - Optional per player mob spawns
+                                    if (j >= EventHooks.getMaxSpawnClusterSize(mob)) {
                                         return;
                                     }
 
@@ -272,13 +_,13 @@
         return p_46994_ > (double)(p_46993_.getType().getCategory().getDespawnDistance() * p_46993_.getType().getCategory().getDespawnDistance())
                 && p_46993_.removeWhenFarAway(p_46994_)
             ? false
-            : p_46993_.checkSpawnRules(p_46992_, MobSpawnType.NATURAL) && p_46993_.checkSpawnObstruction(p_46992_);
+            : EventHooks.checkSpawnPosition(p_46993_, p_46992_, MobSpawnType.NATURAL);
     }
 
     private static Optional<MobSpawnSettings.SpawnerData> getRandomSpawnMobAt(
         ServerLevel p_220430_, StructureManager p_220431_, ChunkGenerator p_220432_, MobCategory p_220433_, RandomSource p_220434_, BlockPos p_220435_
     ) {
-        Holder<Biome> holder = p_220430_.getBiome(p_220435_);
+        Holder<Biome> holder = GlobalConfiguration.get().leafConfigs.optimizeBiome.mobSpawn ? p_220430_.getBiomeCached(p_220435_) : p_220430_.getBiome(p_220435_); // Leaf - cache getBiome
         return p_220433_ == MobCategory.WATER_AMBIENT && holder.is(BiomeTags.REDUCED_WATER_AMBIENT_SPAWNS) && p_220434_.nextFloat() < 0.98F
             ? Optional.empty()
             : mobsAt(p_220430_, p_220431_, p_220432_, p_220433_, p_220435_, holder).getRandom(p_220434_);
@@ -303,9 +_,14 @@
         BlockPos p_220448_,
         @Nullable Holder<Biome> p_220449_
     ) {
-        return isInNetherFortressBounds(p_220448_, p_220444_, p_220447_, p_220445_)
-            ? NetherFortressStructure.FORTRESS_ENEMIES
-            : p_220446_.getMobsAt(p_220449_ != null ? p_220449_ : p_220444_.getBiome(p_220448_), p_220445_, p_220447_, p_220448_);
+        // Forge: Add in potential spawns, and replace hardcoded nether fortress mob list
+        if (isInNetherFortressBounds(p_220448_, p_220444_, p_220447_, p_220445_)) {
+            var monsterSpawns = p_220445_.registryAccess().registryOrThrow(Registries.STRUCTURE).getOrThrow(BuiltinStructures.FORTRESS).spawnOverrides().get(MobCategory.MONSTER);
+            if (monsterSpawns != null) { // structure modifiers can clear the spawn overrides
+                return EventHooks.getPotentialSpawns(p_220444_, p_220447_, p_220448_, monsterSpawns.spawns());
+            }
+        }
+        return EventHooks.getPotentialSpawns(p_220444_, p_220447_, p_220448_, p_220446_.getMobsAt(p_220449_ != null ? p_220449_ : (GlobalConfiguration.get().leafConfigs.optimizeBiome.mobSpawn ? p_220444_.getBiomeCached(p_220448_) : p_220444_.getBiome(p_220448_)), p_220445_, p_220447_, p_220448_)); // Leaf - cache getBiome
     }
 
     public static boolean isInNetherFortressBounds(BlockPos p_220456_, ServerLevel p_220457_, MobCategory p_220458_, StructureManager p_220459_) {
@@ -392,8 +_,7 @@
 
                                 entity.moveTo(d0, (double)blockpos.getY(), d1, p_220454_.nextFloat() * 360.0F, 0.0F);
                                 if (entity instanceof Mob mob
-                                    && mob.checkSpawnRules(p_220451_, MobSpawnType.CHUNK_GENERATION)
-                                    && mob.checkSpawnObstruction(p_220451_)) {
+                                    && EventHooks.checkSpawnPosition(mob, p_220451_, MobSpawnType.CHUNK_GENERATION)) {
                                     spawngroupdata = mob.finalizeSpawn(
                                         p_220451_, p_220451_.getCurrentDifficultyAt(mob.blockPosition()), MobSpawnType.CHUNK_GENERATION, spawngroupdata
                                     );
@@ -505,7 +_,7 @@
             this.spawnPotential.addCharge(blockpos, d0);
             MobCategory mobcategory = entitytype.getCategory();
             this.mobCategoryCounts.addTo(mobcategory, 1);
-            this.localMobCapCalculator.addMob(new ChunkPos(blockpos), mobcategory);
+            if (this.localMobCapCalculator != null)  this.localMobCapCalculator.addMob(new ChunkPos(blockpos), mobcategory);
         }
 
         public int getSpawnableChunkCount() {
@@ -518,6 +_,7 @@
 
         boolean canSpawnForCategory(MobCategory p_186549_, ChunkPos p_186550_) {
             int i = p_186549_.getMaxInstancesPerChunk() * this.spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
+            if (this.localMobCapCalculator == null) return this.mobCategoryCounts.getInt(p_186549_) < i; // Paper - Optional per player mob spawns
             return this.mobCategoryCounts.getInt(p_186549_) >= i ? false : this.localMobCapCalculator.canSpawn(p_186549_, p_186550_);
         }
     }
