--- a/net/minecraft/world/entity/player/Abilities.java
+++ b/net/minecraft/world/entity/player/Abilities.java
@@ -5,11 +_,15 @@
 public class Abilities {
     public boolean invulnerable;
     public boolean flying;
+    /**
+     * @deprecated Neoforge: use {@link net.neoforged.neoforge.common.extensions.IPlayerExtension#mayFly} to read and {@link net.neoforged.neoforge.common.NeoForgeMod#CREATIVE_FLIGHT} Attribute to modify
+     */
+    @Deprecated
     public boolean mayfly;
     public boolean instabuild;
     public boolean mayBuild = true;
-    private float flyingSpeed = 0.05F;
-    private float walkingSpeed = 0.1F;
+    public float flyingSpeed = 0.05F;
+    public float walkingSpeed = 0.1F;
 
     public void addSaveData(CompoundTag p_35946_) {
         CompoundTag compoundtag = new CompoundTag();
