--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -138,6 +_,7 @@
     }
 
     public void placeNewPlayer(Connection p_11262_, ServerPlayer p_11263_, CommonListenerCookie p_301988_) {
+        p_11263_.isRealPlayer = true; // Paper
         GameProfile gameprofile = p_11263_.getGameProfile();
         GameProfileCache gameprofilecache = this.server.getProfileCache();
         String s;
@@ -674,9 +_,23 @@
     }
 
     public void saveAll() {
-        for (int i = 0; i < this.players.size(); i++) {
-            this.save(this.players.get(i));
-        }
+        // Paper start - Incremental chunk and player saving
+        this.saveAll(-1);
+    }
+
+    public void saveAll(int interval) {
+        io.papermc.paper.util.MCUtil.ensureMain("Save Players" , () -> { // Paper - Ensure main
+            int numSaved = 0;
+            long now = MinecraftServer.currentTick;
+            for (int i = 0; i < this.players.size(); ++i) {
+                ServerPlayer entityplayer = this.players.get(i);
+                if (interval == -1 || now - entityplayer.lastSave >= interval) {
+                    this.save(entityplayer);
+                    if (interval != -1 && ++numSaved >= io.papermc.paper.configuration.GlobalConfiguration.get().playerAutoSave.maxPerTick()) { break; }
+                }
+                // Paper end - Incremental chunk and player saving
+            }
+            return null; }); // Paper - ensure main
     }
 
     public UserWhiteList getWhiteList() {
@@ -721,7 +_,11 @@
 
     public void sendAllPlayerInfo(ServerPlayer p_11293_) {
         p_11293_.inventoryMenu.sendAllDataToRemote();
-        p_11293_.resetSentInfo();
+        // Paper start - send all attributes
+        // needs to be done because the ServerPlayer instance is being reused on respawn instead of getting replaced like on vanilla
+        java.util.Collection<net.minecraft.world.entity.ai.attributes.AttributeInstance> syncableAttributes = p_11293_.getAttributes().getSyncableAttributes();
+        p_11293_.connection.send(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket(p_11293_.getId(), syncableAttributes));
+        // Paper end - send all attributes
         p_11293_.connection.send(new ClientboundSetCarriedItemPacket(p_11293_.getInventory().selected));
     }
 
