--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -138,6 +_,7 @@
     }
 
     public void placeNewPlayer(Connection p_11262_, ServerPlayer p_11263_, CommonListenerCookie p_301988_) {
+        p_11263_.isRealPlayer = true; // Paper
         GameProfile gameprofile = p_11263_.getGameProfile();
         GameProfileCache gameprofilecache = this.server.getProfileCache();
         String s;
@@ -190,8 +_,8 @@
                 leveldata.isHardcore(),
                 this.server.levelKeys(),
                 this.getMaxPlayers(),
-                this.viewDistance,
-                this.simulationDistance,
+                io.papermc.paper.FeatureHooks.getViewDistance(serverlevel1), // Paper - view distance
+                io.papermc.paper.FeatureHooks.getSimulationDistance(serverlevel1), // Paper - simulation distance
                 flag1,
                 !flag,
                 flag2,
@@ -268,6 +_,17 @@
 
         p_11263_.initInventoryMenu();
 
+        // Paper start - Send empty chunk, so players aren't stuck in the world loading screen with our chunk system not sending chunks when dead
+        if (p_11263_.isDeadOrDying()) {
+            net.minecraft.core.Holder<net.minecraft.world.level.biome.Biome> plains = serverlevel1.registryAccess().lookupOrThrow(net.minecraft.core.registries.Registries.BIOME)
+                    .getOrThrow(net.minecraft.world.level.biome.Biomes.PLAINS);
+            p_11263_.connection.send(new net.minecraft.network.protocol.game.ClientboundLevelChunkWithLightPacket(
+                    new net.minecraft.world.level.chunk.EmptyLevelChunk(serverlevel1, p_11263_.chunkPosition(), plains),
+                    serverlevel1.getLightEngine(), (java.util.BitSet)null, (java.util.BitSet) null)
+            );
+        }
+        // Paper end - Send empty chunk
+
         net.neoforged.neoforge.attachment.AttachmentSync.syncInitialPlayerAttachments(p_11263_);
         net.neoforged.neoforge.event.EventHooks.firePlayerLoggedIn( p_11263_ );
     }
@@ -454,14 +_,15 @@
         dimensiontransition = event.getDimensionTransition();
 
         ServerLevel serverlevel = dimensiontransition.newLevel();
-        ServerPlayer serverplayer = new ServerPlayer(this.server, serverlevel, p_11237_.getGameProfile(), p_11237_.clientInformation());
+        //ServerPlayer serverplayer = new ServerPlayer(this.server, serverlevel, p_11237_.getGameProfile(), p_11237_.clientInformation());*
+        ServerPlayer serverplayer = p_11237_; // Paper - TODO - recreate instance
         serverplayer.connection = p_11237_.connection;
         serverplayer.restoreFrom(p_11237_, p_11238_);
         serverplayer.setId(p_11237_.getId());
         serverplayer.setMainArm(p_11237_.getMainArm());
 
         // Neo: Allow the event to control if the original spawn position is copied
-        if (event.copyOriginalSpawnPosition()) {
+        if (false && event.copyOriginalSpawnPosition()) { // Paper - Once we not reuse the player entity, this can be flipped again but without the events being fired
             serverplayer.copyRespawnPosition(p_11237_);
         }
 
@@ -469,6 +_,13 @@
             serverplayer.addTag(s);
         }
 
+        // Paper start - Once we not reuse the player entity we can remove this.
+        if (!p_11238_) p_11237_.reset();
+        serverplayer.setServerLevel(serverlevel);
+        serverplayer.unsetRemoved();
+        serverplayer.setShiftKeyDown(false);
+        // Paper end
+
         Vec3 vec3 = dimensiontransition.pos();
         serverplayer.moveTo(vec3.x, vec3.y, vec3.z, dimensiontransition.yRot(), dimensiontransition.xRot());
         if (dimensiontransition.missingRespawnBlock()) {
@@ -674,9 +_,23 @@
     }
 
     public void saveAll() {
-        for (int i = 0; i < this.players.size(); i++) {
-            this.save(this.players.get(i));
-        }
+        // Paper start - Incremental chunk and player saving
+        this.saveAll(-1);
+    }
+
+    public void saveAll(int interval) {
+        io.papermc.paper.util.MCUtil.ensureMain("Save Players" , () -> { // Paper - Ensure main
+            int numSaved = 0;
+            long now = MinecraftServer.currentTick;
+            for (int i = 0; i < this.players.size(); ++i) {
+                ServerPlayer entityplayer = this.players.get(i);
+                if (interval == -1 || now - entityplayer.lastSave >= interval) {
+                    this.save(entityplayer);
+                    if (interval != -1 && ++numSaved >= io.papermc.paper.configuration.GlobalConfiguration.get().playerAutoSave.maxPerTick()) { break; }
+                }
+                // Paper end - Incremental chunk and player saving
+            }
+            return null; }); // Paper - ensure main
     }
 
     public UserWhiteList getWhiteList() {
@@ -721,7 +_,11 @@
 
     public void sendAllPlayerInfo(ServerPlayer p_11293_) {
         p_11293_.inventoryMenu.sendAllDataToRemote();
-        p_11293_.resetSentInfo();
+        // Paper start - send all attributes
+        // needs to be done because the ServerPlayer instance is being reused on respawn instead of getting replaced like on vanilla
+        java.util.Collection<net.minecraft.world.entity.ai.attributes.AttributeInstance> syncableAttributes = p_11293_.getAttributes().getSyncableAttributes();
+        p_11293_.connection.send(new net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket(p_11293_.getId(), syncableAttributes));
+        // Paper end - send all attributes
         p_11293_.connection.send(new ClientboundSetCarriedItemPacket(p_11293_.getInventory().selected));
     }
 
