--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -12,6 +_,8 @@
 import java.util.Set;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Vec3i;
@@ -86,7 +_,7 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.AABB;
 
-public abstract class Mob extends LivingEntity implements EquipmentUser, Leashable, Targeting {
+public abstract class Mob extends LivingEntity implements EquipmentUser, Leashable, Targeting, net.caffeinemc.mods.lithium.common.entity.EquipmentEntity { // Leaf - Lithium - equipment tracking
     private static final EntityDataAccessor<Byte> DATA_MOB_FLAGS_ID = SynchedEntityData.defineId(Mob.class, EntityDataSerializers.BYTE);
     private static final int MOB_FLAG_NO_AI = 1;
     private static final int MOB_FLAG_LEFTHANDED = 2;
@@ -131,6 +_,22 @@
     private Leashable.LeashData leashData;
     private BlockPos restrictCenter = BlockPos.ZERO;
     private float restrictRadius = -1.0F;
+    @Nullable
+    private MobSpawnType spawnType;
+    private boolean spawnCancelled = false;
+
+    // Leaf start - Lithium - equipment tracking
+    private void trackEquipChange(ItemStack prevStack, ItemStack currStack) {
+        this.onEquipmentReplaced(prevStack, currStack);
+    }
+    // Leaf end - Lithium - equipment tracking
+
+    // Leaf start - Async target finding
+    public boolean tickingTarget;
+    public final org.dreeam.leaf.async.ai.Waker getGoalCtx() {
+        return tickingTarget ? this.targetSelector.ctx : this.goalSelector.ctx;
+    }
+    // Leaf end - Async target finding
 
     protected Mob(EntityType<? extends Mob> p_21368_, Level p_21369_) {
         super(p_21368_, p_21369_);
@@ -198,6 +_,32 @@
         return this.lookControl;
     }
 
+    int _pufferfish_inactiveTickDisableCounter = 0; // Pufferfish - throttle inactive goal selector ticking
+    // Paper start
+    @Override
+    public void inactiveTick() {
+        super.inactiveTick();
+
+        boolean isThrottled = GlobalConfiguration.get().leafConfigs.throttleInactiveGoalSelectorTick && _pufferfish_inactiveTickDisableCounter++ % 20 != 0; // Pufferfish - throttle inactive goal selector ticking
+        // Leaf start - Async target finding
+        boolean running = this.targetSelector.ctxState != -1 || this.goalSelector.ctxState != -1;
+        this.tickingTarget = false;
+        if (this.goalSelector.inactiveTick(this.activatedPriority, true) && !isThrottled) { // Pufferfish - pass activated priroity // Pufferfish - throttle inactive goal selector ticking
+            this.goalSelector.tick();
+        }
+        this.tickingTarget = true;
+        if (this.targetSelector.inactiveTick(this.activatedPriority, true)) { // Pufferfish - pass activated priority
+            this.targetSelector.tick();
+        }
+        if (GlobalConfiguration.get().asyncTargetFinding.enabled) {
+            if (!running && (this.targetSelector.ctxState != -1 || this.goalSelector.ctxState != -1)) {
+                ((ServerLevel) this.level()).asyncGoalExecutor.submit(this.getId());
+            }
+        }
+        // Leaf end - Async target finding
+    }
+    // Paper end
+
     public MoveControl getMoveControl() {
         return this.getControlledVehicle() instanceof Mob mob ? mob.getMoveControl() : this.moveControl;
     }
@@ -237,7 +_,10 @@
     }
 
     public void setTarget(@Nullable LivingEntity p_21544_) {
-        this.target = p_21544_;
+        net.neoforged.neoforge.event.entity.living.LivingChangeTargetEvent changeTargetEvent = net.neoforged.neoforge.common.CommonHooks.onLivingChangeTarget(this, p_21544_, net.neoforged.neoforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.MOB_TARGET);
+        if(!changeTargetEvent.isCanceled()) {
+             this.target = changeTargetEvent.getNewAboutToBeSetTarget();
+        }
     }
 
     @Override
@@ -346,6 +_,12 @@
         if (!this.level().isClientSide && this.tickCount % 5 == 0) {
             this.updateControlFlags();
         }
+
+        // Neo: Animal armor tick patch
+        if (this.canUseSlot(EquipmentSlot.BODY)) {
+            ItemStack stack = this.getBodyArmorItem();
+            if (isBodyArmorItem(stack)) stack.onAnimalArmorTick(level(), this);
+        }
     }
 
     protected void updateControlFlags() {
@@ -425,10 +_,14 @@
         if (this.isNoAi()) {
             p_21484_.putBoolean("NoAI", this.isNoAi());
         }
+        if (this.spawnType != null) {
+            p_21484_.putString("neoforge:spawn_type", this.spawnType.name());
+        }
     }
 
     @Override
     public void readAdditionalSaveData(CompoundTag p_21450_) {
+        ItemStack prevBodyArmor = this.bodyArmorItem; // Leaf - Lithium - equipment tracking
         super.readAdditionalSaveData(p_21450_);
         if (p_21450_.contains("CanPickUpLoot", 1)) {
             this.setCanPickUpLoot(p_21450_.getBoolean("CanPickUpLoot"));
@@ -440,7 +_,11 @@
 
             for (int i = 0; i < this.armorItems.size(); i++) {
                 CompoundTag compoundtag = listtag.getCompound(i);
-                this.armorItems.set(i, ItemStack.parseOptional(this.registryAccess(), compoundtag));
+                // Leaf start - Lithium - equipment tracking
+                ItemStack currStack = ItemStack.parseOptional(this.registryAccess(), compoundtag);
+                ItemStack prevStack = this.armorItems.set(i, currStack);
+                this.trackEquipChange(prevStack, currStack);
+                // Leaf end - Lithium - equipment tracking
             }
         }
 
@@ -457,7 +_,11 @@
 
             for (int k = 0; k < this.handItems.size(); k++) {
                 CompoundTag compoundtag1 = listtag2.getCompound(k);
-                this.handItems.set(k, ItemStack.parseOptional(this.registryAccess(), compoundtag1));
+                // Leaf start - Lithium - equipment tracking
+                ItemStack currStack = ItemStack.parseOptional(this.registryAccess(), compoundtag1);
+                ItemStack prevStack = this.handItems.set(k, currStack);
+                this.trackEquipChange(prevStack, currStack);
+                // Leaf end - Lithium - equipment tracking
             }
         }
 
@@ -484,6 +_,19 @@
         }
 
         this.setNoAi(p_21450_.getBoolean("NoAI"));
+
+        if (p_21450_.contains("neoforge:spawn_type")) {
+            try {
+                this.spawnType = MobSpawnType.valueOf(p_21450_.getString("neoforge:spawn_type"));
+            } catch (Exception ex) {
+                p_21450_.remove("neoforge:spawn_type");
+            }
+        }
+        // Leaf start - Lithium - equipment tracking
+        if (prevBodyArmor != this.bodyArmorItem) {
+            this.trackEquipChange(prevBodyArmor, this.bodyArmorItem);
+        }
+        // Leaf end - Lithium - equipment tracking
     }
 
     @Override
@@ -539,7 +_,7 @@
             && this.canPickUpLoot()
             && this.isAlive()
             && !this.dead
-            && this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            && net.neoforged.neoforge.event.EventHooks.canEntityGrief(this.level(), this)) {
             Vec3i vec3i = this.getPickupReach();
 
             for (ItemEntity itementity : this.level()
@@ -666,6 +_,10 @@
 
     private double getApproximateAttackDamageWithItem(ItemStack p_330413_) {
         ItemAttributeModifiers itemattributemodifiers = p_330413_.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
+
+        // Neo: Respect gameplay modifiers
+        itemattributemodifiers = p_330413_.getAttributeModifiers();
+
         return itemattributemodifiers.compute(this.getAttributeBaseValue(Attributes.ATTACK_DAMAGE), EquipmentSlot.MAINHAND);
     }
 
@@ -701,23 +_,31 @@
 
     @Override
     public void checkDespawn() {
+        if (net.neoforged.neoforge.event.EventHooks.checkMobDespawn(this)) return;
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) {
             this.discard();
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             Entity entity = this.level().getNearestPlayer(this, -1.0);
             if (entity != null) {
-                double d0 = entity.distanceToSqr(this);
-                int i = this.getType().getCategory().getDespawnDistance();
-                int j = i * i;
-                if (d0 > (double)j && this.removeWhenFarAway(d0)) {
-                    this.discard();
+                // Paper start - Configurable despawn distances
+                final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
+                final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
+                final double dy = Math.abs(entity.getY() - this.getY());
+                final double dySqr = Math.pow(dy, 2);
+                final double dxSqr = Math.pow(entity.getX() - this.getX(), 2);
+                final double dzSqr = Math.pow(entity.getZ() - this.getZ(), 2);
+                final double distanceSquared = dxSqr + dzSqr + dySqr;
+                // Despawn if hard/soft limit is exceeded
+                if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
+                    this.discard(); // CraftBukkit - add Bukkit remove cause
                 }
 
-                int k = this.getType().getCategory().getNoDespawnDistance();
-                int l = k * k;
-                if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && d0 > (double)l && this.removeWhenFarAway(d0)) {
-                    this.discard();
-                } else if (d0 < (double)l) {
+                if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
+                    if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
+                        this.discard(); // CraftBukkit - add Bukkit remove cause
+                    }
+                } else {
+                    // Paper end - Configurable despawn distances
                     this.noActionTime = 0;
                 }
             }
@@ -734,23 +_,29 @@
         this.sensing.tick();
         profilerfiller.pop();
         int i = this.tickCount + this.getId();
+        // Leaf start - Async target finding
+        boolean running = this.targetSelector.ctxState != -1 || this.goalSelector.ctxState != -1;
         if (i % 2 != 0 && this.tickCount > 1) {
-            profilerfiller.push("targetSelector");
-            this.targetSelector.tickRunningGoals(false);
-            profilerfiller.pop();
-            profilerfiller.push("goalSelector");
-            this.goalSelector.tickRunningGoals(false);
-            profilerfiller.pop();
+            this.tickingTarget = true;
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
+                this.targetSelector.tickRunningGoals(false);
+            this.tickingTarget = false;
+            if (this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
+                this.goalSelector.tickRunningGoals(false);
         } else {
-            profilerfiller.push("targetSelector");
-            this.targetSelector.tick();
-            profilerfiller.pop();
-            profilerfiller.push("goalSelector");
-            this.goalSelector.tick();
-            profilerfiller.pop();
-        }
-
-        profilerfiller.push("navigation");
+            this.tickingTarget = true;
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
+                this.targetSelector.tick();
+            this.tickingTarget = false;
+            if (this.goalSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
+                this.goalSelector.tick();
+        }
+        if (GlobalConfiguration.get().asyncTargetFinding.enabled) {
+            if (!running && (this.targetSelector.ctxState != -1 || this.goalSelector.ctxState != -1)) {
+                ((ServerLevel) this.level()).asyncGoalExecutor.submit(this.getId());
+            }
+        }
+        // Leaf end - Async target findinger.push("navigation");
         this.navigation.tick();
         profilerfiller.pop();
         profilerfiller.push("mob tick");
@@ -1133,6 +_,11 @@
         }
     }
 
+    /**
+     * @deprecated Override-Only. External callers should call via {@link net.neoforged.neoforge.event.EventHooks#finalizeMobSpawn}.
+     */
+    @Deprecated
+    @org.jetbrains.annotations.ApiStatus.OverrideOnly
     @Nullable
     public SpawnGroupData finalizeSpawn(ServerLevelAccessor p_21434_, DifficultyInstance p_21435_, MobSpawnType p_21436_, @Nullable SpawnGroupData p_21437_) {
         RandomSource randomsource = p_21434_.getRandom();
@@ -1144,6 +_,7 @@
         }
 
         this.setLeftHanded(randomsource.nextFloat() < 0.05F);
+        this.spawnType = p_21436_;
         return p_21437_;
     }
 
@@ -1470,14 +_,24 @@
     }
 
     @Override
+    @Deprecated // FORGE: use jumpInFluid instead
     protected void jumpInLiquid(TagKey<Fluid> p_204045_) {
+        this.jumpInLiquidInternal(() -> super.jumpInLiquid(p_204045_));
+    }
+
+    private void jumpInLiquidInternal(Runnable onSuper) {
         if (this.getNavigation().canFloat()) {
-            super.jumpInLiquid(p_204045_);
+            onSuper.run();
         } else {
             this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.3, 0.0));
         }
     }
 
+    @Override
+    public void jumpInFluid(net.neoforged.neoforge.fluids.FluidType type) {
+        this.jumpInLiquidInternal(() -> super.jumpInFluid(type));
+    }
+
     @VisibleForTesting
     public void removeFreeWill() {
         this.removeAllGoals(p_351790_ -> true);
@@ -1503,5 +_,40 @@
     public ItemStack getPickResult() {
         SpawnEggItem spawneggitem = SpawnEggItem.byId(this.getType());
         return spawneggitem == null ? null : new ItemStack(spawneggitem);
+    }
+
+    /**
+    * Returns the type of spawn that created this mob, if applicable.
+    * If it could not be determined, this will return null.
+    * <p>
+    * This is set via {@link Mob#finalizeSpawn}, so you should not call this from within that method, instead using the parameter.
+    */
+    @Nullable
+    public final MobSpawnType getSpawnType() {
+        return this.spawnType;
+    }
+
+    /**
+     * This method exists so that spawns can be cancelled from the {@link net.neoforged.neoforge.event.entity.living.MobSpawnEvent.FinalizeSpawn FinalizeSpawnEvent}
+     * without needing to hook up an additional handler for the {@link net.neoforged.neoforge.event.entity.EntityJoinLevelEvent EntityJoinLevelEvent}.
+     * @return if this mob will be blocked from spawning during {@link Level#addFreshEntity(Entity)}
+     * @apiNote Not public-facing API.
+     */
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public final boolean isSpawnCancelled() {
+        return this.spawnCancelled;
+    }
+
+    /**
+     * Marks this mob as being disallowed to spawn during {@link Level#addFreshEntity(Entity)}.<p>
+     * @throws UnsupportedOperationException if this entity has already been {@link Entity#isAddedToLevel()} added to the level.
+     * @apiNote Not public-facing API.
+     */
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public final void setSpawnCancelled(boolean cancel) {
+        if (this.isAddedToLevel()) {
+            throw new UnsupportedOperationException("Late invocations of Mob#setSpawnCancelled are not permitted.");
+        }
+        this.spawnCancelled = cancel;
     }
 }
