--- a/net/minecraft/core/BlockPos.java
+++ b/net/minecraft/core/BlockPos.java
@@ -30,15 +_,16 @@
    }).stable();
    private static final Logger f_121854_ = LogUtils.getLogger();
    public static final BlockPos f_121853_ = new BlockPos(0, 0, 0);
-   private static final int f_121855_ = 1 + Mth.m_14173_(Mth.m_14125_(30000000));
-   private static final int f_121856_ = f_121855_;
-   public static final int f_121857_ = 64 - f_121855_ - f_121856_;
-   private static final long f_121858_ = (1L << f_121855_) - 1L;
-   private static final long f_121859_ = (1L << f_121857_) - 1L;
-   private static final long f_121860_ = (1L << f_121856_) - 1L;
-   private static final int f_175261_ = 0;
-   private static final int f_121861_ = f_121857_;
-   private static final int f_121862_ = f_121857_ + f_121856_;
+   // Paper start - static constants
+   private static final int f_121855_ = 26;
+   private static final int f_121856_ = 26;
+   public static final int f_121857_ = 12;
+   private static final long f_121858_ = 67108863;
+   private static final long f_121859_ = 4095;
+   private static final long f_121860_ = 67108863;
+   private static final int f_121861_ = 12;
+   private static final int f_121862_ = 38;
+   // Paper end
 
    public BlockPos(int p_121869_, int p_121870_, int p_121871_) {
       super(p_121869_, p_121870_, p_121871_);
@@ -65,23 +_,31 @@
    }
 
    public static long m_121910_(long p_121911_, int p_121912_, int p_121913_, int p_121914_) {
-      return m_121882_(m_121983_(p_121911_) + p_121912_, m_122008_(p_121911_) + p_121913_, m_122015_(p_121911_) + p_121914_);
+      return m_121882_((int) (p_121911_ >> 38) + p_121912_, (int) ((p_121911_ << 52) >> 52) + p_121913_, (int) ((p_121911_ << 26) >> 38) + p_121914_); // Paper - simplify/inline
    }
 
    public static int m_121983_(long p_121984_) {
-      return (int)(p_121984_ << 64 - f_121862_ - f_121855_ >> 64 - f_121855_);
+      return (int) (p_121984_ >> 38); // Paper - simplify/inline
    }
 
    public static int m_122008_(long p_122009_) {
-      return (int)(p_122009_ << 64 - f_121857_ >> 64 - f_121857_);
+      return (int) ((p_122009_ << 52) >> 52); // Paper - simplify/inline
    }
 
    public static int m_122015_(long p_122016_) {
-      return (int)(p_122016_ << 64 - f_121861_ - f_121856_ >> 64 - f_121856_);
+      return (int) ((p_122016_ << 26) >> 38);  // Paper - simplify/inline
    }
 
    public static BlockPos m_122022_(long p_122023_) {
-      return new BlockPos(m_121983_(p_122023_), m_122008_(p_122023_), m_122015_(p_122023_));
+      return new BlockPos((int) (p_122023_ >> 38), (int) ((p_122023_ << 52) >> 52), (int) ((p_122023_ << 26) >> 38)); // Paper - simplify/inline
+   }
+
+   public static BlockPos containing(double x, double y, double z) {
+      return new BlockPos(Mth.m_14107_(x), Mth.m_14107_(y), Mth.m_14107_(z));
+   }
+
+   public static BlockPos containing(Position pos) {
+      return containing(pos.m_7096_(), pos.m_7098_(), pos.m_7094_());
    }
 
    public long m_121878_() {
@@ -89,10 +_,7 @@
    }
 
    public static long m_121882_(int p_121883_, int p_121884_, int p_121885_) {
-      long i = 0L;
-      i |= ((long)p_121883_ & f_121858_) << f_121862_;
-      i |= ((long)p_121884_ & f_121859_) << 0;
-      return i | ((long)p_121885_ & f_121860_) << f_121861_;
+      return (((long) p_121883_ & (long) 67108863) << 38) | (((long) p_121884_ & (long) 4095)) | (((long) p_121885_ & (long) 67108863) << 12); // Paper - inline constants and simplify
    }
 
    public static long m_122027_(long p_122028_) {
@@ -123,56 +_,85 @@
       }
    }
 
+   @Override
    public BlockPos m_7494_() {
-      return this.m_121945_(Direction.UP);
+      return new BlockPos(this.m_123341_(), this.m_123342_() + 1, this.m_123343_()); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_6630_(int p_121972_) {
-      return this.m_5484_(Direction.UP, p_121972_);
+      return p_121972_ == 0 ? this : new BlockPos(this.m_123341_(), this.m_123342_() + p_121972_, this.m_123343_()); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_7495_() {
-      return this.m_121945_(Direction.DOWN);
+      return new BlockPos(this.m_123341_(), this.m_123342_() - 1, this.m_123343_()); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_6625_(int p_122000_) {
-      return this.m_5484_(Direction.DOWN, p_122000_);
+      return p_122000_ == 0 ? this : new BlockPos(this.m_123341_(), this.m_123342_() - p_122000_, this.m_123343_()); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_122012_() {
-      return this.m_121945_(Direction.NORTH);
+      return new BlockPos(this.m_123341_(), this.m_123342_(), this.m_123343_() - 1); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_122013_(int p_122014_) {
-      return this.m_5484_(Direction.NORTH, p_122014_);
+      return p_122014_ == 0 ? this : new BlockPos(this.m_123341_(), this.m_123342_(), this.m_123343_() - p_122014_); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_122019_() {
-      return this.m_121945_(Direction.SOUTH);
+      return new BlockPos(this.m_123341_(), this.m_123342_(), this.m_123343_() + 1); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_122020_(int p_122021_) {
-      return this.m_5484_(Direction.SOUTH, p_122021_);
+      return p_122021_ == 0 ? this : new BlockPos(this.m_123341_(), this.m_123342_(), this.m_123343_() + p_122021_); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_122024_() {
-      return this.m_121945_(Direction.WEST);
+      return new BlockPos(this.m_123341_() - 1, this.m_123342_(), this.m_123343_()); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_122025_(int p_122026_) {
-      return this.m_5484_(Direction.WEST, p_122026_);
+      return p_122026_ == 0 ? this : new BlockPos(this.m_123341_() - p_122026_, this.m_123342_(), this.m_123343_()); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_122029_() {
-      return this.m_121945_(Direction.EAST);
+      return new BlockPos(this.m_123341_() + 1, this.m_123342_(), this.m_123343_()); // Paper - Optimize BlockPosition
    }
 
+   @Override
    public BlockPos m_122030_(int p_122031_) {
-      return this.m_5484_(Direction.EAST, p_122031_);
+      return p_122031_ == 0 ? this : new BlockPos(this.m_123341_() + p_122031_, this.m_123342_(), this.m_123343_()); // Paper - Optimize BlockPosition
    }
 
    public BlockPos m_121945_(Direction p_121946_) {
-      return new BlockPos(this.m_123341_() + p_121946_.m_122429_(), this.m_123342_() + p_121946_.m_122430_(), this.m_123343_() + p_121946_.m_122431_());
+      // Paper Start - Optimize BlockPosition
+      switch(p_121946_) {
+         case UP:
+            return new BlockPos(this.m_123341_(), this.m_123342_() + 1, this.m_123343_());
+         case DOWN:
+            return new BlockPos(this.m_123341_(), this.m_123342_() - 1, this.m_123343_());
+         case NORTH:
+            return new BlockPos(this.m_123341_(), this.m_123342_(), this.m_123343_() - 1);
+         case SOUTH:
+            return new BlockPos(this.m_123341_(), this.m_123342_(), this.m_123343_() + 1);
+         case WEST:
+            return new BlockPos(this.m_123341_() - 1, this.m_123342_(), this.m_123343_());
+         case EAST:
+            return new BlockPos(this.m_123341_() + 1, this.m_123342_(), this.m_123343_());
+         default:
+            return new BlockPos(this.m_123341_() + p_121946_.m_122429_(), this.m_123342_() + p_121946_.m_122430_(), this.m_123343_() + p_121946_.m_122431_());
+      }
+      // Paper End
    }
 
    public BlockPos m_5484_(Direction p_121948_, int p_121949_) {
@@ -435,9 +_,9 @@
       }
 
       public BlockPos.MutableBlockPos m_122178_(int p_122179_, int p_122180_, int p_122181_) {
-         this.m_142451_(p_122179_);
-         this.m_142448_(p_122180_);
-         this.m_142443_(p_122181_);
+         this.f_123285_ = p_122179_; // Paper - force inline
+         this.f_123286_ = p_122180_; // Paper - force inline
+         this.f_123289_ = p_122181_; // Paper - force inline
          return this;
       }
 
@@ -499,17 +_,17 @@
       }
 
       public BlockPos.MutableBlockPos m_142451_(int p_175341_) {
-         super.m_142451_(p_175341_);
+         this.f_123285_ = p_175341_; // Paper
          return this;
       }
 
       public BlockPos.MutableBlockPos m_142448_(int p_175343_) {
-         super.m_142448_(p_175343_);
+         this.f_123286_ = p_175343_; // Paper
          return this;
       }
 
       public BlockPos.MutableBlockPos m_142443_(int p_175345_) {
-         super.m_142443_(p_175345_);
+         this.f_123289_ = p_175345_; // Paper
          return this;
       }
 
