--- a/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
@@ -34,7 +_,7 @@
    }
 
    public boolean m_8036_() {
-      if (!this.f_25837_.f_19853_.m_46469_().m_46207_(GameRules.f_46132_)) {
+      if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_25837_.f_19853_, this.f_25837_)) {
          return false;
       } else if (this.f_25600_ > 0) {
          --this.f_25600_;
@@ -113,7 +_,9 @@
 
    @Nullable
    private BlockPos m_25852_(BlockPos p_25853_, BlockGetter p_25854_) {
-      if (p_25854_.m_8055_(p_25853_).m_60713_(this.f_25836_)) {
+      net.minecraft.world.level.block.state.BlockState block = p_25854_.getBlockStateIfLoaded(p_25853_); // Paper
+      if (block == null) return null; // Paper
+      if (block.m_60713_(this.f_25836_)) { // Paper
          return p_25853_;
       } else {
          BlockPos[] ablockpos = new BlockPos[]{p_25853_.m_7495_(), p_25853_.m_122024_(), p_25853_.m_122029_(), p_25853_.m_122012_(), p_25853_.m_122019_(), p_25853_.m_7495_().m_7495_()};
@@ -133,6 +_,7 @@
       if (chunkaccess == null) {
          return false;
       } else {
+         if (!chunkaccess.m_8055_(p_25851_).canEntityDestroy(p_25850_, p_25851_, this.f_25837_)) return false;
          return chunkaccess.m_8055_(p_25851_).m_60713_(this.f_25836_) && chunkaccess.m_8055_(p_25851_.m_7494_()).m_60795_() && chunkaccess.m_8055_(p_25851_.m_6630_(2)).m_60795_();
       }
    }
