--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -75,18 +_,18 @@
     public static final Marker PACKET_RECEIVED_MARKER = Util.make(MarkerFactory.getMarker("PACKET_RECEIVED"), p_202562_ -> p_202562_.add(PACKET_MARKER));
     public static final Marker PACKET_SENT_MARKER = Util.make(MarkerFactory.getMarker("PACKET_SENT"), p_202557_ -> p_202557_.add(PACKET_MARKER));
     public static final Supplier<NioEventLoopGroup> NETWORK_WORKER_GROUP = Suppliers.memoize(
-        () -> new NioEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Client IO #%d").setDaemon(true).build())
+        () -> new NioEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Client IO #%d").setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()) // Paper
     );
     public static final Supplier<EpollEventLoopGroup> NETWORK_EPOLL_WORKER_GROUP = Suppliers.memoize(
-        () -> new EpollEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Epoll Client IO #%d").setDaemon(true).build())
+        () -> new EpollEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Epoll Client IO #%d").setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()) // Paper
     );
     public static final Supplier<DefaultEventLoopGroup> LOCAL_WORKER_GROUP = Suppliers.memoize(
-        () -> new DefaultEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Local Client IO #%d").setDaemon(true).build())
+        () -> new DefaultEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty Local Client IO #%d").setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()) // Paper
     );
     private static final ProtocolInfo<ServerHandshakePacketListener> INITIAL_PROTOCOL = HandshakeProtocols.SERVERBOUND;
     private final PacketFlow receiving;
     private volatile boolean sendLoginDisconnect = true;
-    private final Queue<Consumer<Connection>> pendingActions = Queues.newConcurrentLinkedQueue();
+    private final Queue<WrappedConsumer> pendingActions = Queues.newConcurrentLinkedQueue(); // Paper - Optimize network
     private Channel channel;
     private SocketAddress address;
     @Nullable
@@ -110,6 +_,13 @@
     @Nullable
     private ProtocolInfo<?> inboundProtocol;
 
+    public boolean preparing = true;
+    private static boolean enableExplicitFlush = Boolean.getBoolean("paper.explicit-flush"); // Paper - Disable explicit network manager flushing
+    // Paper end
+    // Paper start - Optimize network
+    public boolean isPending = true;
+    // Paper end - Optimize network
+
     public Connection(PacketFlow p_129482_) {
         this.receiving = p_129482_;
     }
@@ -119,6 +_,7 @@
         super.channelActive(p_129525_);
         this.channel = p_129525_.channel();
         this.address = this.channel.remoteAddress();
+        this.preparing = false; // Spigot
         if (this.delayedDisconnect != null) {
             this.disconnect(this.delayedDisconnect);
         }
@@ -133,6 +_,21 @@
 
     @Override
     public void exceptionCaught(ChannelHandlerContext p_129533_, Throwable p_129534_) {
+        // Paper start - Handle large packets disconnecting client
+        if (p_129534_ instanceof io.netty.handler.codec.EncoderException && p_129534_.getCause() instanceof PacketEncoder.PacketTooLargeException packetTooLargeException) {
+            final Packet<?> packet = packetTooLargeException.getPacket();
+            if (packet.packetTooLarge(this)) {
+                ProtocolSwapHandler.handleOutboundTerminalPacket(p_129533_, packet);
+                return;
+            } else if (packet.isSkippable()) {
+                Connection.LOGGER.debug("Skipping packet due to errors", p_129534_.getCause());
+                ProtocolSwapHandler.handleOutboundTerminalPacket(p_129533_, packet);
+                return;
+            } else {
+                p_129534_ = p_129534_.getCause();
+            }
+        }
+        // Paper end - Handle large packets disconnecting client
         if (p_129534_ instanceof SkipPacketException) {
             LOGGER.debug("Skipping packet due to errors", p_129534_.getCause());
         } else {
@@ -161,7 +_,8 @@
 
                     if (flag) {
                         LOGGER.debug("Failed to sent packet", p_129534_);
-                        if (this.getSending() == PacketFlow.CLIENTBOUND) {
+                        boolean doesDisconnectExist = this.packetListener.protocol() != ConnectionProtocol.STATUS && this.packetListener.protocol() != ConnectionProtocol.HANDSHAKING; // Paper
+                        if (this.getSending() == PacketFlow.CLIENTBOUND && doesDisconnectExist) { // Paper
                             Packet<?> packet = (Packet<?>)(this.sendLoginDisconnect
                                 ? new ClientboundLoginDisconnectPacket(component)
                                 : new ClientboundDisconnectPacket(component));
@@ -190,6 +_,7 @@
                     try {
                         genericsFtw(p_129488_, packetlistener);
                     } catch (RunningOnDifferentThreadException runningondifferentthreadexception) {
+                    } catch (io.papermc.paper.util.ServerStopRejectedExecutionException ignored) { // Paper - do not prematurely disconnect players on stop
                     } catch (RejectedExecutionException rejectedexecutionexception) {
                         this.disconnect(Component.translatable("multiplayer.disconnect.server_shutdown"));
                     } catch (ClassCastException classcastexception) {
@@ -334,11 +_,37 @@
     }
 
     public void send(Packet<?> p_295839_, @Nullable PacketSendListener p_294866_, boolean p_294265_) {
-        if (this.isConnected()) {
-            this.flushQueue();
+        // Paper start - Optimize network: Handle oversized packets better
+        final boolean connected = this.isConnected();
+        if (!connected && !this.preparing) {
+            return;
+        }
+
+        if (connected && (InnerUtil.canSendImmediate(this, p_295839_)
+                || (io.papermc.paper.util.MCUtil.isMainThread() && this.pendingActions.isEmpty()
+                && (p_295839_.getExtraPackets() == null || p_295839_.getExtraPackets().isEmpty())))) {
             this.sendPacket(p_295839_, p_294866_, p_294265_);
         } else {
-            this.pendingActions.add(p_293706_ -> p_293706_.sendPacket(p_295839_, p_294866_, p_294265_));
+            // Write the packets to the queue, then flush - antixray hooks there already
+            final java.util.List<Packet<?>> extraPackets = InnerUtil.buildExtraPackets(p_295839_);
+            final boolean hasExtraPackets = extraPackets != null && !extraPackets.isEmpty();
+            if (!hasExtraPackets) {
+                this.pendingActions.add(new PacketSendAction(p_295839_, p_294866_, p_294265_));
+            } else {
+                final java.util.List<PacketSendAction> actions = new java.util.ArrayList<>(1 + extraPackets.size());
+                actions.add(new PacketSendAction(p_295839_, null, false)); // Delay the future listener until the end of the extra packets
+
+                for (int i = 0, len = extraPackets.size(); i < len;) {
+                    final Packet<?> extraPacket = extraPackets.get(i);
+                    final boolean end = ++i == len;
+                    actions.add(new PacketSendAction(extraPacket, end ? p_294866_ : null, end)); // Append listener to the end
+                }
+
+                this.pendingActions.addAll(actions);
+            }
+
+            this.flushQueue();
+            // Paper end - Optimize network
         }
     }
 
@@ -347,7 +_,7 @@
             this.flushQueue();
             p_294674_.accept(this);
         } else {
-            this.pendingActions.add(p_294674_);
+            this.pendingActions.add(new WrappedConsumer(p_294674_)); // Paper - Optimize network
         }
     }
 
@@ -361,6 +_,11 @@
     }
 
     private void doSendPacket(Packet<?> p_243260_, @Nullable PacketSendListener p_243290_, boolean p_294125_) {
+        // Paper start - Optimize network
+        if (!this.isConnected()) {
+            return;
+        }
+        // Paper end - Optimize network
         ChannelFuture channelfuture = p_294125_ ? this.channel.writeAndFlush(p_243260_) : this.channel.write(p_243260_);
         if (p_243290_ != null) {
             channelfuture.addListener(p_243167_ -> {
@@ -383,7 +_,7 @@
         if (this.isConnected()) {
             this.flush();
         } else {
-            this.pendingActions.add(Connection::flush);
+            this.pendingActions.add(new WrappedConsumer(Connection::flush)); // Paper - Optimize network
         }
     }
 
@@ -395,21 +_,69 @@
         }
     }
 
-    private void flushQueue() {
-        if (this.channel != null && this.channel.isOpen()) {
+    // Paper start - Optimize network: Rewrite this to be safer if ran off main thread
+    private boolean flushQueue() {
+        if (!this.isConnected()) {
+            return true;
+        }
+        if (io.papermc.paper.util.MCUtil.isMainThread()) {
+            return this.processQueue();
+        } else if (this.isPending) {
+            // Should only happen during login/status stages
             synchronized (this.pendingActions) {
-                Consumer<Connection> consumer;
-                while ((consumer = this.pendingActions.poll()) != null) {
-                    consumer.accept(this);
-                }
-            }
-        }
-    }
-
+                return this.processQueue();
+            }
+        }
+        return false;
+    }
+
+    private boolean processQueue() {
+        if (this.pendingActions.isEmpty()) {
+            return true;
+        }
+
+        // If we are on main, we are safe here in that nothing else should be processing queue off main anymore
+        // But if we are not on main due to login/status, the parent is synchronized on packetQueue
+        final java.util.Iterator<WrappedConsumer> iterator = this.pendingActions.iterator();
+        while (iterator.hasNext()) {
+            final WrappedConsumer queued = iterator.next(); // poll -> peek
+
+            // Fix NPE (Spigot bug caused by handleDisconnection())
+            if (queued == null) {
+                return true;
+            }
+
+            if (queued.isConsumed()) {
+                continue;
+            }
+
+            iterator.remove();
+            if (queued.tryMarkConsumed()) {
+                queued.accept(this);
+            }
+        }
+        return true;
+    }
+    // Paper end - Optimize network
+
+    private static final int MAX_PER_TICK = io.papermc.paper.configuration.GlobalConfiguration.get().misc.maxJoinsPerTick; // Paper - Buffer joins to world
+    private static int joinAttemptsThisTick; // Paper - Buffer joins to world
+    private static int currTick; // Paper - Buffer joins to world
     public void tick() {
         this.flushQueue();
+        // Paper start - Buffer joins to world
+        if (Connection.currTick != net.minecraft.server.MinecraftServer.currentTick) {
+            Connection.currTick = net.minecraft.server.MinecraftServer.currentTick;
+            Connection.joinAttemptsThisTick = 0;
+        }
+        // Paper end - Buffer joins to world
         if (this.packetListener instanceof TickablePacketListener tickablepacketlistener) {
-            tickablepacketlistener.tick();
+            // Paper start - Buffer joins to world
+            if (!(this.packetListener instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl loginPacketListener)
+                    || loginPacketListener.state != net.minecraft.server.network.ServerLoginPacketListenerImpl.State.VERIFYING
+                    || Connection.joinAttemptsThisTick++ < MAX_PER_TICK) {
+                tickablepacketlistener.tick();
+            } // Paper end - Buffer joins to world
         }
 
         if (!this.isConnected() && !this.disconnectionHandled) {
@@ -417,7 +_,7 @@
         }
 
         if (this.channel != null) {
-            this.channel.flush();
+            if (enableExplicitFlush) this.channel.eventLoop().execute(() -> this.channel.flush()); // Paper - Disable explicit network manager flushing; we don't need to explicit flush here, but allow opt in incase issues are found to a better version
         }
 
         if (this.tickCount++ % 20 == 0) {
@@ -453,12 +_,14 @@
     }
 
     public void disconnect(DisconnectionDetails p_350867_) {
+        this.preparing = false; // Spigot
+        this.clearPacketQueue(); // Paper - Optimize network
         if (this.channel == null) {
             this.delayedDisconnect = p_350867_;
         }
 
         if (this.isConnected()) {
-            this.channel.close().awaitUninterruptibly();
+            this.channel.close(); // We can't wait as this may be called from an event loop.
             this.disconnectionDetails = p_350867_;
         }
     }
@@ -604,6 +_,14 @@
         }
     }
 
+    // Paper start - add proper async disconnect
+    public void enableAutoRead() {
+        if (this.channel != null) {
+            this.channel.config().setAutoRead(true);
+        }
+    }
+    // Paper end - add proper async disconnect
+
     public void setupCompression(int p_129485_, boolean p_182682_) {
         if (p_129485_ >= 0) {
             if (this.channel.pipeline().get("decompress") instanceof CompressionDecoder compressiondecoder) {
@@ -631,7 +_,7 @@
     public void handleDisconnection() {
         if (this.channel != null && !this.channel.isOpen()) {
             if (this.disconnectionHandled) {
-                LOGGER.warn("handleDisconnection() called twice");
+                //LOGGER.warn("handleDisconnection() called twice"); // Paper - Don't log useless message
             } else {
                 this.disconnectionHandled = true;
                 PacketListener packetlistener = this.getPacketListener();
@@ -642,6 +_,7 @@
                     );
                     packetlistener1.onDisconnect(disconnectiondetails);
                 }
+                this.clearPacketQueue(); // Paper - Optimize network
             }
         }
     }
@@ -669,4 +_,88 @@
     public ProtocolInfo<?> getInboundProtocol() {
         return Objects.requireNonNull(this.inboundProtocol, "Inbound protocol not set?");
     }
+
+    // Paper start - Optimize network
+    public void clearPacketQueue() {
+        this.pendingActions.clear();
+    }
+
+    private static class InnerUtil { // Attempt to hide these methods from ProtocolLib, so it doesn't accidently pick them up.
+
+        @Nullable
+        private static java.util.List<Packet<?>> buildExtraPackets(final Packet<?> packet) {
+            final java.util.List<Packet<?>> extra = packet.getExtraPackets();
+            if (extra == null || extra.isEmpty()) {
+                return null;
+            }
+
+            final java.util.List<Packet<?>> ret = new java.util.ArrayList<>(1 + extra.size());
+            buildExtraPackets0(extra, ret);
+            return ret;
+        }
+
+        private static void buildExtraPackets0(final java.util.List<Packet<?>> extraPackets, final java.util.List<Packet<?>> into) {
+            for (final Packet<?> extra : extraPackets) {
+                into.add(extra);
+                final java.util.List<Packet<?>> extraExtra = extra.getExtraPackets();
+                if (extraExtra != null && !extraExtra.isEmpty()) {
+                    buildExtraPackets0(extraExtra, into);
+                }
+            }
+        }
+
+        private static boolean canSendImmediate(final Connection networkManager, final net.minecraft.network.protocol.Packet<?> packet) {
+            return networkManager.isPending || networkManager.packetListener.protocol() != ConnectionProtocol.PLAY ||
+                    packet instanceof net.minecraft.network.protocol.common.ClientboundKeepAlivePacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundSystemChatPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundDisguisedChatPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundCommandSuggestionsPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundSetTitleTextPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundSetSubtitleTextPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundSetActionBarTextPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundSetTitlesAnimationPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundClearTitlesPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundSoundPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundSoundEntityPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundStopSoundPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundLevelParticlesPacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerInfoUpdatePacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerInfoRemovePacket ||
+                    packet instanceof net.minecraft.network.protocol.game.ClientboundBossEventPacket ||
+                    packet instanceof net.minecraft.network.protocol.ping.ClientboundPongResponsePacket;
+        }
+    }
+
+    private static class WrappedConsumer implements Consumer<Connection> {
+        private final Consumer<Connection> delegate;
+        private final java.util.concurrent.atomic.AtomicBoolean consumed = new java.util.concurrent.atomic.AtomicBoolean(false);
+
+        private WrappedConsumer(final Consumer<Connection> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void accept(final Connection connection) {
+            this.delegate.accept(connection);
+        }
+
+        public boolean tryMarkConsumed() {
+            return consumed.compareAndSet(false, true);
+        }
+
+        public boolean isConsumed() {
+            return consumed.get();
+        }
+    }
+
+    private static final class PacketSendAction extends WrappedConsumer {
+        private final Packet<?> packet;
+
+        private PacketSendAction(final Packet<?> packet, @Nullable final PacketSendListener packetSendListener, final boolean flush) {
+            super(connection -> connection.sendPacket(packet, packetSendListener, flush));
+            this.packet = packet;
+        }
+    }
+    // Paper end - Optimize network
 }
