--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -143,9 +_,11 @@
       return (Brain<Goat>)super.m_6274_();
    }
 
+   private int behaviorTick = 0; // Pufferfish
    protected void m_8024_() {
       this.f_19853_.m_46473_().m_6180_("goatBrain");
-      this.m_6274_().m_21865_((ServerLevel)this.f_19853_, this);
+      if (this.behaviorTick++ % this.activatedPriority == 0) // Pufferfish
+         this.m_6274_().m_21865_((ServerLevel)this.f_19853_, this);
       this.f_19853_.m_46473_().m_7238_();
       this.f_19853_.m_46473_().m_6180_("goatActivityUpdate");
       GoatAi.m_149455_(this);
