--- a/net/minecraft/world/entity/ai/attributes/Attribute.java
+++ b/net/minecraft/world/entity/ai/attributes/Attribute.java
@@ -16,10 +_,15 @@
     private boolean syncable;
     private final String descriptionId;
     private Attribute.Sentiment sentiment = Attribute.Sentiment.POSITIVE;
+    // Leaf start - optimize attribute
+    public final int id;
+    private static int size = 0;
+    // Leaf end - optimize attribute
 
     protected Attribute(String p_22080_, double p_22081_) {
         this.defaultValue = p_22081_;
         this.descriptionId = p_22080_;
+        this.id = size++; // Leaf - optimize attribute
     }
 
     public double getDefaultValue() {
