--- a/net/minecraft/world/phys/shapes/ArrayVoxelShape.java
+++ b/net/minecraft/world/phys/shapes/ArrayVoxelShape.java
@@ -13,14 +_,14 @@
 
     protected ArrayVoxelShape(DiscreteVoxelShape p_82572_, double[] p_82573_, double[] p_82574_, double[] p_82575_) {
         this(
-            p_82572_,
-            DoubleArrayList.wrap(Arrays.copyOf(p_82573_, p_82572_.getXSize() + 1)),
-            DoubleArrayList.wrap(Arrays.copyOf(p_82574_, p_82572_.getYSize() + 1)),
-            DoubleArrayList.wrap(Arrays.copyOf(p_82575_, p_82572_.getZSize() + 1))
+                p_82572_,
+                DoubleArrayList.wrap(Arrays.copyOf(p_82573_, p_82572_.getXSize() + 1)),
+                DoubleArrayList.wrap(Arrays.copyOf(p_82574_, p_82572_.getYSize() + 1)),
+                DoubleArrayList.wrap(Arrays.copyOf(p_82575_, p_82572_.getZSize() + 1))
         );
     }
 
-    ArrayVoxelShape(DiscreteVoxelShape p_82567_, DoubleList p_82568_, DoubleList p_82569_, DoubleList p_82570_) {
+    public ArrayVoxelShape(DiscreteVoxelShape p_82567_, DoubleList p_82568_, DoubleList p_82569_, DoubleList p_82570_) {
         super(p_82567_);
         int i = p_82567_.getXSize() + 1;
         int j = p_82567_.getYSize() + 1;
@@ -31,9 +_,10 @@
             this.zs = p_82570_;
         } else {
             throw (IllegalArgumentException)Util.pauseInIde(
-                new IllegalArgumentException("Lengths of point arrays must be consistent with the size of the VoxelShape.")
+                    new IllegalArgumentException("Lengths of point arrays must be consistent with the size of the VoxelShape.")
             );
         }
+        ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)this).moonrise$initCache(); // Paper - optimise collisions
     }
 
     @Override
