--- a/net/minecraft/world/phys/shapes/OffsetDoubleList.java
+++ b/net/minecraft/world/phys/shapes/OffsetDoubleList.java
@@ -4,8 +_,8 @@
 import it.unimi.dsi.fastutil.doubles.DoubleList;
 
 public class OffsetDoubleList extends AbstractDoubleList {
-    private final DoubleList delegate;
-    private final double offset;
+    public final DoubleList delegate;
+    public final double offset;
 
     public OffsetDoubleList(DoubleList p_83031_, double p_83032_) {
         this.delegate = p_83031_;
