--- a/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
@@ -26,6 +_,7 @@
 
     @Override
     public void place(TreeDecorator.Context p_226028_) {
+        if (p_226028_.logs().isEmpty()) return; // Paper - Fix crash when trying to generate without logs
         RandomSource randomsource = p_226028_.random();
         if (!(randomsource.nextFloat() >= this.probability)) {
             List<BlockPos> list = p_226028_.logs();
