--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -253,7 +_,7 @@
     }
 
     public static boolean isShapeFullBlock(VoxelShape p_49917_) {
-        return SHAPE_FULL_BLOCK_CACHE.getUnchecked(p_49917_);
+        return ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)p_49917_).moonrise$isFullBlock(); // Paper - optimise collisions
     }
 
     public void animateTick(BlockState p_220827_, Level p_220828_, BlockPos p_220829_, RandomSource p_220830_) {
