--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -135,9 +_,29 @@
     @Override
     protected void onPlace(BlockState p_54754_, Level p_54755_, BlockPos p_54756_, BlockState p_54757_, boolean p_54758_) {
         if (!net.neoforged.neoforge.fluids.FluidInteractionRegistry.canInteract(p_54755_, p_54756_)) {
-            p_54755_.scheduleTick(p_54756_, p_54754_.getFluidState().getType(), this.fluid.getTickDelay(p_54755_));
-        }
-    }
+            p_54755_.scheduleTick(p_54756_, p_54754_.getFluidState().getType(), this.getFlowSpeed(p_54755_, p_54756_)); // Paper - Configurable speed for water flowing over lava
+        }
+    }
+
+    // Paper start - Configurable speed for water flowing over lava
+    public int getFlowSpeed(Level level, BlockPos pos) {
+        if (net.minecraft.core.registries.BuiltInRegistries.FLUID.wrapAsHolder(this.fluid).is(FluidTags.WATER)) {
+            if (
+                    isLava(level, pos.north(1)) ||
+                            isLava(level, pos.south(1)) ||
+                            isLava(level, pos.west(1)) ||
+                            isLava(level, pos.east(1))
+            ) {
+                return level.paperConfig().environment.waterOverLavaFlowSpeed;
+            }
+        }
+        return this.fluid.getTickDelay(level);
+    }
+    private static boolean isLava(Level level, BlockPos pos) {
+        final FluidState fluidState = level.getFluidIfLoaded(pos);
+        return fluidState != null && fluidState.is(FluidTags.LAVA);
+    }
+    // Paper end - Configurable speed for water flowing over lava
 
     @Override
     protected BlockState updateShape(BlockState p_54723_, Direction p_54724_, BlockState p_54725_, LevelAccessor p_54726_, BlockPos p_54727_, BlockPos p_54728_) {
