--- a/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
@@ -3,6 +_,9 @@
 import java.util.EnumSet;
 import java.util.List;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.ai.goal.Goal;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
@@ -23,8 +_,49 @@
         this.setFlags(EnumSet.of(Goal.Flag.TARGET));
     }
 
+   // Leaf start - Async Target Finding
+    protected boolean poll() {
+        if (!(this.mob.getGoalCtx().result() instanceof LivingEntity target)) return false;
+        ServerLevel serverLevel = getServerLevel(this.mob);
+        if (serverLevel == null || !target.isAlive() || !attackTargeting.test(golem, target)) return false;
+        if ((target instanceof Player player && (player.isSpectator() || player.isCreative()))) return false;
+        this.potentialTarget = target;
+        return true;
+    }
+
+    private void findTargetAsync() {
+        final IronGolem mob = this.golem;
+        final var ctx = mob.getGoalCtx();
+        if (!ctx.state) return;
+        AABB bound = this.golem.getBoundingBox().inflate(10.0, 8.0, 10.0);
+        final ServerLevel serverLevel = getServerLevel(mob);
+        ctx.wake = () -> {
+            List<Villager> nearbyEntities = serverLevel.getNearbyEntities(Villager.class, attackTargeting, mob, bound);
+            List<Player> nearbyPlayers = serverLevel.getNearbyPlayers(attackTargeting, mob, bound);
+            for (Villager villager : nearbyEntities) {
+                for (Player player : nearbyPlayers) {
+                    int playerReputation = villager.getPlayerReputation(player);
+                    if (playerReputation <= -100) {
+                        return player;
+                    }
+                }
+            }
+            return null;
+        };
+    }
+    // Leaf end - Async Target Finding
+
     @Override
     public boolean canUse() {
+        // Leaf start - Async target finding
+        if (poll()) {
+            return true;
+        }
+        if (GlobalConfiguration.get().asyncTargetFinding.searchEntity) {
+            this.findTargetAsync();
+            return false;
+        }
+        // Leaf end - Async target finding
         AABB aabb = this.golem.getBoundingBox().inflate(10.0, 8.0, 10.0);
         List<? extends LivingEntity> list = this.golem.level().getNearbyEntities(Villager.class, this.attackTargeting, this.golem, aabb);
         List<Player> list1 = this.golem.level().getNearbyPlayers(this.attackTargeting, this.golem, aabb);
