--- a/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/HurtByTargetGoal.java
@@ -4,6 +_,8 @@
 import java.util.Iterator;
 import java.util.List;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
@@ -73,6 +_,46 @@
     protected void alertOthers() {
         double d0 = this.getFollowDistance();
         AABB aabb = AABB.unitCubeFromLowerCorner(this.mob.position()).inflate(d0, 10.0, d0);
+
+        // Leaf start - Async alert other
+        if (GlobalConfiguration.get().asyncTargetFinding.alertOther) {
+            final var self = this.mob;
+            final var ctx = self.getGoalCtx();
+            if (!ctx.state) return;
+            final var serverLevel = getServerLevel(self);
+            final var toIgnoreAlert = this.toIgnoreAlert;
+            ctx.wake = () -> {
+                var toAlert = new java.util.ArrayList<Mob>();
+                List<? extends Mob> entitiesOfClass = serverLevel
+                        .getEntitiesOfClass(self.getClass(), aabb, EntitySelector.NO_SPECTATORS);
+                for (Mob mob : entitiesOfClass) {
+                    if (self == mob
+                            || mob.getTarget() != null
+                            || (self instanceof TamableAnimal && ((TamableAnimal) self).getOwner() != ((TamableAnimal) mob).getOwner())
+                            || mob.isAlliedTo(self.getLastHurtByMob())) {
+                        continue;
+                    }
+                    if (toIgnoreAlert == null) {
+                        toAlert.add(mob);
+                        continue;
+                    }
+                    boolean flag = false;
+                    for (Class<?> clazz : toIgnoreAlert) {
+                        if (mob.getClass() == clazz) {
+                            flag = true;
+                            break;
+                        }
+                    }
+                    if (!flag) {
+                        toAlert.add(mob);
+                    }
+                }
+                return toAlert;
+            };
+            return;
+        }
+        // Leaf end - Async alert other
+
         List<? extends Mob> list = this.mob.level().getEntitiesOfClass((Class<? extends Mob>)this.mob.getClass(), aabb, EntitySelector.NO_SPECTATORS);
         Iterator iterator = list.iterator();
 
@@ -110,6 +_,36 @@
             this.alertOther(mob, this.mob.getLastHurtByMob());
         }
     }
+
+    // Leaf start - Async alert other
+    public void poll() {
+        if (!(this.mob.getGoalCtx().result() instanceof List toAlert)) return;
+        LivingEntity lastHurtByMob = this.mob.getLastHurtByMob();
+        if (lastHurtByMob.getType() == EntityType.PLAYER && getServerLevel(this.mob).getGameRules().getBoolean(GameRules.RULE_UNIVERSAL_ANGER)) {
+            return;
+        }
+        for (Class<?> clazz : this.toIgnoreDamage) {
+            if (clazz.isAssignableFrom(lastHurtByMob.getClass())) {
+                return;
+            }
+        }
+        if (this.mob.getTarget() == null) {
+            return;
+        }
+        if (!canContinueToUse()) {
+            return;
+        }
+        if (!this.canAttack(lastHurtByMob, HURT_BY_TARGETING)) {
+            return;
+        }
+        for (var obj : toAlert) {
+            Mob mob = (Mob) obj;
+            if (mob.getTarget() == null) {
+                alertOther(mob, lastHurtByMob);
+            }
+        }
+    }
+    // Leaf end - Async alert other
 
     protected void alertOther(Mob p_26042_, LivingEntity p_26043_) {
         p_26042_.setTarget(p_26043_);
