--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -13,17 +_,41 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
 
-public class LevelChunkSection {
+public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_counting.BlockCountingChunkSection { // Paper - block counting
     public static final int SECTION_WIDTH = 16;
     public static final int SECTION_HEIGHT = 16;
     public static final int SECTION_SIZE = 4096;
     public static final int BIOME_CONTAINER_BITS = 2;
-    private short nonEmptyBlockCount;
+    public short nonEmptyBlockCount;
     private short tickingBlockCount;
     private short tickingFluidCount;
-    private final PalettedContainer<BlockState> states;
+    public final PalettedContainer<BlockState> states;
     private PalettedContainerRO<Holder<Biome>> biomes;
 
+    // Paper start - block counting
+    private static final it.unimi.dsi.fastutil.shorts.ShortArrayList FULL_LIST = new it.unimi.dsi.fastutil.shorts.ShortArrayList(16*16*16);
+    static {
+        for (short i = 0; i < (16*16*16); ++i) {
+            FULL_LIST.add(i);
+        }
+    }
+
+    private boolean isClient;
+    private static final short CLIENT_FORCED_SPECIAL_COLLIDING_BLOCKS = (short)9999;
+    private short specialCollidingBlocks;
+    private final ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks = new ca.spottedleaf.moonrise.common.list.ShortList();
+
+    @Override
+    public final boolean moonrise$hasSpecialCollidingBlocks() {
+        return this.specialCollidingBlocks != 0;
+    }
+
+    @Override
+    public final ca.spottedleaf.moonrise.common.list.ShortList moonrise$getTickingBlockList() {
+        return this.tickingBlocks;
+    }
+    // Paper end - block counting
+
     public LevelChunkSection(PalettedContainer<BlockState> p_282846_, PalettedContainerRO<Holder<Biome>> p_281695_) {
         this.states = p_282846_;
         this.biomes = p_281695_;
@@ -55,6 +_,45 @@
         return this.setBlockState(p_62987_, p_62988_, p_62989_, p_62990_, true);
     }
 
+    // Paper start - block counting
+    private void updateBlockCallback(final int x, final int y, final int z, final BlockState newState,
+                                     final BlockState oldState) {
+        if (oldState == newState) {
+            return;
+        }
+
+        if (this.isClient) {
+            if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isSpecialCollidingBlock(newState)) {
+                this.specialCollidingBlocks = CLIENT_FORCED_SPECIAL_COLLIDING_BLOCKS;
+            }
+            return;
+        }
+
+        final boolean isSpecialOld = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isSpecialCollidingBlock(oldState);
+        final boolean isSpecialNew = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isSpecialCollidingBlock(newState);
+        if (isSpecialOld != isSpecialNew) {
+            if (isSpecialOld) {
+                --this.specialCollidingBlocks;
+            } else {
+                ++this.specialCollidingBlocks;
+            }
+        }
+
+        final boolean oldTicking = oldState.isRandomlyTicking();
+        final boolean newTicking = newState.isRandomlyTicking();
+        if (oldTicking != newTicking) {
+            final ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks = this.tickingBlocks;
+            final short position = (short)(x | (z << 4) | (y << (4+4)));
+
+            if (oldTicking) {
+                tickingBlocks.remove(position);
+            } else {
+                tickingBlocks.add(position);
+            }
+        }
+    }
+    // Paper end - block counting
+
     public BlockState setBlockState(int p_62992_, int p_62993_, int p_62994_, BlockState p_62995_, boolean p_62996_) {
         BlockState blockstate;
         if (p_62996_) {
@@ -65,7 +_,7 @@
 
         FluidState fluidstate = blockstate.getFluidState();
         FluidState fluidstate1 = p_62995_.getFluidState();
-        if (!blockstate.isAir()) {
+        if (!blockstate.isEmpty()) { // Neo: Fix MC-232360 for modded blocks (Makes modded isAir blocks not be replaced with Blocks.AIR in all-air chunk sections)
             this.nonEmptyBlockCount--;
             if (blockstate.isRandomlyTicking()) {
                 this.tickingBlockCount--;
@@ -76,7 +_,7 @@
             this.tickingFluidCount--;
         }
 
-        if (!p_62995_.isAir()) {
+        if (!p_62995_.isEmpty()) { // Neo: Fix MC-232360 for modded blocks (Makes modded isAir blocks not be replaced with Blocks.AIR in all-air chunk sections)
             this.nonEmptyBlockCount++;
             if (p_62995_.isRandomlyTicking()) {
                 this.tickingBlockCount++;
@@ -87,6 +_,8 @@
             this.tickingFluidCount++;
         }
 
+        this.updateBlockCallback(p_62992_, p_62993_, p_62994_, p_62995_, blockstate); // Paper - block counting
+
         return blockstate;
     }
 
@@ -107,35 +_,103 @@
     }
 
     public void recalcBlockCounts() {
-        class BlockCounter implements PalettedContainer.CountConsumer<BlockState> {
-            public int nonEmptyBlockCount;
-            public int tickingBlockCount;
-            public int tickingFluidCount;
-
-            public void accept(BlockState p_204444_, int p_204445_) {
-                FluidState fluidstate = p_204444_.getFluidState();
-                if (!p_204444_.isAir()) {
-                    this.nonEmptyBlockCount += p_204445_;
-                    if (p_204444_.isRandomlyTicking()) {
-                        this.tickingBlockCount += p_204445_;
+        // Paper start - block counting
+        // reset, then recalculate
+        this.nonEmptyBlockCount = (short)0;
+        this.tickingBlockCount = (short)0;
+        this.tickingFluidCount = (short)0;
+        this.specialCollidingBlocks = (short)0;
+        this.tickingBlocks.clear();
+
+        if (this.maybeHas((final BlockState state) -> !state.isAir())) {
+            final PalettedContainer.Data<BlockState> data = this.states.data;
+            final Palette<BlockState> palette = data.palette();
+            final int paletteSize = palette.getSize();
+            final net.minecraft.util.BitStorage storage = data.storage();
+
+            final it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<it.unimi.dsi.fastutil.shorts.ShortArrayList> counts;
+            if (paletteSize == 1) {
+                counts = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>(1);
+                counts.put(0, FULL_LIST);
+            } else {
+                counts = ((ca.spottedleaf.moonrise.patches.block_counting.BlockCountingBitStorage)storage).moonrise$countEntries();
+            }
+
+            for (final java.util.Iterator<it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry<it.unimi.dsi.fastutil.shorts.ShortArrayList>> iterator = counts.int2ObjectEntrySet().fastIterator(); iterator.hasNext();) {
+                final it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry<it.unimi.dsi.fastutil.shorts.ShortArrayList> entry = iterator.next();
+                final int paletteIdx = entry.getIntKey();
+                final it.unimi.dsi.fastutil.shorts.ShortArrayList coordinates = entry.getValue();
+                final int paletteCount = coordinates.size();
+
+                final BlockState state = palette.valueFor(paletteIdx);
+
+                if (state.isAir()) {
+                    continue;
+                }
+
+                if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isSpecialCollidingBlock(state)) {
+                    this.specialCollidingBlocks += (short)paletteCount;
+                }
+                this.nonEmptyBlockCount += (short)paletteCount;
+                if (state.isRandomlyTicking()) {
+                    this.tickingBlockCount += (short)paletteCount;
+                    final short[] raw = coordinates.elements();
+                    final int rawLen = raw.length;
+
+                    final ca.spottedleaf.moonrise.common.list.ShortList tickingBlocks = this.tickingBlocks;
+
+                    tickingBlocks.setMinCapacity(Math.min((rawLen + tickingBlocks.size()) * 3 / 2, 16*16*16));
+
+                    java.util.Objects.checkFromToIndex(0, paletteCount, raw.length);
+                    for (int i = 0; i < paletteCount; ++i) {
+                        tickingBlocks.add(raw[i]);
                     }
                 }
 
-                if (!fluidstate.isEmpty()) {
-                    this.nonEmptyBlockCount += p_204445_;
-                    if (fluidstate.isRandomlyTicking()) {
-                        this.tickingFluidCount += p_204445_;
+                final FluidState fluid = state.getFluidState();
+
+                if (!fluid.isEmpty()) {
+                    //this.nonEmptyBlockCount += count; // fix vanilla bug: make non-empty block count correct
+                    if (fluid.isRandomlyTicking()) {
+                        this.tickingFluidCount += (short)paletteCount;
                     }
                 }
             }
         }
-
-        BlockCounter levelchunksection$1blockcounter = new BlockCounter();
-        this.states.count(levelchunksection$1blockcounter);
-        this.nonEmptyBlockCount = (short)levelchunksection$1blockcounter.nonEmptyBlockCount;
-        this.tickingBlockCount = (short)levelchunksection$1blockcounter.tickingBlockCount;
-        this.tickingFluidCount = (short)levelchunksection$1blockcounter.tickingFluidCount;
+        // Paper end - block counting
     }
+
+
+//    public void recalcBlockCounts() {
+//        class BlockCounter implements PalettedContainer.CountConsumer<BlockState> {
+//            public int nonEmptyBlockCount;
+//            public int tickingBlockCount;
+//            public int tickingFluidCount;
+//
+//            public void accept(BlockState p_204444_, int p_204445_) {
+//                FluidState fluidstate = p_204444_.getFluidState();
+//                if (!p_204444_.isEmpty()) { // Neo: Fix MC-232360 for modded blocks (Makes modded isAir blocks not be replaced with Blocks.AIR in all-air chunk sections)
+//                    this.nonEmptyBlockCount += p_204445_;
+//                    if (p_204444_.isRandomlyTicking()) {
+//                        this.tickingBlockCount += p_204445_;
+//                    }
+//                }
+//
+//                if (!fluidstate.isEmpty()) {
+//                    this.nonEmptyBlockCount += p_204445_;
+//                    if (fluidstate.isRandomlyTicking()) {
+//                        this.tickingFluidCount += p_204445_;
+//                    }
+//                }
+//            }
+//        }
+//
+//        BlockCounter levelchunksection$1blockcounter = new BlockCounter();
+//        this.states.count(levelchunksection$1blockcounter);
+//        this.nonEmptyBlockCount = (short)levelchunksection$1blockcounter.nonEmptyBlockCount;
+//        this.tickingBlockCount = (short)levelchunksection$1blockcounter.tickingBlockCount;
+//        this.tickingFluidCount = (short)levelchunksection$1blockcounter.tickingFluidCount;
+//    }
 
     public PalettedContainer<BlockState> getStates() {
         return this.states;
@@ -151,6 +_,7 @@
         PalettedContainer<Holder<Biome>> palettedcontainer = this.biomes.recreate();
         palettedcontainer.read(p_63005_);
         this.biomes = palettedcontainer;
+        this.recalcBlockCounts(); // Paper - block counting
     }
 
     public void readBiomes(FriendlyByteBuf p_275669_) {
