--- a/net/minecraft/world/level/chunk/PalettedContainer.java
+++ b/net/minecraft/world/level/chunk/PalettedContainer.java
@@ -27,6 +_,7 @@
 import net.minecraft.util.ZeroBitStorage;
 
 public class PalettedContainer<T> implements PaletteResize<T>, PalettedContainerRO<T> {
+    // TODO: forceBits-parametered setBits function. -C
     private static final int MIN_PALETTE_BITS = 0;
     private final PaletteResize<T> dummyPaletteResize = (p_238275_, p_238276_) -> 0;
     private final IdMap<T> registry;
@@ -152,7 +_,7 @@
         return this.get(this.strategy.getIndex(p_63088_, p_63089_, p_63090_));
     }
 
-    protected T get(int p_63086_) {
+    public T get(int p_63086_) {
         PalettedContainer.Data<T> data = this.data;
         return data.palette.valueFor(data.storage.get(p_63086_));
     }
