--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -83,6 +_,18 @@
         return new ChunkAccess.TicksToSave(this.blockTicks, this.fluidTicks);
     }
 
+    // Paper start - If loaded util
+    @Override
+    public final FluidState getFluidIfLoaded(BlockPos blockposition) {
+        return this.getFluidState(blockposition);
+    }
+
+    @Override
+    public final BlockState getBlockStateIfLoaded(BlockPos blockposition) {
+        return this.getBlockState(blockposition);
+    }
+    // Paper end
+
     @Override
     public BlockState getBlockState(BlockPos p_63264_) {
         int i = p_63264_.getY();
@@ -133,7 +_,7 @@
                     }
 
                     if (LightEngine.hasDifferentLightProperties(this, p_63217_, blockstate, p_63218_)) {
-                        this.skyLightSources.update(this, i1, j, k1);
+                        // Paper - rewrite chunk system
                         this.lightEngine.checkBlock(p_63217_);
                     }
                 }
