--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -133,7 +_,7 @@
                     }
 
                     if (LightEngine.hasDifferentLightProperties(this, p_63217_, blockstate, p_63218_)) {
-                        this.skyLightSources.update(this, i1, j, k1);
+                        // Paper - rewrite chunk system
                         this.lightEngine.checkBlock(p_63217_);
                     }
                 }
