--- a/net/minecraft/Util.java
+++ b/net/minecraft/Util.java
@@ -135,7 +_,7 @@
     }
 
     public static long getNanos() {
-        return timeSource.getAsLong();
+        return System.nanoTime(); // Paper
     }
 
     public static long getEpochMillis() {
@@ -239,7 +_,7 @@
         throw p_137560_ instanceof RuntimeException ? (RuntimeException)p_137560_ : new RuntimeException(p_137560_);
     }
 
-    private static void onThreadException(Thread p_137496_, Throwable p_137497_) {
+    public static void onThreadException(Thread p_137496_, Throwable p_137497_) {
         pauseInIde(p_137497_);
         if (p_137497_ instanceof CompletionException) {
             p_137497_ = p_137497_.getCause();
