--- a/net/minecraft/core/component/PatchedDataComponentMap.java
+++ b/net/minecraft/core/component/PatchedDataComponentMap.java
@@ -14,10 +_,11 @@
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
 
-public final class PatchedDataComponentMap implements DataComponentMap {
+public final class PatchedDataComponentMap implements DataComponentMap, net.caffeinemc.mods.lithium.common.util.change_tracking.ChangePublisher<PatchedDataComponentMap> { // Leaf - Lithium - equipment tracking
     private final DataComponentMap prototype;
     private Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch;
     private boolean copyOnWrite;
+    private net.caffeinemc.mods.lithium.common.util.change_tracking.ChangeSubscriber<PatchedDataComponentMap> subscriber; // Leaf - Lithium - equipment tracking
 
     public PatchedDataComponentMap(DataComponentMap p_332070_) {
         this(p_332070_, Reference2ObjectMaps.emptyMap(), true);
@@ -64,6 +_,7 @@
 
     @Nullable
     public <T> T set(DataComponentType<? super T> p_330791_, @Nullable T p_330369_) {
+        net.neoforged.neoforge.common.CommonHooks.validateComponent(p_330369_);
         this.ensureMapOwnership();
         T t = this.prototype.get((DataComponentType<? extends T>)p_330791_);
         Optional<T> optional;
@@ -126,6 +_,11 @@
     }
 
     private void ensureMapOwnership() {
+        // Leaf start - Lithium - equipment tracking
+        if (this.subscriber != null) {
+            this.subscriber.notify(this, 0);
+        }
+        // Leaf end - Lithium - equipment tracking
         if (this.copyOnWrite) {
             this.patch = new Reference2ObjectArrayMap<>(this.patch);
             this.copyOnWrite = false;
@@ -194,6 +_,10 @@
         return i;
     }
 
+    public boolean isPatchEmpty() {
+        return this.patch.isEmpty();
+    }
+
     public DataComponentPatch asPatch() {
         if (this.patch.isEmpty()) {
             return DataComponentPatch.EMPTY;
@@ -208,6 +_,22 @@
         return new PatchedDataComponentMap(this.prototype, this.patch, true);
     }
 
+    // Leaf start - Lithium - equipment tracking
+    @Override
+    public void subscribe(net.caffeinemc.mods.lithium.common.util.change_tracking.ChangeSubscriber<PatchedDataComponentMap> subscriber, int subscriberData) {
+        if (subscriberData != 0) {
+            throw new UnsupportedOperationException("ComponentMapImpl does not support subscriber data");
+        }
+        this.subscriber = net.caffeinemc.mods.lithium.common.util.change_tracking.ChangeSubscriber.combine(this.subscriber, 0, subscriber, 0);
+    }
+
+    @Override
+    public int unsubscribe(net.caffeinemc.mods.lithium.common.util.change_tracking.ChangeSubscriber<PatchedDataComponentMap> subscriber) {
+        this.subscriber = net.caffeinemc.mods.lithium.common.util.change_tracking.ChangeSubscriber.without(this.subscriber, subscriber);
+        return 0;
+    }
+    // Leaf end - Lithium - equipment tracking
+
     @Override
     public boolean equals(Object p_332106_) {
         if (this == p_332106_) {
@@ -225,7 +_,23 @@
 
     @Override
     public int hashCode() {
-        return this.prototype.hashCode() + this.patch.hashCode() * 31;
+        return this.prototype.hashCode() + hashPatch(this.patch) * 31;
+    }
+
+    // Neo: Change implementation of hashCode to reduce collisions.
+    // For a map, hashCode is specified as the sum of the hash codes of its entries.
+    // We do that, but change the entry hash code to 31^<key hash> * <value hash>,
+    // where <key hash> is the lower bits of the identity hash code of the key.
+    private static int hashPatch(Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch) {
+        int h = 0, n = patch.size();
+        var iterator = it.unimi.dsi.fastutil.objects.Reference2ObjectMaps.fastIterator(patch);
+        while (n-- != 0) {
+            var entry = iterator.next();
+            int exponent = System.identityHashCode(entry.getKey()) & 0xff;
+            int entryHash = com.google.common.math.IntMath.pow(31, exponent) * entry.getValue().hashCode();
+            h += entryHash;
+        }
+        return h;
     }
 
     @Override
