--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -91,6 +_,7 @@
 import net.minecraft.world.Container;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.MenuProvider;
+import net.minecraft.world.damagesource.CombatTracker;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.effect.MobEffects;
@@ -109,6 +_,7 @@
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.AbstractArrow;
+import net.minecraft.world.food.FoodData;
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.ContainerListener;
 import net.minecraft.world.inventory.ContainerSynchronizer;
@@ -145,10 +_,12 @@
 import net.minecraft.world.scores.Score;
 import net.minecraft.world.scores.Team;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
+import org.goldenforge.GoldenForge;
 import org.slf4j.Logger;
 
-public class ServerPlayer extends Player {
+public class ServerPlayer extends Player implements net.minecraftforge.common.extensions.IForgeServerPlayer {
    private static final Logger LOGGER = LogUtils.getLogger();
+   public long lastSave = MinecraftServer.currentTick; // Paper
    private static final int NEUTRAL_MOB_DEATH_NOTIFICATION_RADII_XZ = 32;
    private static final int NEUTRAL_MOB_DEATH_NOTIFICATION_RADII_Y = 10;
    public ServerGamePacketListenerImpl connection;
@@ -243,6 +_,53 @@
    public int containerCounter;
    public boolean wonGame;
 
+   public boolean isRealPlayer; // Paper
+   public double lastEntitySpawnRadiusSquared = -1.0;
+   public final com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> cachedSingleHashSet; // Paper
+
+   // Paper start - replace player chunk loader
+   private final java.util.concurrent.atomic.AtomicReference<io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.ViewDistances> viewDistances = new java.util.concurrent.atomic.AtomicReference<>(new io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.ViewDistances(-1, -1, -1));
+   public io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
+
+   public io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.ViewDistances getViewDistances() {
+      return this.viewDistances.get();
+   }
+
+   private void updateViewDistance(final java.util.function.Function<io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.ViewDistances, io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.ViewDistances> update) {
+      for (io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.ViewDistances curr = this.viewDistances.get();;) {
+         if (this.viewDistances.compareAndSet(curr, update.apply(curr))) {
+            return;
+         }
+      }
+   }
+
+   public void setTickViewDistance(final int distance) {
+      if ((distance < io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MIN_VIEW_DISTANCE || distance > io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MAX_VIEW_DISTANCE)) {
+         throw new IllegalArgumentException("Tick view distance must be a number between " + io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MIN_VIEW_DISTANCE + " and " + (io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MAX_VIEW_DISTANCE) + ", got: " + distance);
+      }
+      this.updateViewDistance((input) -> {
+         return input.setTickViewDistance(distance);
+      });
+   }
+
+   public void setLoadViewDistance(final int distance) {
+      if (distance != -1 && (distance < io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MIN_VIEW_DISTANCE || distance > io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MAX_VIEW_DISTANCE + 1)) {
+         throw new IllegalArgumentException("Load view distance must be a number between " + io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MIN_VIEW_DISTANCE + " and " + (io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MAX_VIEW_DISTANCE + 1) + " or -1, got: " + distance);
+      }
+      this.updateViewDistance((input) -> {
+         return input.setLoadViewDistance(distance);
+      });
+   }
+
+   public void setSendViewDistance(final int distance) {
+      if (distance != -1 && (distance < io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MIN_VIEW_DISTANCE || distance > io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MAX_VIEW_DISTANCE + 1)) {
+         throw new IllegalArgumentException("Send view distance must be a number between " + io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MIN_VIEW_DISTANCE + " and " + (io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.MAX_VIEW_DISTANCE + 1) + " or -1, got: " + distance);
+      }
+      this.updateViewDistance((input) -> {
+         return input.setSendViewDistance(distance);
+      });
+   }
+   // Paper end - replace player chunk loader
    public ServerPlayer(MinecraftServer p_254143_, ServerLevel p_254435_, GameProfile p_253651_, ClientInformation p_299301_) {
       super(p_254435_, p_254435_.getSharedSpawnPos(), p_254435_.getSharedSpawnAngle(), p_253651_);
       this.textFilter = p_254143_.createTextFilterForPlayer(this);
@@ -253,7 +_,47 @@
       this.setMaxUpStep(1.0F);
       this.fudgeSpawnLocation(p_254435_);
       this.updateOptions(p_299301_);
-   }
+      this.cachedSingleHashSet = new com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<>(this); // Paper
+   }
+
+   // Yes, this doesn't match Vanilla, but it's the best we can do for now.
+   // If this is an issue, PRs are welcome
+   public final BlockPos getSpawnPoint(ServerLevel worldserver) {
+      BlockPos blockposition = worldserver.getSharedSpawnPos();
+
+      if (worldserver.dimensionType().hasSkyLight() && worldserver.serverLevelData.getGameType() != GameType.ADVENTURE) {
+         int i = Math.max(0, this.server.getSpawnRadius(worldserver));
+         int j = Mth.floor(worldserver.getWorldBorder().getDistanceToBorder((double) blockposition.getX(), (double) blockposition.getZ()));
+
+         if (j < i) {
+            i = j;
+         }
+
+         if (j <= 1) {
+            i = 1;
+         }
+
+         long k = (long) (i * 2 + 1);
+         long l = k * k;
+         int i1 = l > 2147483647L ? Integer.MAX_VALUE : (int) l;
+         int j1 = this.getCoprime(i1);
+         int k1 = RandomSource.create().nextInt(i1);
+
+         for (int l1 = 0; l1 < i1; ++l1) {
+            int i2 = (k1 + j1 * l1) % i1;
+            int j2 = i2 % (i * 2 + 1);
+            int k2 = i2 / (i * 2 + 1);
+            BlockPos blockposition1 = PlayerRespawnLogic.getOverworldRespawnPos(worldserver, blockposition.getX() + j2 - i, blockposition.getZ() + k2 - i);
+
+            if (blockposition1 != null) {
+               return blockposition1;
+            }
+         }
+      }
+
+      return blockposition;
+   }
+   // CraftBukkit end
 
    private void fudgeSpawnLocation(ServerLevel p_9202_) {
       BlockPos blockpos = p_9202_.getSharedSpawnPos();
@@ -374,6 +_,76 @@
 
    }
 
+   public int newExp = 0;
+   public int newLevel = 0;
+   public int newTotalExp = 0;
+   public boolean keepLevel = false;
+
+   public void reset() {
+      float exp = 0;
+      boolean keepInventory = this.level().getGameRules().getBoolean(GameRules.RULE_KEEPINVENTORY);
+
+      if (this.keepLevel) { // CraftBukkit - SPIGOT-6687: Only use keepLevel (was pre-set with RULE_KEEPINVENTORY value in PlayerDeathEvent)
+         exp = this.experienceProgress;
+         this.newTotalExp = this.totalExperience;
+         this.newLevel = this.experienceLevel;
+      }
+
+      this.setHealth(this.getMaxHealth());
+      this.stopUsingItem(); // CraftBukkit - SPIGOT-6682: Clear active item on reset
+      this.setAirSupply(this.getMaxAirSupply()); // Paper
+      this.setRemainingFireTicks(0);
+      this.fallDistance = 0;
+      this.foodData = new FoodData();
+      this.experienceLevel = this.newLevel;
+      this.totalExperience = this.newTotalExp;
+      this.experienceProgress = 0;
+      this.deathTime = 0;
+      this.setArrowCount(0); // CraftBukkit - ArrowBodyCountChangeEvent
+      this.effectsDirty = true;
+      this.containerMenu = this.inventoryMenu;
+      this.lastHurtByPlayer = null;
+      this.lastHurtByMob = null;
+      this.combatTracker = new CombatTracker(this);
+      this.lastSentExp = -1;
+      if (this.keepLevel) { // CraftBukkit - SPIGOT-6687: Only use keepLevel (was pre-set with RULE_KEEPINVENTORY value in PlayerDeathEvent)
+         this.experienceProgress = exp;
+      } else {
+         this.giveExperiencePoints(this.newExp);
+      }
+      this.keepLevel = false;
+      this.setDeltaMovement(0, 0, 0); // CraftBukkit - SPIGOT-6948: Reset velocity on death
+   }
+
+   // SPIGOT-1903, MC-98153
+   public void forceSetPositionRotation(double x, double y, double z, float yaw, float pitch) {
+      this.moveTo(x, y, z, yaw, pitch);
+      this.connection.resetPosition();
+   }
+
+   // CraftBukkit start - World fallback code, either respawn location or global spawn
+   public void spawnIn(ServerLevel world) {
+      this.setLevel(world);
+      if (world == null) {
+         this.revive();
+         Vec3 position = null;
+         if (this.respawnDimension != null) {
+            world = this.server.getLevel(this.respawnDimension);
+            if (world != null && this.getRespawnPosition() != null) {
+               position = Player.findRespawnPositionAndUseSpawnBlock((ServerLevel) world, this.getRespawnPosition(), this.getRespawnAngle(), false, false).orElse(null);
+            }
+         }
+         if (world == null || position == null) {
+            world = MinecraftServer.getServer().overworld();
+            position = Vec3.atCenterOf(world.getSharedSpawnPos());
+         }
+         this.setLevel(world);
+         this.setPosRaw(position.x(), position.y(), position.z()); // Paper - don't register to chunks yet
+      }
+      this.gameMode.setLevel((ServerLevel) world);
+   }
+   // CraftBukkit end
+
    public void setExperiencePoints(int p_8986_) {
       float f = (float)this.getXpNeededForNextLevel();
       float f1 = (f - 1.0F) / f;
@@ -569,6 +_,7 @@
 
    public void die(DamageSource p_9035_) {
       this.gameEvent(GameEvent.ENTITY_DIE);
+      if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, p_9035_)) return;
       boolean flag = this.level().getGameRules().getBoolean(GameRules.RULE_SHOWDEATHMESSAGES);
       if (flag) {
          Component component = this.getCombatTracker().getDeathMessage();
@@ -715,11 +_,13 @@
    }
 
    @Nullable
-   public Entity changeDimension(ServerLevel p_9180_) {
+   public Entity changeDimension(ServerLevel p_9180_, net.minecraftforge.common.util.ITeleporter teleporter) {
+      if (!net.minecraftforge.event.ForgeEventFactory.onTravelToDimension(this, p_9180_.dimension()))
+         return null;
       this.isChangingDimension = true;
       ServerLevel serverlevel = this.serverLevel();
       ResourceKey<Level> resourcekey = serverlevel.dimension();
-      if (resourcekey == Level.END && p_9180_.dimension() == Level.OVERWORLD) {
+      if (resourcekey == Level.END && p_9180_.dimension() == Level.OVERWORLD && teleporter.isVanilla()) { //Forge: Fix non-vanilla teleporters triggering end credits
          this.unRide();
          this.serverLevel().removePlayerImmediately(this, Entity.RemovalReason.CHANGED_DIMENSION);
          if (!this.wonGame) {
@@ -736,13 +_,14 @@
          PlayerList playerlist = this.server.getPlayerList();
          playerlist.sendPlayerPermissionLevel(this);
          serverlevel.removePlayerImmediately(this, Entity.RemovalReason.CHANGED_DIMENSION);
-         this.unsetRemoved();
-         PortalInfo portalinfo = this.findDimensionEntryPoint(p_9180_);
+         this.revive();
+         PortalInfo portalinfo = teleporter.getPortalInfo(this, p_9180_, this::findDimensionEntryPoint);
          if (portalinfo != null) {
+            Entity e = teleporter.placeEntity(this, serverlevel, p_9180_, this.getYRot(), spawnPortal -> {//Forge: Start vanilla logic
             serverlevel.getProfiler().push("moving");
             if (resourcekey == Level.OVERWORLD && p_9180_.dimension() == Level.NETHER) {
                this.enteredNetherPosition = this.position();
-            } else if (p_9180_.dimension() == Level.END) {
+            } else if (spawnPortal && p_9180_.dimension() == Level.END) {
                this.createEndPlatform(p_9180_, BlockPos.containing(portalinfo.pos));
             }
 
@@ -754,6 +_,9 @@
             p_9180_.addDuringPortalTeleport(this);
             serverlevel.getProfiler().pop();
             this.triggerDimensionChangeTriggers(serverlevel);
+            return this;//forge: this is part of the ITeleporter patch
+            });//Forge: End vanilla logic
+            if (e != this) throw new java.lang.IllegalArgumentException(String.format(java.util.Locale.ENGLISH, "Teleporter %s returned not the player entity but instead %s, expected PlayerEntity %s", teleporter, e, this));
             this.connection.send(new ClientboundPlayerAbilitiesPacket(this.getAbilities()));
             playerlist.sendLevelInfo(this, p_9180_);
             playerlist.sendAllPlayerInfo(this);
@@ -762,10 +_,12 @@
                this.connection.send(new ClientboundUpdateMobEffectPacket(this.getId(), mobeffectinstance));
             }
 
+            if (teleporter.playTeleportSound(this, serverlevel, p_9180_))
             this.connection.send(new ClientboundLevelEventPacket(1032, BlockPos.ZERO, 0, false));
             this.lastSentExp = -1;
             this.lastSentHealth = -1.0F;
             this.lastSentFood = -1;
+            net.minecraftforge.event.ForgeEventFactory.onPlayerChangedDimension(this, resourcekey, p_9180_.dimension());
          }
 
          return this;
@@ -801,7 +_,7 @@
       }
    }
 
-   private void triggerDimensionChangeTriggers(ServerLevel p_9210_) {
+   public void triggerDimensionChangeTriggers(ServerLevel p_9210_) {
       ResourceKey<Level> resourcekey = p_9210_.dimension();
       ResourceKey<Level> resourcekey1 = this.level().dimension();
       CriteriaTriggers.CHANGED_DIMENSION.trigger(this, resourcekey, resourcekey1);
@@ -829,6 +_,9 @@
    }
 
    public Either<Player.BedSleepingProblem, Unit> startSleepInBed(BlockPos p_9115_) {
+      var optAt = java.util.Optional.of(p_9115_);
+      var ret = net.minecraftforge.event.ForgeEventFactory.onPlayerSleepInBed(this, optAt);
+      if (ret != null) return Either.left(ret);
       Direction direction = this.level().getBlockState(p_9115_).getValue(HorizontalDirectionalBlock.FACING);
       if (!this.isSleeping() && this.isAlive()) {
          if (!this.level().dimensionType().natural()) {
@@ -839,7 +_,7 @@
             return Either.left(Player.BedSleepingProblem.OBSTRUCTED);
          } else {
             this.setRespawnPosition(this.level().dimension(), p_9115_, this.getYRot(), false, true);
-            if (this.level().isDay()) {
+            if (!net.minecraftforge.event.ForgeEventFactory.onSleepingTimeCheck(this, optAt)) {
                return Either.left(Player.BedSleepingProblem.NOT_POSSIBLE_NOW);
             } else {
                if (!this.isCreative()) {
@@ -877,6 +_,7 @@
    }
 
    private boolean bedInRange(BlockPos p_9117_, Direction p_9118_) {
+      if (p_9118_ == null) return false;
       return this.isReachableBedBlock(p_9117_) || this.isReachableBedBlock(p_9117_.relative(p_9118_.getOpposite()));
    }
 
@@ -958,6 +_,7 @@
             this.connection.send(new ClientboundOpenScreenPacket(abstractcontainermenu.containerId, abstractcontainermenu.getType(), p_9033_.getDisplayName()));
             this.initMenu(abstractcontainermenu);
             this.containerMenu = abstractcontainermenu;
+            net.minecraftforge.event.ForgeEventFactory.onPlayerOpenContainer(this, this.containerMenu);
             return OptionalInt.of(this.containerCounter);
          }
       }
@@ -976,6 +_,7 @@
       this.connection.send(new ClientboundHorseScreenOpenPacket(this.containerCounter, p_9060_.getContainerSize(), p_9059_.getId()));
       this.containerMenu = new HorseInventoryMenu(this.containerCounter, this.getInventory(), p_9060_, p_9059_);
       this.initMenu(this.containerMenu);
+      net.minecraftforge.event.ForgeEventFactory.onPlayerOpenContainer(this, this.containerMenu);
    }
 
    public void openItemGui(ItemStack p_9082_, InteractionHand p_9083_) {
@@ -1001,6 +_,7 @@
    public void doCloseContainer() {
       this.containerMenu.removed(this);
       this.inventoryMenu.transferState(this.containerMenu);
+      net.minecraftforge.event.ForgeEventFactory.onPlayerCloseContainer(this, this.containerMenu);
       this.containerMenu = this.inventoryMenu;
    }
 
@@ -1127,13 +_,22 @@
       this.lastSentExp = -1;
       this.lastSentHealth = -1.0F;
       this.lastSentFood = -1;
-      this.recipeBook.copyOverData(p_9016_.recipeBook);
+      //this.recipeBook.copyOverData(p_9016_.recipeBook);
       this.seenCredits = p_9016_.seenCredits;
       this.enteredNetherPosition = p_9016_.enteredNetherPosition;
       this.chunkTrackingView = p_9016_.chunkTrackingView;
       this.setShoulderEntityLeft(p_9016_.getShoulderEntityLeft());
       this.setShoulderEntityRight(p_9016_.getShoulderEntityRight());
       this.setLastDeathLocation(p_9016_.getLastDeathLocation());
+
+      //Copy over a section of the Entity Data from the old player.
+      //Allows mods to specify data that persists after players respawn.
+      CompoundTag old = p_9016_.getPersistentData();
+      if (old.contains(PERSISTED_NBT_TAG))
+          getPersistentData().put(PERSISTED_NBT_TAG, old.get(PERSISTED_NBT_TAG));
+      net.minecraftforge.event.ForgeEventFactory.onPlayerClone(this, p_9016_, !p_9017_);
+      this.tabListHeader = p_9016_.tabListHeader;
+      this.tabListFooter = p_9016_.tabListFooter;
    }
 
    protected void onEffectAdded(MobEffectInstance p_143393_, @Nullable Entity p_143394_) {
@@ -1214,6 +_,7 @@
    }
 
    public boolean setGameMode(GameType p_143404_) {
+      p_143404_ = net.minecraftforge.common.ForgeHooks.onChangeGameType(this, this.gameMode.getGameModeForPlayer(), p_143404_);
       if (!this.gameMode.changeGameModeForPlayer(p_143404_)) {
          return false;
       } else {
@@ -1348,6 +_,8 @@
    public void setCamera(@Nullable Entity p_9214_) {
       Entity entity = this.getCamera();
       this.camera = (Entity)(p_9214_ == null ? this : p_9214_);
+      while (this.camera instanceof net.minecraftforge.entity.PartEntity<?> partEntity)
+         this.camera = partEntity.getParent(); // FORGE: fix MC-46486
       if (entity != this.camera) {
          Level level = this.camera.level();
          if (level instanceof ServerLevel) {
@@ -1387,7 +_,11 @@
 
    @Nullable
    public Component getTabListDisplayName() {
-      return null;
+      if (!this.hasTabListName) {
+         this.tabListDisplayName = net.minecraftforge.event.ForgeEventFactory.getPlayerTabListDisplayName(this);
+         this.hasTabListName = true;
+      }
+      return this.tabListDisplayName;
    }
 
    public void swing(InteractionHand p_9031_) {
@@ -1412,21 +_,23 @@
       this.stopRiding();
       if (p_9000_ == this.level()) {
          this.connection.teleport(p_9001_, p_9002_, p_9003_, p_9004_, p_9005_);
-      } else {
+      } else if (!net.minecraftforge.event.ForgeEventFactory.onTravelToDimension(this, p_9000_.dimension())) {
          ServerLevel serverlevel = this.serverLevel();
          LevelData leveldata = p_9000_.getLevelData();
          this.connection.send(new ClientboundRespawnPacket(this.createCommonSpawnInfo(p_9000_), (byte)3));
          this.connection.send(new ClientboundChangeDifficultyPacket(leveldata.getDifficulty(), leveldata.isDifficultyLocked()));
          this.server.getPlayerList().sendPlayerPermissionLevel(this);
          serverlevel.removePlayerImmediately(this, Entity.RemovalReason.CHANGED_DIMENSION);
-         this.unsetRemoved();
+         this.revive();
          this.moveTo(p_9001_, p_9002_, p_9003_, p_9004_, p_9005_);
          this.setServerLevel(p_9000_);
          p_9000_.addDuringCommandTeleport(this);
          this.triggerDimensionChangeTriggers(serverlevel);
          this.connection.teleport(p_9001_, p_9002_, p_9003_, p_9004_, p_9005_);
+         this.gameMode.setLevel(p_9000_);
          this.server.getPlayerList().sendLevelInfo(this, p_9000_);
          this.server.getPlayerList().sendAllPlayerInfo(this);
+         net.minecraftforge.event.ForgeEventFactory.onPlayerChangedDimension(this, serverlevel.dimension(), p_9000_.dimension());
       }
 
    }
@@ -1449,6 +_,8 @@
    }
 
    public void setRespawnPosition(ResourceKey<Level> p_9159_, @Nullable BlockPos p_9160_, float p_9161_, boolean p_9162_, boolean p_9163_) {
+      if (net.minecraftforge.event.ForgeEventFactory.onPlayerSpawnSet(this, p_9160_ == null ? Level.OVERWORLD : p_9159_, p_9160_, p_9162_))
+         return;
       if (p_9160_ != null) {
          boolean flag = p_9160_.equals(this.respawnPosition) && p_9159_.equals(this.respawnDimension);
          if (p_9163_ && !flag) {
@@ -1493,6 +_,9 @@
       if (itementity == null) {
          return null;
       } else {
+         if (captureDrops() != null)
+            captureDrops().add(itementity);
+         else
          this.level().addFreshEntity(itementity);
          ItemStack itemstack = itementity.getItem();
          if (p_9087_) {
@@ -1507,6 +_,75 @@
       }
    }
 
+   /**
+    * Returns the language last reported by the player as their local language.
+    * Defaults to en_us if the value is unknown.
+    */
+   public String getLanguage() {
+      return this.language;
+   }
+
+   private Component tabListHeader = Component.empty();
+   private Component tabListFooter = Component.empty();
+
+   public Component getTabListHeader() {
+       return this.tabListHeader;
+   }
+
+   /**
+    * Set the tab list header while preserving the footer.
+    *
+    * @param header the new header, or {@link Component#empty()} to clear
+    */
+   public void setTabListHeader(final Component header) {
+       this.setTabListHeaderFooter(header, this.tabListFooter);
+   }
+
+   public Component getTabListFooter() {
+       return this.tabListFooter;
+   }
+
+   /**
+    * Set the tab list footer while preserving the header.
+    *
+    * @param footer the new footer, or {@link Component#empty()} to clear
+    */
+   public void setTabListFooter(final Component footer) {
+       this.setTabListHeaderFooter(this.tabListHeader, footer);
+   }
+
+   /**
+    * Set the tab list header and footer at once.
+    *
+    * @param header the new header, or {@link Component#empty()} to clear
+    * @param footer the new footer, or {@link Component#empty()} to clear
+    */
+   public void setTabListHeaderFooter(final Component header, final Component footer) {
+       if (java.util.Objects.equals(header, this.tabListHeader)
+           && java.util.Objects.equals(footer, this.tabListFooter)) {
+           return;
+       }
+
+       this.tabListHeader = java.util.Objects.requireNonNull(header, "header");
+       this.tabListFooter = java.util.Objects.requireNonNull(footer, "footer");
+
+       this.connection.send(new net.minecraft.network.protocol.game.ClientboundTabListPacket(header, footer));
+   }
+
+   // We need this as tablistDisplayname may be null even if the event was fired.
+   private boolean hasTabListName = false;
+   private Component tabListDisplayName = null;
+   /**
+    * Force the name displayed in the tab list to refresh, by firing {@link net.minecraftforge.event.entity.player.PlayerEvent.TabListNameFormat}.
+    */
+   public void refreshTabListName() {
+      Component oldName = this.tabListDisplayName;
+      this.tabListDisplayName = net.minecraftforge.event.ForgeEventFactory.getPlayerTabListDisplayName(this);
+      if (!java.util.Objects.equals(oldName, this.tabListDisplayName)) {
+         this.getServer().getPlayerList().broadcastAll(new net.minecraft.network.protocol.game.ClientboundPlayerInfoUpdatePacket(net.minecraft.network.protocol.game.ClientboundPlayerInfoUpdatePacket.Action.UPDATE_DISPLAY_NAME, this));
+      }
+   }
+
    public TextFilter getTextFilter() {
       return this.textFilter;
    }
@@ -1566,11 +_,14 @@
 
    public boolean drop(boolean p_182295_) {
       Inventory inventory = this.getInventory();
+      ItemStack selected = inventory.getSelected();
+      if (selected.isEmpty() || !selected.onDroppedByPlayer(this)) return false;
+      if (isUsingItem() && getUsedItemHand() == InteractionHand.MAIN_HAND && (p_182295_ || selected.getCount() == 1)) stopUsingItem(); // Forge: fix MC-231097 on the serverside
       ItemStack itemstack = inventory.removeFromSelected(p_182295_);
       this.containerMenu.findSlot(inventory, inventory.selected).ifPresent((p_287377_) -> {
          this.containerMenu.setRemoteSlot(p_287377_, inventory.getSelected());
       });
-      return this.drop(itemstack, false, true) != null;
+      return net.minecraftforge.common.ForgeHooks.onPlayerTossEvent(this, itemstack, true) != null;
    }
 
    public boolean allowsListing() {
