--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -86,6 +_,8 @@
     @Override
     protected void randomTick(BlockState p_221139_, ServerLevel p_221140_, BlockPos p_221141_, RandomSource p_221142_) {
         int i = p_221139_.getValue(MOISTURE);
+        if (i > 0 && p_221140_.paperConfig().tickRates.wetFarmland != 1 && (p_221140_.paperConfig().tickRates.wetFarmland < 1 || (net.minecraft.server.MinecraftServer.currentTick + p_221141_.hashCode()) % p_221140_.paperConfig().tickRates.wetFarmland != 0)) { return; } // Paper - Configurable random tick rates for blocks
+        if (i == 0 && p_221140_.paperConfig().tickRates.dryFarmland != 1 && (p_221140_.paperConfig().tickRates.dryFarmland < 1 || (net.minecraft.server.MinecraftServer.currentTick + p_221141_.hashCode()) % p_221140_.paperConfig().tickRates.dryFarmland != 0)) { return; } // Paper - Configurable random tick rates for blocks
         if (!isNearWater(p_221140_, p_221141_) && !p_221140_.isRainingAt(p_221141_.above())) {
             if (i > 0) {
                 p_221140_.setBlock(p_221141_, p_221139_.setValue(MOISTURE, Integer.valueOf(i - 1)), 2);
@@ -100,10 +_,7 @@
     @Override
     public void fallOn(Level p_153227_, BlockState p_153228_, BlockPos p_153229_, Entity p_153230_, float p_153231_) {
         if (!p_153227_.isClientSide
-            && p_153227_.random.nextFloat() < p_153231_ - 0.5F
-            && p_153230_ instanceof LivingEntity
-            && (p_153230_ instanceof Player || p_153227_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))
-            && p_153230_.getBbWidth() * p_153230_.getBbWidth() * p_153230_.getBbHeight() > 0.512F) {
+            && net.neoforged.neoforge.common.CommonHooks.onFarmlandTrample(p_153227_, p_153229_, Blocks.DIRT.defaultBlockState(), p_153231_, p_153230_)) { // Forge: Move logic to Entity#canTrample
             turnToDirt(p_153230_, p_153228_, p_153227_, p_153229_);
         }
 
@@ -121,13 +_,14 @@
     }
 
     private static boolean isNearWater(LevelReader p_53259_, BlockPos p_53260_) {
+        BlockState state = p_53259_.getBlockState(p_53260_);
         for (BlockPos blockpos : BlockPos.betweenClosed(p_53260_.offset(-4, 0, -4), p_53260_.offset(4, 1, 4))) {
-            if (p_53259_.getFluidState(blockpos).is(FluidTags.WATER)) {
+            if (state.canBeHydrated(p_53259_, p_53260_, p_53259_.getFluidState(blockpos), blockpos)) {
                 return true;
             }
         }
 
-        return false;
+        return net.neoforged.neoforge.common.FarmlandWaterManager.hasBlockWaterTicket(p_53259_, p_53260_);
     }
 
     @Override
