--- a/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
@@ -16,6 +_,7 @@
         super(p_56817_);
     }
 
+    // Goldenforge note: snowrealmagic inject here, maybe we should use mixins to inject paper changes here
     private static boolean canBeGrass(BlockState p_56824_, LevelReader p_56825_, BlockPos p_56826_) {
         BlockPos blockpos = p_56826_.above();
         BlockState blockstate = p_56825_.getBlockState(blockpos);
@@ -25,7 +_,7 @@
             return false;
         } else {
             int i = LightEngine.getLightBlockInto(
-                p_56825_, p_56824_, p_56826_, blockstate, blockpos, Direction.UP, blockstate.getLightBlock(p_56825_, blockpos)
+                    p_56825_, p_56824_, p_56826_, blockstate, blockpos, Direction.UP, blockstate.getLightBlock(p_56825_, blockpos)
             );
             return i < p_56825_.getMaxLightLevel();
         }
@@ -51,7 +_,7 @@
                     BlockPos blockpos = p_222510_.offset(p_222511_.nextInt(3) - 1, p_222511_.nextInt(5) - 3, p_222511_.nextInt(3) - 1);
                     if (p_222509_.getBlockState(blockpos).is(Blocks.DIRT) && canPropagate(blockstate, p_222509_, blockpos)) {
                         p_222509_.setBlockAndUpdate(
-                            blockpos, blockstate.setValue(SNOWY, Boolean.valueOf(p_222509_.getBlockState(blockpos.above()).is(Blocks.SNOW)))
+                                blockpos, blockstate.setValue(SNOWY, Boolean.valueOf(p_222509_.getBlockState(blockpos.above()).is(Blocks.SNOW)))
                         );
                     }
                 }
