--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -8,6 +_,7 @@
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.suggestion.Suggestions;
 import com.mojang.logging.LogUtils;
+import io.papermc.paper.configuration.GlobalConfiguration;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMaps;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry;
 import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
@@ -24,6 +_,7 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
@@ -184,6 +_,7 @@
 import net.minecraft.world.phys.shapes.BooleanOp;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.Location;
 import org.slf4j.Logger;
 
 public class ServerGamePacketListenerImpl
@@ -236,14 +_,22 @@
     private final FutureChain chatMessageChain;
     private boolean waitingForSwitchToConfig;
 
+    private boolean hasMoved = false;
+    private double lastPosX = Double.MAX_VALUE;
+    private double lastPosY = Double.MAX_VALUE;
+    private double lastPosZ = Double.MAX_VALUE;
+    private float lastPitch = Float.MAX_VALUE;
+    private float lastYaw = Float.MAX_VALUE;
+
     public ServerGamePacketListenerImpl(MinecraftServer p_9770_, Connection p_9771_, ServerPlayer p_9772_, CommonListenerCookie p_301978_) {
         super(p_9770_, p_9771_, p_301978_);
+        super.player = p_9772_; // Goldenforge: mixin friendly
         this.chunkSender = new PlayerChunkSender(p_9771_.isMemoryConnection());
         this.player = p_9772_;
         p_9772_.connection = this;
         p_9772_.getTextFilter().join();
         this.signedMessageDecoder = SignedMessageChain.Decoder.unsigned(p_9772_.getUUID(), p_9770_::enforceSecureProfile);
-        this.chatMessageChain = new FutureChain(p_9770_);
+        this.chatMessageChain = new FutureChain(p_9770_.chatExecutor); // CraftBukkit - async chat
     }
 
     @Override
@@ -307,7 +_,8 @@
 
         if (this.player.getLastActionTime() > 0L
             && this.server.getPlayerIdleTimeout() > 0
-            && Util.getMillis() - this.player.getLastActionTime() > (long)this.server.getPlayerIdleTimeout() * 1000L * 60L) {
+            && Util.getMillis() - this.player.getLastActionTime() > (long)this.server.getPlayerIdleTimeout() * 1000L * 60L && !this.player.wonGame) { // Paper - Prevent AFK kick while watching end credits
+            this.player.resetLastActionTime(); // CraftBukkit - SPIGOT-854
             this.disconnect(Component.translatable("multiplayer.disconnect.idling"));
         }
     }
@@ -397,9 +_,9 @@
                 double d0 = entity.getX();
                 double d1 = entity.getY();
                 double d2 = entity.getZ();
-                double d3 = clampHorizontal(p_9876_.getX());
-                double d4 = clampVertical(p_9876_.getY());
-                double d5 = clampHorizontal(p_9876_.getZ());
+                double d3 = clampHorizontal(p_9876_.getX()); final double toX = d3; // Paper - OBFHELPER
+                double d4 = clampVertical(p_9876_.getY()); final double toY = d4; // Paper - OBFHELPER
+                double d5 = clampHorizontal(p_9876_.getZ()); final double toZ = d5; // Paper - OBFHELPER
                 float f = Mth.wrapDegrees(p_9876_.getYRot());
                 float f1 = Mth.wrapDegrees(p_9876_.getXRot());
                 double d6 = d3 - this.vehicleFirstGoodX;
@@ -407,6 +_,26 @@
                 double d8 = d5 - this.vehicleFirstGoodZ;
                 double d9 = entity.getDeltaMovement().lengthSqr();
                 double d10 = d6 * d6 + d7 * d7 + d8 * d8;
+                // Paper start - fix large move vectors killing the server
+                double currDeltaX = toX - d0;
+                double currDeltaY = toY - d1;
+                double currDeltaZ = toZ - d2;
+                d10 = Math.max(d10, (currDeltaX * currDeltaX + currDeltaY * currDeltaY + currDeltaZ * currDeltaZ) - 1);
+                double otherFieldX = toX - this.vehicleLastGoodX;
+                double otherFieldY = toY - this.vehicleLastGoodY;
+                double otherFieldZ = toZ - this.vehicleLastGoodZ;
+                d10 = Math.max(d10, (otherFieldX * otherFieldX + otherFieldY * otherFieldY + otherFieldZ * otherFieldZ) - 1);
+                // Paper end - fix large move vectors killing the server
+
+                // Paper start - Prevent moving into unloaded chunks
+                if (this.player.level().paperConfig().chunks.preventMovingIntoUnloadedChunks && (
+                        !serverlevel.areChunksLoadedForMove(this.player.getBoundingBox().expandTowards(new Vec3(toX, toY, toZ).subtract(this.player.position()))) ||
+                                !serverlevel.areChunksLoadedForMove(entity.getBoundingBox().expandTowards(new Vec3(toX, toY, toZ).subtract(entity.position())))
+                )) {
+                    this.connection.send(new ClientboundMoveVehiclePacket(entity));
+                    return;
+                }
+                // Paper end - Prevent moving into unloaded chunks
                 if (d10 - d9 > 100.0 && !this.isSingleplayerOwner()) {
                     LOGGER.warn("{} (vehicle of {}) moved too quickly! {},{},{}", entity.getName().getString(), this.player.getName().getString(), d6, d7, d8);
                     this.send(new ClientboundMoveVehiclePacket(entity));
@@ -439,9 +_,11 @@
                 }
 
                 entity.absMoveTo(d3, d4, d5, f, f1);
+                resyncPlayerWithVehicle(entity); // Neo - Resync player position on vehicle moving
                 boolean flag3 = serverlevel.noCollision(entity, entity.getBoundingBox().deflate(0.0625));
                 if (flag && (flag2 || !flag3)) {
                     entity.absMoveTo(d0, d1, d2, f, f1);
+                    resyncPlayerWithVehicle(entity); // Neo - Resync player position on vehicle moving
                     this.send(new ClientboundMoveVehiclePacket(entity));
                     return;
                 }
@@ -450,6 +_,7 @@
                 Vec3 vec3 = new Vec3(entity.getX() - d0, entity.getY() - d1, entity.getZ() - d2);
                 this.player.setKnownMovement(vec3);
                 this.player.checkMovementStatistics(vec3.x, vec3.y, vec3.z);
+                this.player.checkRidingStatistics(vec3.x, vec3.y, vec3.z); // Neo: check riding stats too as vanilla checks them in rideTick based on the assumption that Entity#rideTick will move the entity, which we break
                 this.clientVehicleIsFloating = d11 >= -0.03125
                     && !flag1
                     && !this.server.isFlightAllowed()
@@ -462,10 +_,50 @@
         }
     }
 
+    private void resyncPlayerWithVehicle(Entity vehicle) {
+        Vec3 oldPos = this.player.position();
+        float yRot = this.player.getYRot();
+        float xRot = this.player.getXRot();
+        float yHeadRot = this.player.getYHeadRot();
+
+        vehicle.positionRider(this.player);
+
+        // preserve old rotation and store old position in xo/yo/zo
+        this.player.setYRot(yRot);
+        this.player.setXRot(xRot);
+        this.player.setYHeadRot(yHeadRot);
+        this.player.xo = oldPos.x;
+        this.player.yo = oldPos.y;
+        this.player.zo = oldPos.z;
+    }
+
     private boolean noBlocksAround(Entity p_9794_) {
-        return p_9794_.level()
-            .getBlockStates(p_9794_.getBoundingBox().inflate(0.0625).expandTowards(0.0, -0.55, 0.0))
-            .allMatch(BlockBehaviour.BlockStateBase::isAir);
+        // Paper start - stop using streams, this is already a known fixed problem in Entity#move
+        AABB box = p_9794_.getBoundingBox().inflate(0.0625).expandTowards(0.0, -0.55, 0.0);
+        int minX = Mth.floor(box.minX);
+        int minY = Mth.floor(box.minY);
+        int minZ = Mth.floor(box.minZ);
+        int maxX = Mth.floor(box.maxX);
+        int maxY = Mth.floor(box.maxY);
+        int maxZ = Mth.floor(box.maxZ);
+
+        Level level = p_9794_.level();
+        BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
+
+        for (int y = minY; y <= maxY; ++y) {
+            for (int z = minZ; z <= maxZ; ++z) {
+                for (int x = minX; x <= maxX; ++x) {
+                    pos.set(x, y, z);
+                    BlockState blockState = level.getBlockStateIfLoaded(pos);
+                    if (blockState != null && !blockState.isAir()) {
+                        return false;
+                    }
+                }
+            }
+        }
+
+        return true;
+        // Paper end - stop using streams, this is already a known fixed problem in Entity#move
     }
 
     @Override
@@ -478,7 +_,7 @@
             }
 
             this.player
-                .absMoveTo(
+                    .absMoveTo(
                     this.awaitingPositionFromClient.x,
                     this.awaitingPositionFromClient.y,
                     this.awaitingPositionFromClient.z,
@@ -493,6 +_,7 @@
             }
 
             this.awaitingPositionFromClient = null;
+            this.player.serverLevel().getChunkSource().move(this.player); // CraftBukkit
         }
     }
 
@@ -878,7 +_,7 @@
                                 }
 
                                 if (!this.player.isChangingDimension()
-                                    && (!this.player.level().getGameRules().getBoolean(GameRules.RULE_DISABLE_ELYTRA_MOVEMENT_CHECK) || !flag)) {
+                                        && (!this.player.level().getGameRules().getBoolean(GameRules.RULE_DISABLE_ELYTRA_MOVEMENT_CHECK) || !flag)) {
                                     float f2 = flag ? 300.0F : 100.0F;
                                     if (d10 - d9 > (double)(f2 * (float)i) && !this.isSingleplayerOwner()) {
                                         LOGGER.warn("{} moved too quickly! {},{},{}", this.player.getName().getString(), d6, d7, d8);
@@ -909,6 +_,178 @@
                             d10 = d6 * d6 + d7 * d7 + d8 * d8;
                             boolean flag2 = false;
                             if (!this.player.isChangingDimension()
+                                    && d10 > 0.0625
+                                    && !this.player.isSleeping()
+                                    && !this.player.gameMode.isCreative()
+                                    && this.player.gameMode.getGameModeForPlayer() != GameType.SPECTATOR) {
+                                flag2 = true;
+                                LOGGER.warn("{} moved wrongly!", this.player.getName().getString());
+                            }
+
+                            if (this.player.noPhysics
+                                    || this.player.isSleeping()
+                                    || (!flag2 || !serverlevel.noCollision(this.player, aabb))
+                                    && !this.isPlayerCollidingWithAnythingNew(serverlevel, aabb, d0, d1, d2)) {
+                                this.player.absMoveTo(d0, d1, d2, f, f1);
+                                boolean flag3 = this.player.isAutoSpinAttack();
+                                this.clientIsFloating = d7 >= -0.03125
+                                        && !flag1
+                                        && this.player.gameMode.getGameModeForPlayer() != GameType.SPECTATOR
+                                        && !this.server.isFlightAllowed()
+                                        && !this.player.getAbilities().mayfly
+                                        && !this.player.hasEffect(MobEffects.LEVITATION)
+                                        && !flag
+                                        && !flag3
+                                        && this.noBlocksAround(this.player);
+                                this.player.serverLevel().getChunkSource().move(this.player);
+                                Vec3 vec3 = new Vec3(this.player.getX() - d3, this.player.getY() - d4, this.player.getZ() - d5);
+                                this.player.setOnGroundWithMovement(p_9874_.isOnGround(), vec3);
+                                this.player.doCheckFallDamage(this.player.getX() - d3, this.player.getY() - d4, this.player.getZ() - d5, p_9874_.isOnGround());
+                                this.player.setKnownMovement(vec3);
+                                if (flag4) {
+                                    this.player.resetFallDistance();
+                                }
+
+                                if (p_9874_.isOnGround()
+                                        || this.player.hasLandedInLiquid()
+                                        || this.player.onClimbable()
+                                        || this.player.isSpectator()
+                                        || flag
+                                        || flag3) {
+                                    this.player.tryResetCurrentImpulseContext();
+                                }
+
+                                this.player.checkMovementStatistics(this.player.getX() - d3, this.player.getY() - d4, this.player.getZ() - d5);
+                                this.lastGoodX = this.player.getX();
+                                this.lastGoodY = this.player.getY();
+                                this.lastGoodZ = this.player.getZ();
+                            } else {
+                                this.teleport(d3, d4, d5, f, f1);
+                                this.player.doCheckFallDamage(this.player.getX() - d3, this.player.getY() - d4, this.player.getZ() - d5, p_9874_.isOnGround());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+   // @Override
+    public void handleMovePlayer_paper(ServerboundMovePlayerPacket p_9874_) {
+        PacketUtils.ensureRunningOnSameThread(p_9874_, this, this.player.serverLevel());
+        if (containsInvalidValues(p_9874_.getX(0.0), p_9874_.getY(0.0), p_9874_.getZ(0.0), p_9874_.getYRot(0.0F), p_9874_.getXRot(0.0F))) {
+            this.disconnect(Component.translatable("multiplayer.disconnect.invalid_player_movement"));
+        } else {
+            ServerLevel serverlevel = this.player.serverLevel();
+            if (!this.player.wonGame && !this.player.isImmobile()) { // CraftBukkit
+                if (this.tickCount == 0) {
+                    this.resetPosition();
+                }
+
+                if (!this.updateAwaitingTeleport()) {
+                    double d0 = clampHorizontal(p_9874_.getX(this.player.getX())); final double toX = d0; // Paper - OBFHELPER
+                    double d1 = clampVertical(p_9874_.getY(this.player.getY())); final double toY = d1; // Paper - OBFHELPER
+                    double d2 = clampHorizontal(p_9874_.getZ(this.player.getZ())); final double toZ = d2; // Paper - OBFHELPER
+                    float f = Mth.wrapDegrees(p_9874_.getYRot(this.player.getYRot()));
+                    float f1 = Mth.wrapDegrees(p_9874_.getXRot(this.player.getXRot()));
+                    if (this.player.isPassenger()) {
+                        this.player.absMoveTo(this.player.getX(), this.player.getY(), this.player.getZ(), f, f1);
+                        this.player.serverLevel().getChunkSource().move(this.player);
+                    } else {
+                        // CraftBukkit - Make sure the move is valid but then reset it for plugins to modify
+                        double prevX = this.player.getX();
+                        double prevY = this.player.getY();
+                        double prevZ = this.player.getZ();
+                        float prevYaw = this.player.getYRot();
+                        float prevPitch = this.player.getXRot();
+                        // CraftBukkit end
+                        double d3 = this.player.getX();
+                        double d4 = this.player.getY();
+                        double d5 = this.player.getZ();
+                        double d6 = d0 - this.firstGoodX;
+                        double d7 = d1 - this.firstGoodY;
+                        double d8 = d2 - this.firstGoodZ;
+                        double d9 = this.player.getDeltaMovement().lengthSqr();
+                        double d10 = d6 * d6 + d7 * d7 + d8 * d8;
+                        // Paper start - fix large move vectors killing the server
+                        double currDeltaX = toX - d3;
+                        double currDeltaY = toY - d4;
+                        double currDeltaZ = toZ - d5;
+                        d10 = Math.max(d10, (currDeltaX * currDeltaX + currDeltaY * currDeltaY + currDeltaZ * currDeltaZ) - 1);
+                        double otherFieldX = d0 - this.lastGoodX;
+                        double otherFieldY = d1 - this.lastGoodY;
+                        double otherFieldZ = d2 - this.lastGoodZ;
+                        d10 = Math.max(d10, (otherFieldX * otherFieldX + otherFieldY * otherFieldY + otherFieldZ * otherFieldZ) - 1);
+                        // Paper end - fix large move vectors killing the server
+                        if (this.player.isSleeping()) {
+                            if (d10 > 1.0) {
+                                this.teleport(this.player.getX(), this.player.getY(), this.player.getZ(), f, f1);
+                            }
+                        } else {
+                            boolean flag = this.player.isFallFlying();
+                            if (serverlevel.tickRateManager().runsNormally()) {
+                                this.receivedMovePacketCount++;
+                                int i = this.receivedMovePacketCount - this.knownMovePacketCount;
+                                if (i > 5) {
+                                    LOGGER.debug("{} is sending move packets too frequently ({} packets since last tick)", this.player.getName().getString(), i);
+                                    i = 1;
+                                }
+
+                                double speed;
+                                if (this.player.getAbilities().flying) {
+                                    speed = this.player.getAbilities().flyingSpeed * 20f;
+                                } else {
+                                    speed = this.player.getAbilities().walkingSpeed * 10f;
+                                }
+                                speed *= 2f; // TODO: Get the speed of the vehicle instead of the player
+
+
+                                // Paper start - Prevent moving into unloaded chunks
+                                if (this.player.level().paperConfig().chunks.preventMovingIntoUnloadedChunks && (this.player.getX() != toX || this.player.getZ() != toZ) && !serverlevel.areChunksLoadedForMove(this.player.getBoundingBox().expandTowards(new Vec3(toX, toY, toZ).subtract(this.player.position())))) {
+                                    this.teleport(this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot(), Collections.emptySet());
+                                    return;
+                                }
+                                // Paper end - Prevent moving into unloaded chunks
+
+                                if (!this.player.isChangingDimension()
+                                    && (!this.player.level().getGameRules().getBoolean(GameRules.RULE_DISABLE_ELYTRA_MOVEMENT_CHECK) || !flag)) {
+                                    float f2 = flag ? 300.0F : 100.0F;
+                                    if (d10 - d9 > Math.max(100.0, Math.pow((double) (GlobalConfiguration.get().misc.movedTooQuicklyMultiplier * (float) i * speed), 2)) && !this.isSingleplayerOwner()) {
+                                        LOGGER.warn("{} moved too quickly! {},{},{}", this.player.getName().getString(), d6, d7, d8);
+                                        this.teleport(this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot());
+                                        return;
+                                    }
+                                }
+                            }
+
+                            AABB aabb = this.player.getBoundingBox();
+                            d6 = d0 - this.lastGoodX;
+                            d7 = d1 - this.lastGoodY;
+                            d8 = d2 - this.lastGoodZ;
+                            boolean flag4 = d7 > 0.0;
+                            if (this.player.onGround() && !p_9874_.isOnGround() && flag4) {
+                                this.player.jumpFromGround();
+                            }
+
+                            boolean flag1 = this.player.verticalCollisionBelow;
+                            this.player.move(MoverType.PLAYER, new Vec3(d6, d7, d8));
+                            this.player.onGround = p_9874_.isOnGround(); // CraftBukkit - SPIGOT-5810, SPIGOT-5835, SPIGOT-6828: reset by this.player.move
+                            boolean didCollide = toX != this.player.getX() || toY != this.player.getY() || toZ != this.player.getZ(); // Paper - needed here as the difference in Y can be reset - also note: this is only a guess at whether collisions took place, floating point errors can make this true when it shouldn't be...
+                            // Paper start - prevent position desync
+                            if (this.awaitingPositionFromClient != null) {
+                                return; // ... thanks Mojang for letting move calls teleport across dimensions.
+                            }
+                            // Paper end - prevent position desync
+                            d6 = d0 - this.player.getX();
+                            d7 = d1 - this.player.getY();
+                            if (d7 > -0.5 || d7 < 0.5) {
+                                d7 = 0.0;
+                            }
+
+                            d8 = d2 - this.player.getZ();
+                            d10 = d6 * d6 + d7 * d7 + d8 * d8;
+                            boolean flag2 = false;
+                            if (!this.player.isChangingDimension()
                                 && d10 > 0.0625
                                 && !this.player.isSleeping()
                                 && !this.player.gameMode.isCreative()
@@ -917,17 +_,69 @@
                                 LOGGER.warn("{} moved wrongly!", this.player.getName().getString());
                             }
 
-                            if (this.player.noPhysics
-                                || this.player.isSleeping()
-                                || (!flag2 || !serverlevel.noCollision(this.player, aabb))
-                                    && !this.isPlayerCollidingWithAnythingNew(serverlevel, aabb, d0, d1, d2)) {
+                            // Paper start - Add fail move event
+                            // Paper start - optimise out extra getCubes
+                            boolean teleportBack = !this.player.noPhysics && !this.player.isSleeping() && flag2;
+                            this.player.absMoveTo(d0, d1, d2, f, f1); // prevent desync by tping to the set position, dropped for unknown reasons by mojang
+                            if (!this.player.noPhysics && !this.player.isSleeping() && !teleportBack) {
+                                AABB newBox = this.player.getBoundingBox();
+                                if (didCollide || !aabb.equals(newBox)) {
+                                    // note: only call after setLocation, or else getBoundingBox is wrong
+                                    teleportBack = this.hasNewCollision(serverlevel, this.player, aabb, newBox);
+                                } // else: no collision at all detected, why do we care?
+                            }
+                            // Paper end - optimise out extra getCubes
+                            if (teleportBack) {
+                                teleportBack = false;
+                            }
+                            if (!teleportBack) {
+                                // CraftBukkit start - fire PlayerMoveEvent
+                                // Reset to old location first
+                                this.player.absMoveTo(prevX, prevY, prevZ, prevYaw, prevPitch);
+
+                                if (!this.hasMoved) {
+                                    this.lastPosX = prevX;
+                                    this.lastPosY = prevY;
+                                    this.lastPosZ = prevZ;
+                                    this.lastYaw = prevYaw;
+                                    this.lastPitch = prevPitch;
+                                    this.hasMoved = true;
+                                }
+
+                                Location to = Location.toBukkit(this.player.position(), player.serverLevel(), this.player.yRot, this.player.getXRot()); // Start off the To location as the Players current location.
+
+                                // If the packet contains movement information then we update the To location with the correct XYZ.
+                                if (p_9874_.hasPosition()) {
+                                    to.setX(p_9874_.getX(0.0));
+                                    to.setY(p_9874_.getY(0.0));
+                                    to.setZ(p_9874_.getZ(0.0));
+                                }
+
+                                // If the packet contains look information then we update the To location with the correct Yaw & Pitch.
+                                if (p_9874_.hasRotation()) {
+                                    to.setYaw(p_9874_.getYRot(0.0F));
+                                    to.setPitch(p_9874_.getXRot(0.0F));
+                                }
+
+                                // Prevent 40 event-calls for less than a single pixel of movement >.>
+                                double delta = Math.pow(this.lastPosX - to.getX(), 2) + Math.pow(this.lastPosY - to.getY(), 2) + Math.pow(this.lastPosZ - to.getZ(), 2);
+                                float deltaAngle = Math.abs(this.lastYaw - to.getYaw()) + Math.abs(this.lastPitch - to.getPitch());
+
+                                if ((delta > 1f / 256 || deltaAngle > 10f) && !this.player.isImmobile()) {
+                                    this.lastPosX = to.getX();
+                                    this.lastPosY = to.getY();
+                                    this.lastPosZ = to.getZ();
+                                    this.lastYaw = to.getYaw();
+                                    this.lastPitch = to.getPitch();
+                                }
+                                // CraftBukkit end
                                 this.player.absMoveTo(d0, d1, d2, f, f1);
                                 boolean flag3 = this.player.isAutoSpinAttack();
                                 this.clientIsFloating = d7 >= -0.03125
                                     && !flag1
                                     && this.player.gameMode.getGameModeForPlayer() != GameType.SPECTATOR
                                     && !this.server.isFlightAllowed()
-                                    && !this.player.getAbilities().mayfly
+                                    && !this.player.mayFly()
                                     && !this.player.hasEffect(MobEffects.LEVITATION)
                                     && !flag
                                     && !flag3
@@ -967,7 +_,7 @@
 
     private boolean updateAwaitingTeleport() {
         if (this.awaitingPositionFromClient != null) {
-            if (this.tickCount - this.awaitingTeleportTime > 20) {
+            if (false && this.tickCount - this.awaitingTeleportTime > 20) { // Paper - this will greatly screw with clients with > 1000ms RTT
                 this.awaitingTeleportTime = this.tickCount;
                 this.teleport(
                     this.awaitingPositionFromClient.x,
@@ -985,6 +_,33 @@
         }
     }
 
+    // Paper start - optimise out extra getCubes
+    private boolean hasNewCollision(final ServerLevel level, final Entity entity, final AABB oldBox, final AABB newBox) {
+        final List<AABB> collisionsBB = new java.util.ArrayList<>();
+        final List<VoxelShape> collisionsVoxel = new java.util.ArrayList<>();
+        ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisions(
+                level, entity, newBox, collisionsVoxel, collisionsBB,
+                ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_COLLIDE_WITH_UNLOADED_CHUNKS | ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER,
+                null, null
+        );
+
+        for (int i = 0, len = collisionsBB.size(); i < len; ++i) {
+            final AABB box = collisionsBB.get(i);
+            if (!ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersect(box, oldBox)) {
+                return true;
+            }
+        }
+
+        for (int i = 0, len = collisionsVoxel.size(); i < len; ++i) {
+            final VoxelShape voxel = collisionsVoxel.get(i);
+            if (!ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersectNoEmpty(voxel, oldBox)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+    // Paper end - optimise out extra getCubes
     private boolean isPlayerCollidingWithAnythingNew(LevelReader p_289008_, AABB p_288986_, double p_288990_, double p_288991_, double p_288967_) {
         AABB aabb = this.player.getBoundingBox().move(p_288990_ - this.player.getX(), p_288991_ - this.player.getY(), p_288967_ - this.player.getZ());
         Iterable<VoxelShape> iterable = p_289008_.getCollisions(this.player, aabb.deflate(1.0E-5F));
@@ -1024,6 +_,7 @@
     @Override
     public void handlePlayerAction(ServerboundPlayerActionPacket p_9889_) {
         PacketUtils.ensureRunningOnSameThread(p_9889_, this, this.player.serverLevel());
+        if (this.player.isImmobile()) return; // CraftBukkit
         BlockPos blockpos = p_9889_.getPos();
         this.player.resetLastActionTime();
         ServerboundPlayerActionPacket.Action serverboundplayeractionpacket$action = p_9889_.getAction();
@@ -1031,8 +_,10 @@
             case SWAP_ITEM_WITH_OFFHAND:
                 if (!this.player.isSpectator()) {
                     ItemStack itemstack = this.player.getItemInHand(InteractionHand.OFF_HAND);
-                    this.player.setItemInHand(InteractionHand.OFF_HAND, this.player.getItemInHand(InteractionHand.MAIN_HAND));
-                    this.player.setItemInHand(InteractionHand.MAIN_HAND, itemstack);
+                    var event = net.neoforged.neoforge.common.CommonHooks.onLivingSwapHandItems(this.player);
+                    if (event.isCanceled()) return;
+                    this.player.setItemInHand(InteractionHand.OFF_HAND, event.getItemSwappedToOffHand());
+                    this.player.setItemInHand(InteractionHand.MAIN_HAND, event.getItemSwappedToMainHand());
                     this.player.stopUsingItem();
                 }
 
@@ -1055,12 +_,18 @@
             case START_DESTROY_BLOCK:
             case ABORT_DESTROY_BLOCK:
             case STOP_DESTROY_BLOCK:
+                // Paper start - Don't allow digging into unloaded chunks
+                if (this.player.level().getChunkIfLoadedImmediately(blockpos.getX() >> 4, blockpos.getZ() >> 4) == null) {
+                    this.player.connection.ackBlockChangesUpTo(p_9889_.getSequence());
+                    return;
+                }
+                // Paper end - Don't allow digging into unloaded chunks
                 this.player
                     .gameMode
                     .handleBlockBreakAction(
                         blockpos, serverboundplayeractionpacket$action, p_9889_.getDirection(), this.player.level().getMaxBuildHeight(), p_9889_.getSequence()
                     );
-                this.player.connection.ackBlockChangesUpTo(p_9889_.getSequence());
+                this.player.connection.ackBlockChangesUpTo =p_9889_.getSequence();
                 return;
             default:
                 throw new IllegalArgumentException("Invalid player action");
@@ -1079,13 +_,19 @@
     @Override
     public void handleUseItemOn(ServerboundUseItemOnPacket p_9930_) {
         PacketUtils.ensureRunningOnSameThread(p_9930_, this, this.player.serverLevel());
-        this.player.connection.ackBlockChangesUpTo(p_9930_.getSequence());
+        if (this.player.isImmobile()) return; // CraftBukkit
+        this.player.connection.ackBlockChangesUpTo = p_9930_.getSequence();
         ServerLevel serverlevel = this.player.serverLevel();
         InteractionHand interactionhand = p_9930_.getHand();
         ItemStack itemstack = this.player.getItemInHand(interactionhand);
         if (itemstack.isItemEnabled(serverlevel.enabledFeatures())) {
             BlockHitResult blockhitresult = p_9930_.getHitResult();
             Vec3 vec3 = blockhitresult.getLocation();
+            // Paper start - improve distance check
+            if (!Double.isFinite(vec3.x()) || !Double.isFinite(vec3.y()) || !Double.isFinite(vec3.z())) {
+                return;
+            }
+            // Paper end - improve distance check
             BlockPos blockpos = blockhitresult.getBlockPos();
             if (this.player.canInteractWithBlock(blockpos, 1.0)) {
                 Vec3 vec31 = vec3.subtract(Vec3.atCenterOf(blockpos));
@@ -1112,7 +_,7 @@
                             } else if (interactionresult.shouldSwing()) {
                                 this.player.swing(interactionhand, true);
                             }
-                        }
+                        } else { this.player.containerMenu.sendAllDataToRemote(); } // Paper - Fix inventory desync; MC-99075
                     } else {
                         Component component1 = Component.translatable("build.tooHigh", i - 1).withStyle(ChatFormatting.RED);
                         this.player.sendSystemMessage(component1, true);
@@ -1135,6 +_,7 @@
     @Override
     public void handleUseItem(ServerboundUseItemPacket p_9932_) {
         PacketUtils.ensureRunningOnSameThread(p_9932_, this, this.player.serverLevel());
+        if (this.player.isImmobile()) return; // CraftBukkit
         this.ackBlockChangesUpTo(p_9932_.getSequence());
         ServerLevel serverlevel = this.player.serverLevel();
         InteractionHand interactionhand = p_9932_.getHand();
@@ -1178,6 +_,13 @@
 
     @Override
     public void onDisconnect(DisconnectionDetails p_350554_) {
+        // CraftBukkit start - Rarely it would send a disconnect line twice
+        if (this.processedDisconnect) {
+            return;
+        } else {
+            this.processedDisconnect = true;
+        }
+        // CraftBukkit end
         LOGGER.info("{} lost connection: {}", this.player.getName().getString(), p_350554_.reason().getString());
         this.removePlayerFromWorld();
         super.onDisconnect(p_350554_);
@@ -1205,6 +_,7 @@
     @Override
     public void handleSetCarriedItem(ServerboundSetCarriedItemPacket p_9909_) {
         PacketUtils.ensureRunningOnSameThread(p_9909_, this, this.player.serverLevel());
+        if (this.player.isImmobile()) return; // CraftBukkit
         if (p_9909_.getSlot() >= 0 && p_9909_.getSlot() < Inventory.getSelectionSize()) {
             if (this.player.getInventory().selected != p_9909_.getSlot() && this.player.getUsedItemHand() == InteractionHand.MAIN_HAND) {
                 this.player.stopUsingItem();
@@ -1219,6 +_,12 @@
 
     @Override
     public void handleChat(ServerboundChatPacket p_9841_) {
+        // CraftBukkit start - async chat
+        // SPIGOT-3638
+        if (this.server.isStopped()) {
+            return;
+        }
+        // CraftBukkit end
         Optional<LastSeenMessages> optional = this.unpackAndApplyLastSeen(p_9841_.lastSeenMessages());
         if (!optional.isEmpty()) {
             this.tryHandleChat(p_9841_.message(), () -> {
@@ -1230,13 +_,14 @@
                     return;
                 }
 
-                CompletableFuture<FilteredText> completablefuture = this.filterTextPacket(playerchatmessage.signedContent());
-                Component component = this.server.getChatDecorator().decorate(this.player, playerchatmessage.decoratedContent());
+                CompletableFuture<FilteredText> completablefuture = this.filterTextPacket(playerchatmessage.signedContent()).thenApplyAsync(Function.identity(), this.server.chatExecutor); // CraftBukkit - async chat
+                Component component = net.neoforged.neoforge.common.CommonHooks.getServerChatSubmittedDecorator().decorate(this.player, playerchatmessage.decoratedContent());
                 this.chatMessageChain.append(completablefuture, p_300785_ -> {
+                    if (component == null) return; // Forge: ServerChatEvent was canceled if this is null.
                     PlayerChatMessage playerchatmessage1 = playerchatmessage.withUnsignedContent(component).filter(p_300785_.mask());
                     this.broadcastChatMessage(playerchatmessage1);
                 });
-            });
+            }, false); // CraftBukkit - async chat
         }
     }
 
@@ -1245,7 +_,7 @@
         this.tryHandleChat(p_215225_.command(), () -> {
             this.performUnsignedChatCommand(p_215225_.command());
             this.detectRateSpam();
-        });
+        }, true); // CraftBukkit - sync commands
     }
 
     private void performUnsignedChatCommand(String p_338482_) {
@@ -1267,7 +_,7 @@
             this.tryHandleChat(p_338604_.command(), () -> {
                 this.performSignedChatCommand(p_338604_, optional.get());
                 this.detectRateSpam();
-            });
+            }, true); // CraftBukkit - sync commands
         }
     }
 
@@ -1348,14 +_,20 @@
         return commanddispatcher.parse(p_242938_, this.player.createCommandSourceStack());
     }
 
-    private void tryHandleChat(String p_338775_, Runnable p_338235_) {
+    private void tryHandleChat(String p_338775_, Runnable p_338235_, boolean sync) { // CraftBukkit
         if (isChatMessageIllegal(p_338775_)) {
             this.disconnect(Component.translatable("multiplayer.disconnect.illegal_characters"));
         } else if (this.player.getChatVisibility() == ChatVisiblity.HIDDEN) {
             this.send(new ClientboundSystemChatPacket(Component.translatable("chat.disabled.options").withStyle(ChatFormatting.RED), false));
         } else {
             this.player.resetLastActionTime();
-            this.server.execute(p_338235_);
+            // CraftBukkit start
+            if (sync) {
+                this.server.execute(p_338235_);
+            } else {
+                p_338235_.run();
+            }
+            // CraftBukkit end
         }
     }
 
@@ -1522,6 +_,7 @@
     @Override
     public void handleInteract(ServerboundInteractPacket p_9866_) {
         PacketUtils.ensureRunningOnSameThread(p_9866_, this, this.player.serverLevel());
+        if (this.player.isImmobile()) return; // CraftBukkit
         final ServerLevel serverlevel = this.player.serverLevel();
         final Entity entity = p_9866_.getTarget(serverlevel);
         this.player.resetLastActionTime();
@@ -1532,7 +_,7 @@
             }
 
             AABB aabb = entity.getBoundingBox();
-            if (this.player.canInteractWithEntity(aabb, 1.0)) {
+            if (this.player.canInteractWithEntity(aabb, io.papermc.paper.configuration.GlobalConfiguration.get().misc.clientInteractionLeniencyDistance.or(3.0))) { // Paper - configurable lenience value for interact range
                 p_9866_.dispatch(
                     new ServerboundInteractPacket.Handler() {
                         private void performInteraction(InteractionHand p_143679_, ServerGamePacketListenerImpl.EntityInteraction p_143680_) {
@@ -1561,7 +_,11 @@
 
                         @Override
                         public void onInteraction(InteractionHand p_143682_, Vec3 p_143683_) {
-                            this.performInteraction(p_143682_, (p_143686_, p_143687_, p_143688_) -> p_143687_.interactAt(p_143686_, p_143683_, p_143688_));
+                            this.performInteraction(p_143682_, (p_143686_, p_143687_, p_143688_) -> {
+                                InteractionResult onInteractEntityAtResult = net.neoforged.neoforge.common.CommonHooks.onInteractEntityAt(player, entity, p_143683_, p_143682_);
+                                if (onInteractEntityAtResult != null) return onInteractEntityAtResult;
+                                return p_143687_.interactAt(p_143686_, p_143683_, p_143688_);
+                            });
                         }
 
                         @Override
@@ -1712,7 +_,12 @@
             if (customdata.contains("x") && customdata.contains("y") && customdata.contains("z")) {
                 BlockPos blockpos = BlockEntity.getPosFromTag(customdata.getUnsafe());
                 if (this.player.level().isLoaded(blockpos)) {
-                    BlockEntity blockentity = this.player.level().getBlockEntity(blockpos);
+                    // Paper start - Prevent tile entity copies loading chunks
+                    BlockEntity blockentity = null;
+                    if (this.player.distanceToSqr(blockpos.getX(), blockpos.getY(), blockpos.getZ()) < 32 * 32 && this.player.serverLevel().isLoadedAndInBounds(blockpos)) {
+                        blockentity = this.player.level().getBlockEntity(blockpos);
+                    }
+                    // Paper end - Prevent tile entity copies loading chunks
                     if (blockentity != null) {
                         blockentity.saveToItem(itemstack, this.player.level().registryAccess());
                     }
@@ -1753,13 +_,15 @@
     @Override
     public void handlePlayerAbilities(ServerboundPlayerAbilitiesPacket p_9887_) {
         PacketUtils.ensureRunningOnSameThread(p_9887_, this, this.player.serverLevel());
-        this.player.getAbilities().flying = p_9887_.isFlying() && this.player.getAbilities().mayfly;
+        this.player.getAbilities().flying = p_9887_.isFlying() && this.player.mayFly();
     }
 
     @Override
     public void handleClientInformation(ServerboundClientInformationPacket p_301979_) {
         PacketUtils.ensureRunningOnSameThread(p_301979_, this, this.player.serverLevel());
+        net.minecraft.server.level.ClientInformation oldInfo = this.player.clientInformation();
         this.player.updateOptions(p_301979_.information());
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.ClientInformationUpdatedEvent(this.player, oldInfo, p_301979_.information()));
     }
 
     @Override
@@ -1812,7 +_,7 @@
             this.connection
                 .setupInboundProtocol(
                     ConfigurationProtocols.SERVERBOUND,
-                    new ServerConfigurationPacketListenerImpl(this.server, this.connection, this.createCookie(this.player.clientInformation()))
+                    new ServerConfigurationPacketListenerImpl(this.server, this.connection, this.createCookie(this.player.clientInformation(), this.connectionType)).setPlayer(this.player) // Goldenforge: mixin friendly injection
                 );
         }
     }
@@ -1847,6 +_,7 @@
 
     @Override
     public void handleCustomPayload(ServerboundCustomPayloadPacket p_333887_) {
+        super.handleCustomPayload(p_333887_); // Neo: Call super to invoke modded payload handling.
     }
 
     @Override
