--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -62,6 +_,7 @@
 import net.minecraft.world.level.storage.loot.LootTable;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
+import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
@@ -95,6 +_,17 @@
       this.f_60449_ = p_60452_.f_60893_;
       this.f_60438_ = p_60452_.f_60903_;
       this.f_60439_ = p_60452_;
+      final ResourceLocation lootTableCache = p_60452_.f_60894_;
+      if (lootTableCache != null) {
+         this.lootTableSupplier = () -> lootTableCache;
+      } else if (p_60452_.lootTableSupplier != null) {
+         this.lootTableSupplier = p_60452_.lootTableSupplier;
+      } else {
+         this.lootTableSupplier = () -> {
+            ResourceLocation registryName = net.minecraftforge.registries.ForgeRegistries.BLOCKS.getKey((Block) this);
+            return new ResourceLocation(registryName.m_135827_(), "blocks/" + registryName.m_135815_());
+         };
+      }
    }
 
    /** @deprecated */
@@ -143,7 +_,7 @@
    /** @deprecated */
    @Deprecated
    public void m_6810_(BlockState p_60515_, Level p_60516_, BlockPos p_60517_, BlockState p_60518_, boolean p_60519_) {
-      if (p_60515_.m_155947_() && !p_60515_.m_60713_(p_60518_.m_60734_())) {
+      if (p_60515_.m_155947_() && (!p_60515_.m_60713_(p_60518_.m_60734_()) || !p_60518_.m_155947_())) {
          p_60516_.m_46747_(p_60517_);
       }
 
@@ -220,7 +_,7 @@
    /** @deprecated */
    @Deprecated
    public boolean m_6864_(BlockState p_60470_, BlockPlaceContext p_60471_) {
-      return this.f_60442_.m_76336_() && (p_60471_.m_43722_().m_41619_() || !p_60471_.m_43722_().m_150930_(this.m_5456_()));
+      return p_60470_.m_60767_().m_76336_() && (p_60471_.m_43722_().m_41619_() || p_60471_.m_43722_().m_41720_() != this.m_5456_());
    }
 
    /** @deprecated */
@@ -350,8 +_,8 @@
       if (f == -1.0F) {
          return 0.0F;
       } else {
-         int i = p_60467_.m_36298_(p_60466_) ? 30 : 100;
-         return p_60467_.m_36281_(p_60466_) / f / (float)i;
+         int i = net.minecraftforge.common.ForgeHooks.isCorrectToolForDrops(p_60466_, p_60467_) ? 30 : 100;
+         return p_60467_.getDigSpeed(p_60466_, p_60469_) / f / (float)i;
       }
    }
 
@@ -384,8 +_,7 @@
 
    public final ResourceLocation m_60589_() {
       if (this.f_60440_ == null) {
-         ResourceLocation resourcelocation = Registry.f_122824_.m_7981_(this.m_7374_());
-         this.f_60440_ = new ResourceLocation(resourcelocation.m_135827_(), "blocks/" + resourcelocation.m_135815_());
+         this.f_60440_ = this.lootTableSupplier.get();
       }
 
       return this.f_60440_;
@@ -408,6 +_,14 @@
       return this.f_60439_.f_60888_;
    }
 
+   protected boolean isAir(BlockState state) {
+      return ((BlockStateBase)state).f_60596_;
+   }
+
+   /* ======================================== FORGE START ===================================== */
+   private final java.util.function.Supplier<ResourceLocation> lootTableSupplier;
+   /* ========================================= FORGE END ====================================== */
+
    public abstract static class BlockStateBase extends StateHolder<Block, BlockState> {
       private final int f_60594_;
       private final boolean f_60595_;
@@ -443,13 +_,121 @@
          this.f_60605_ = blockbehaviour$properties.f_60901_;
          this.f_60606_ = blockbehaviour$properties.f_60902_;
          this.f_222962_ = blockbehaviour$properties.f_222978_.apply(this.m_7160_());
-      }
+         this.conditionallyFullOpaque = this.f_60601_ & this.f_60595_; // Paper
+         // Paper start - optimise collisions
+         this.id1 = it.unimi.dsi.fastutil.HashCommon.murmurHash3(it.unimi.dsi.fastutil.HashCommon.murmurHash3(ID_GENERATOR.getAndIncrement() + RANDOM_OFFSET) + RANDOM_OFFSET);
+         this.id2 = it.unimi.dsi.fastutil.HashCommon.murmurHash3(it.unimi.dsi.fastutil.HashCommon.murmurHash3(ID_GENERATOR.getAndIncrement() + RANDOM_OFFSET) + RANDOM_OFFSET);
+         // Paper end - optimise collisions
+
+      }
+
+      // Paper start
+      protected boolean shapeExceedsCube = true;
+      public final boolean shapeExceedsCube() {
+         return this.shapeExceedsCube;
+      }
+      // Paper end
+      // Paper start - starlight
+      protected int opacityIfCached = -1;
+      // ret -1 if opacity is dynamic, or -1 if the block is conditionally full opaque, else return opacity in [0, 15]
+      public final int getOpacityIfCached() {
+         return this.opacityIfCached;
+      }
+
+      protected final boolean conditionallyFullOpaque;
+      public final boolean isConditionallyFullOpaque() {
+         return this.conditionallyFullOpaque;
+      }
+      // Paper end - starlight
+      // Paper start - optimise collisions
+      private static final int RANDOM_OFFSET = 704237939;
+      private static final Direction[] DIRECTIONS_CACHED = Direction.values();
+      private static final java.util.concurrent.atomic.AtomicInteger ID_GENERATOR = new java.util.concurrent.atomic.AtomicInteger();
+      private final int id1, id2;
+      private boolean occludesFullBlock;
+      private boolean emptyCollisionShape;
+      private VoxelShape constantCollisionShape;
+      private AABB constantAABBCollision;
+      private static void initCaches(final VoxelShape shape) {
+         shape.isFullBlock();
+         shape.occludesFullBlock();
+         shape.m_83299_();
+         if (!shape.m_83281_()) {
+            shape.m_83215_();
+         }
+      }
+
+      public final boolean hasCache() {
+         return this.f_60593_ != null;
+      }
+
+      public final boolean occludesFullBlock() {
+         return this.occludesFullBlock;
+      }
+
+      public final boolean emptyCollisionShape() {
+         return this.emptyCollisionShape;
+      }
+
+      public final int uniqueId1() {
+         return this.id1;
+      }
+
+      public final int uniqueId2() {
+         return this.id2;
+      }
+
+      public final VoxelShape getConstantCollisionShape() {
+         return this.constantCollisionShape;
+      }
+
+      public final AABB getConstantCollisionAABB() {
+         return this.constantAABBCollision;
+      }
+      // Paper end - optimise collisions
 
       public void m_60611_() {
          if (!this.m_60734_().m_49967_()) {
             this.f_60593_ = new BlockBehaviour.BlockStateBase.Cache(this.m_7160_());
          }
 
+         this.shapeExceedsCube = this.f_60593_ == null || this.f_60593_.f_60843_; // Paper - moved from actual method to here
+         this.opacityIfCached = this.f_60593_ == null || this.isConditionallyFullOpaque() ? -1 : this.f_60593_.f_60848_; // Paper - starlight - cache opacity for light
+
+         // Paper start - optimise collisions
+         if (this.f_60593_ != null) {
+            final VoxelShape collisionShape = this.f_60593_.f_60842_;
+            try {
+               this.constantCollisionShape = this.m_60742_(null, null, null);
+               this.constantAABBCollision = this.constantCollisionShape == null ? null : this.constantCollisionShape.getSingleAABBRepresentation();
+            } catch (final Throwable throwable) {
+               this.constantCollisionShape = null;
+               this.constantAABBCollision = null;
+            }
+            this.occludesFullBlock = collisionShape.occludesFullBlock();
+            this.emptyCollisionShape = collisionShape.m_83281_();
+            // init caches
+            initCaches(collisionShape);
+            if (collisionShape != Shapes.m_83040_() && collisionShape != Shapes.m_83144_()) {
+               for (final Direction direction : DIRECTIONS_CACHED) {
+                  // initialise the directional face shape cache as well
+                  final VoxelShape shape = Shapes.m_83121_(collisionShape, direction);
+                  initCaches(shape);
+               }
+            }
+            if (this.f_60593_.f_60849_ != null) {
+               for (final VoxelShape shape : this.f_60593_.f_60849_) {
+                  initCaches(shape);
+               }
+            }
+         } else {
+            this.occludesFullBlock = false;
+            this.emptyCollisionShape = false;
+            this.constantCollisionShape = null;
+            this.constantAABBCollision = null;
+         }
+         // Paper end - optimise collisions
+
       }
 
       public Block m_60734_() {
@@ -485,25 +_,29 @@
       }
 
       public boolean m_60779_() {
-         return this.f_60593_ == null || this.f_60593_.f_60843_;
+         return this.shapeExceedsCube; // Paper - moved into shape cache init
       }
 
-      public boolean m_60787_() {
+      public final boolean m_60787_() { // Paper
          return this.f_60595_;
       }
 
-      public int m_60791_() {
+      /** @deprecated use {@link BlockState#getLightEmission(BlockGetter, BlockPos)} */
+      @Deprecated
+      public final int m_60791_() { // Paper
          return this.f_60594_;
       }
 
-      public boolean m_60795_() {
-         return this.f_60596_;
+      public final boolean m_60795_() { // Paper
+         return this.m_60734_().isAir((BlockState)this);
       }
 
       public MaterialColor m_60780_(BlockGetter p_60781_, BlockPos p_60782_) {
-         return this.f_60598_;
+         return m_60734_().getMapColor(this.m_7160_(), p_60781_, p_60782_, this.f_60598_);
       }
 
+      /** @deprecated use {@link BlockState#rotate(LevelAccessor, BlockPos, Rotation)} */
+      @Deprecated
       public BlockState m_60717_(Rotation p_60718_) {
          return this.m_60734_().m_6843_(this.m_7160_(), p_60718_);
       }
@@ -569,7 +_,7 @@
          }
       }
 
-      public boolean m_60815_() {
+      public final boolean m_60815_() { // Paper
          return this.f_60601_;
       }
 
@@ -762,11 +_,11 @@
          return this.m_60734_() == p_60714_;
       }
 
-      public FluidState m_60819_() {
+      public FluidState m_60819_() {  // Paper
          return this.m_60734_().m_5888_(this.m_7160_());
       }
 
-      public boolean m_60823_() {
+      public final boolean m_60823_() { // Paper
          return this.m_60734_().m_6724_(this.m_7160_());
       }
 
@@ -886,8 +_,9 @@
       ResourceLocation f_60894_;
       boolean f_60895_ = true;
       boolean f_60896_;
+      private java.util.function.Supplier<ResourceLocation> lootTableSupplier;
       BlockBehaviour.StateArgumentPredicate<EntityType<?>> f_60897_ = (p_60935_, p_60936_, p_60937_, p_60938_) -> {
-         return p_60935_.m_60783_(p_60936_, p_60937_, Direction.UP) && p_60935_.m_60791_() < 14;
+         return p_60935_.m_60783_(p_60936_, p_60937_, Direction.UP) && p_60935_.getLightEmission(p_60936_, p_60937_) < 14;
       };
       BlockBehaviour.StatePredicate f_60898_ = (p_60985_, p_60986_, p_60987_) -> {
          return p_60985_.m_60767_().m_76337_() && p_60985_.m_60838_(p_60986_, p_60987_);
@@ -1018,9 +_,15 @@
          return this;
       }
 
+      @Deprecated // FORGE: Use the variant that takes a Supplier below
       public BlockBehaviour.Properties m_60916_(Block p_60917_) {
-         this.f_60894_ = p_60917_.m_60589_();
+         this.lootTableSupplier = () -> net.minecraftforge.registries.ForgeRegistries.BLOCKS.getDelegateOrThrow(p_60917_).get().m_60589_();
          return this;
+      }
+
+      public BlockBehaviour.Properties lootFrom(java.util.function.Supplier<? extends Block> blockIn) {
+          this.lootTableSupplier = () -> blockIn.get().m_60589_();
+          return this;
       }
 
       public BlockBehaviour.Properties m_60996_() {
