--- a/net/minecraft/client/renderer/chunk/RenderChunkRegion.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunkRegion.java
@@ -42,6 +_,11 @@
     }
 
     @Override
+    public @org.jetbrains.annotations.Nullable FluidState getFluidIfLoaded(BlockPos blockposition) {
+        return null;
+    }
+
+    @Override
     public FluidState getFluidState(BlockPos p_112943_) {
         return this.getChunk(SectionPos.blockToSectionCoord(p_112943_.getX()), SectionPos.blockToSectionCoord(p_112943_.getZ()))
             .getBlockState(p_112943_)
