--- a/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
@@ -34,7 +_,11 @@
    }
 
    protected void m_6725_(ServerLevel p_24700_, Villager p_24701_, long p_24702_) {
-      if (p_24702_ % 100L == 0L) {
+      // Pufferfish start
+      if (p_24701_.nextGolemPanic < 0) p_24701_.nextGolemPanic = p_24702_ + 100;
+      if (--p_24701_.nextGolemPanic < p_24702_) {
+         p_24701_.nextGolemPanic = -1;
+         // Pufferfish end
          p_24701_.m_35397_(p_24700_, p_24702_, 3);
       }
 
