--- a/net/minecraft/client/renderer/chunk/RenderChunkRegion.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunkRegion.java
@@ -34,6 +_,12 @@
       return this.f_112905_[i][j].m_200453_(p_112947_);
    }
 
+   @org.jetbrains.annotations.Nullable
+   @Override
+   public BlockState getBlockStateIfLoaded(BlockPos blockposition) {
+      return null;
+   }
+
    public FluidState m_6425_(BlockPos p_112943_) {
       int i = SectionPos.m_123171_(p_112943_.m_123341_()) - this.f_112899_;
       int j = SectionPos.m_123171_(p_112943_.m_123343_()) - this.f_112900_;
@@ -65,5 +_,15 @@
 
    public int m_141928_() {
       return this.f_112908_.m_141928_();
+   }
+
+   @Override
+   public float getShade(float normalX, float normalY, float normalZ, boolean shade) {
+      return this.f_112908_.getShade(normalX, normalY, normalZ, shade);
+   }
+
+   @Override
+   public net.minecraftforge.client.model.data.ModelDataManager getModelDataManager() {
+      return f_112908_.getModelDataManager();
    }
 }
