--- a/net/minecraft/world/level/chunk/storage/ChunkSerializer.java
+++ b/net/minecraft/world/level/chunk/storage/ChunkSerializer.java
@@ -7,10 +_,8 @@
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
 import it.unimi.dsi.fastutil.longs.LongSet;
 import it.unimi.dsi.fastutil.shorts.ShortList;
-import java.util.Arrays;
-import java.util.EnumSet;
-import java.util.Map;
-import java.util.Objects;
+
+import java.util.*;
 import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -68,7 +_,7 @@
 
 public class ChunkSerializer {
     private static final Codec<PalettedContainer<BlockState>> BLOCK_STATE_CODEC = PalettedContainer.codecRW(
-        Block.BLOCK_STATE_REGISTRY, BlockState.CODEC, PalettedContainer.Strategy.SECTION_STATES, Blocks.AIR.defaultBlockState()
+            Block.BLOCK_STATE_REGISTRY, BlockState.CODEC, PalettedContainer.Strategy.SECTION_STATES, Blocks.AIR.defaultBlockState()
     );
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final String TAG_UPGRADE_DATA = "UpgradeData";
@@ -82,6 +_,30 @@
     public static final String BLOCK_LIGHT_TAG = "BlockLight";
     public static final String SKY_LIGHT_TAG = "SkyLight";
 
+    // Paper start - guard against serializing mismatching coordinates
+    // TODO Note: This needs to be re-checked each update
+    public static ChunkPos getChunkCoordinate(final CompoundTag chunkData) {
+        final int dataVersion = ChunkStorage.getVersion(chunkData);
+        if (dataVersion < 2842) { // Level tag is removed after this version
+            final CompoundTag levelData = chunkData.getCompound("Level");
+            return new ChunkPos(levelData.getInt("xPos"), levelData.getInt("zPos"));
+        } else {
+            return new ChunkPos(chunkData.getInt("xPos"), chunkData.getInt("zPos"));
+        }
+    }
+    // Paper end - guard against serializing mismatching coordinates
+    // TODO: Check on update
+    public static long getLastWorldSaveTime(final CompoundTag chunkData) {
+        final int dataVersion = ChunkStorage.getVersion(chunkData);
+        if (dataVersion < 2842) { // Level tag is removed after this version
+            final CompoundTag levelData = chunkData.getCompound("Level");
+            return levelData.getLong("LastUpdate");
+        } else {
+            return chunkData.getLong("LastUpdate");
+        }
+    }
+    // Paper end - Attempt to recalculate regionfile header if it is corrupt
+
     public static ProtoChunk read(ServerLevel p_188231_, PoiManager p_188232_, RegionStorageInfo p_352213_, ChunkPos p_188233_, CompoundTag p_188234_) {
         ChunkPos chunkpos = new ChunkPos(p_188234_.getInt("xPos"), p_188234_.getInt("zPos"));
         if (!Objects.equals(p_188233_, chunkpos)) {
@@ -109,29 +_,31 @@
                 PalettedContainer<BlockState> palettedcontainer;
                 if (compoundtag.contains("block_states", 10)) {
                     palettedcontainer = BLOCK_STATE_CODEC.parse(NbtOps.INSTANCE, compoundtag.getCompound("block_states"))
-                        .promotePartial(p_188283_ -> logErrors(p_188233_, k, p_188283_))
-                        .getOrThrow(ChunkSerializer.ChunkReadException::new);
+                            .promotePartial(p_188283_ -> logErrors(p_188233_, k, p_188283_))
+                            .getOrThrow(ChunkSerializer.ChunkReadException::new);
                 } else {
                     palettedcontainer = new PalettedContainer<>(
-                        Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState(), PalettedContainer.Strategy.SECTION_STATES
+                            Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState(), PalettedContainer.Strategy.SECTION_STATES
                     );
                 }
 
                 PalettedContainerRO<Holder<Biome>> palettedcontainerro;
                 if (compoundtag.contains("biomes", 10)) {
                     palettedcontainerro = codec.parse(NbtOps.INSTANCE, compoundtag.getCompound("biomes"))
-                        .promotePartial(p_188274_ -> logErrors(p_188233_, k, p_188274_))
-                        .getOrThrow(ChunkSerializer.ChunkReadException::new);
+                            .promotePartial(p_188274_ -> logErrors(p_188233_, k, p_188274_))
+                            .getOrThrow(ChunkSerializer.ChunkReadException::new);
                 } else {
                     palettedcontainerro = new PalettedContainer<>(
-                        registry.asHolderIdMap(), registry.getHolderOrThrow(Biomes.PLAINS), PalettedContainer.Strategy.SECTION_BIOMES
+                            registry.asHolderIdMap(), registry.getHolderOrThrow(Biomes.PLAINS), PalettedContainer.Strategy.SECTION_BIOMES
                     );
                 }
 
                 LevelChunkSection levelchunksection = new LevelChunkSection(palettedcontainer, palettedcontainerro);
                 alevelchunksection[l] = levelchunksection;
                 SectionPos sectionpos = SectionPos.of(p_188233_, k);
-                p_188232_.checkConsistencyWithBlocks(sectionpos, levelchunksection);
+
+                // Paper - rewrite chunk system - moved to final load stage
+                //p_188232_.checkConsistencyWithBlocks(sectionpos, levelchunksection);
             }
 
             boolean flag3 = compoundtag.contains("BlockLight", 7);
@@ -157,9 +_,9 @@
         BlendingData blendingdata;
         if (p_188234_.contains("blending_data", 10)) {
             blendingdata = BlendingData.CODEC
-                .parse(new Dynamic<>(NbtOps.INSTANCE, p_188234_.getCompound("blending_data")))
-                .resultOrPartial(LOGGER::error)
-                .orElse(null);
+                    .parse(new Dynamic<>(NbtOps.INSTANCE, p_188234_.getCompound("blending_data")))
+                    .resultOrPartial(LOGGER::error)
+                    .orElse(null);
         } else {
             blendingdata = null;
         }
@@ -167,41 +_,41 @@
         ChunkAccess chunkaccess;
         if (chunktype == ChunkType.LEVELCHUNK) {
             LevelChunkTicks<Block> levelchunkticks = LevelChunkTicks.load(
-                p_188234_.getList("block_ticks", 10), p_258988_ -> BuiltInRegistries.BLOCK.getOptional(ResourceLocation.tryParse(p_258988_)), p_188233_
+                    p_188234_.getList("block_ticks", 10), p_258988_ -> BuiltInRegistries.BLOCK.getOptional(ResourceLocation.tryParse(p_258988_)), p_188233_
             );
             LevelChunkTicks<Fluid> levelchunkticks1 = LevelChunkTicks.load(
-                p_188234_.getList("fluid_ticks", 10), p_258990_ -> BuiltInRegistries.FLUID.getOptional(ResourceLocation.tryParse(p_258990_)), p_188233_
+                    p_188234_.getList("fluid_ticks", 10), p_258990_ -> BuiltInRegistries.FLUID.getOptional(ResourceLocation.tryParse(p_258990_)), p_188233_
             );
             chunkaccess = new LevelChunk(
-                p_188231_.getLevel(),
-                p_188233_,
-                upgradedata,
-                levelchunkticks,
-                levelchunkticks1,
-                k1,
-                alevelchunksection,
-                postLoadChunk(p_188231_, p_188234_),
-                blendingdata
+                    p_188231_.getLevel(),
+                    p_188233_,
+                    upgradedata,
+                    levelchunkticks,
+                    levelchunkticks1,
+                    k1,
+                    alevelchunksection,
+                    postLoadChunk(p_188231_, p_188234_),
+                    blendingdata
             );
             if (p_188234_.contains(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY, net.minecraft.nbt.Tag.TAG_LIST))
                 Objects.requireNonNull(((LevelChunk)chunkaccess).getAuxLightManager(p_188233_)).deserializeNBT(p_188231_.registryAccess(), p_188234_.getList(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY, net.minecraft.nbt.Tag.TAG_COMPOUND));
         } else {
             ProtoChunkTicks<Block> protochunkticks = ProtoChunkTicks.load(
-                p_188234_.getList("block_ticks", 10), p_258992_ -> BuiltInRegistries.BLOCK.getOptional(ResourceLocation.tryParse(p_258992_)), p_188233_
+                    p_188234_.getList("block_ticks", 10), p_258992_ -> BuiltInRegistries.BLOCK.getOptional(ResourceLocation.tryParse(p_258992_)), p_188233_
             );
             ProtoChunkTicks<Fluid> protochunkticks1 = ProtoChunkTicks.load(
-                p_188234_.getList("fluid_ticks", 10), p_258991_ -> BuiltInRegistries.FLUID.getOptional(ResourceLocation.tryParse(p_258991_)), p_188233_
+                    p_188234_.getList("fluid_ticks", 10), p_258991_ -> BuiltInRegistries.FLUID.getOptional(ResourceLocation.tryParse(p_258991_)), p_188233_
             );
             ProtoChunk protochunk = new ProtoChunk(
-                p_188233_, upgradedata, alevelchunksection, protochunkticks, protochunkticks1, p_188231_, registry, blendingdata
+                    p_188233_, upgradedata, alevelchunksection, protochunkticks, protochunkticks1, p_188231_, registry, blendingdata
             );
             chunkaccess = protochunk;
             protochunk.setInhabitedTime(k1);
             if (p_188234_.contains("below_zero_retrogen", 10)) {
                 BelowZeroRetrogen.CODEC
-                    .parse(new Dynamic<>(NbtOps.INSTANCE, p_188234_.getCompound("below_zero_retrogen")))
-                    .resultOrPartial(LOGGER::error)
-                    .ifPresent(protochunk::setBelowZeroRetrogen);
+                        .parse(new Dynamic<>(NbtOps.INSTANCE, p_188234_.getCompound("below_zero_retrogen")))
+                        .resultOrPartial(LOGGER::error)
+                        .ifPresent(protochunk::setBelowZeroRetrogen);
             }
 
             ChunkStatus chunkstatus = ChunkStatus.byName(p_188234_.getString("Status"));
@@ -244,6 +_,7 @@
             }
         }
 
+        ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.loadLightHook(p_188231_, p_188233_, p_188234_, (ChunkAccess)chunkaccess); // Paper - rewrite chunk system - note: it's ok to pass the raw value instead of wrapped
         if (chunktype == ChunkType.LEVELCHUNK) {
             net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkDataEvent.Load(chunkaccess, p_188234_, chunktype));
             return new ImposterProtoChunk((LevelChunk)chunkaccess, false);
@@ -280,9 +_,34 @@
 
     private static Codec<PalettedContainerRO<Holder<Biome>>> makeBiomeCodec(Registry<Biome> p_188261_) {
         return PalettedContainer.codecRO(
-            p_188261_.asHolderIdMap(), p_188261_.holderByNameCodec(), PalettedContainer.Strategy.SECTION_BIOMES, p_188261_.getHolderOrThrow(Biomes.PLAINS)
-        );
-    }
+                p_188261_.asHolderIdMap(), p_188261_.holderByNameCodec(), PalettedContainer.Strategy.SECTION_BIOMES, p_188261_.getHolderOrThrow(Biomes.PLAINS)
+        );
+    }
+
+    // Paper start - async chunk saving
+    // must be called sync
+    public static ca.spottedleaf.moonrise.patches.chunk_system.async_save.AsyncChunkSaveData getAsyncSaveData(ServerLevel world, ChunkAccess chunk) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(world, chunk.locX, chunk.locZ, "Preparing async chunk save data");
+
+        final CompoundTag tickLists = new CompoundTag();
+        ChunkSerializer.saveTicks(world, tickLists, chunk.getTicksForSerialization());
+
+        ListTag blockEntitiesSerialized = new ListTag();
+        for (final BlockPos blockPos : chunk.getBlockEntitiesPos()) {
+            final CompoundTag blockEntityNbt = chunk.getBlockEntityNbtForSaving(blockPos, world.registryAccess());
+            if (blockEntityNbt != null) {
+                blockEntitiesSerialized.add(blockEntityNbt);
+            }
+        }
+
+        return new ca.spottedleaf.moonrise.patches.chunk_system.async_save.AsyncChunkSaveData(
+                tickLists.get(BLOCK_TICKS_TAG),
+                tickLists.get(FLUID_TICKS_TAG),
+                blockEntitiesSerialized,
+                world.getGameTime()
+        );
+    }
+    // Paper end - async chunk saving
 
     public static CompoundTag write(ServerLevel p_63455_, ChunkAccess p_63456_) {
         ChunkPos chunkpos = p_63456_.getPos();
@@ -290,23 +_,24 @@
         compoundtag.putInt("xPos", chunkpos.x);
         compoundtag.putInt("yPos", p_63456_.getMinSection());
         compoundtag.putInt("zPos", chunkpos.z);
-        compoundtag.putLong("LastUpdate", p_63455_.getGameTime());
+        ca.spottedleaf.moonrise.patches.chunk_system.async_save.AsyncChunkSaveData asyncsavedata = p_63456_.asyncsavedata;
+        compoundtag.putLong("LastUpdate", asyncsavedata != null ? asyncsavedata.worldTime() : p_63455_.getGameTime()); // Paper - async chunk saving // Paper - diff on change
         compoundtag.putLong("InhabitedTime", p_63456_.getInhabitedTime());
         compoundtag.putString("Status", BuiltInRegistries.CHUNK_STATUS.getKey(p_63456_.getPersistedStatus()).toString());
         BlendingData blendingdata = p_63456_.getBlendingData();
         if (blendingdata != null) {
             BlendingData.CODEC
-                .encodeStart(NbtOps.INSTANCE, blendingdata)
-                .resultOrPartial(LOGGER::error)
-                .ifPresent(p_196909_ -> compoundtag.put("blending_data", p_196909_));
+                    .encodeStart(NbtOps.INSTANCE, blendingdata)
+                    .resultOrPartial(LOGGER::error)
+                    .ifPresent(p_196909_ -> compoundtag.put("blending_data", p_196909_));
         }
 
         BelowZeroRetrogen belowzeroretrogen = p_63456_.getBelowZeroRetrogen();
         if (belowzeroretrogen != null) {
             BelowZeroRetrogen.CODEC
-                .encodeStart(NbtOps.INSTANCE, belowzeroretrogen)
-                .resultOrPartial(LOGGER::error)
-                .ifPresent(p_188279_ -> compoundtag.put("below_zero_retrogen", p_188279_));
+                    .encodeStart(NbtOps.INSTANCE, belowzeroretrogen)
+                    .resultOrPartial(LOGGER::error)
+                    .ifPresent(p_188279_ -> compoundtag.put("below_zero_retrogen", p_188279_));
         }
 
         UpgradeData upgradedata = p_63456_.getUpgradeData();
@@ -354,16 +_,30 @@
             compoundtag.putBoolean("isLightOn", true);
         }
 
-        ListTag listtag1 = new ListTag();
-
-        for (BlockPos blockpos : p_63456_.getBlockEntitiesPos()) {
-            CompoundTag compoundtag3 = p_63456_.getBlockEntityNbtForSaving(blockpos, p_63455_.registryAccess());
-            if (compoundtag3 != null) {
-                listtag1.add(compoundtag3);
+        // Paper start - async chunk saving
+        ListTag nbttaglist1;
+        Iterator<BlockPos> iterator;
+        if (asyncsavedata != null) {
+            nbttaglist1 = asyncsavedata.blockEntities();
+            iterator = java.util.Collections.emptyIterator();
+        } else {
+            nbttaglist1 = new ListTag();
+            iterator = p_63456_.getBlockEntitiesPos().iterator();
+        }
+        // Paper end - async chunk saving
+
+        CompoundTag nbttagcompound2;
+
+        while (iterator.hasNext()) {
+            BlockPos blockposition = (BlockPos) iterator.next();
+
+            nbttagcompound2 = p_63456_.getBlockEntityNbtForSaving(blockposition, p_63455_.registryAccess());
+            if (nbttagcompound2 != null) {
+                nbttaglist1.add(nbttagcompound2);
             }
         }
 
-        compoundtag.put("block_entities", listtag1);
+        compoundtag.put("block_entities", nbttaglist1);
         if (p_63456_.getPersistedStatus().getChunkType() == ChunkType.PROTOCHUNK) {
             ProtoChunk protochunk = (ProtoChunk)p_63456_;
             ListTag listtag2 = new ListTag();
@@ -382,11 +_,18 @@
         }
         else if (p_63456_ instanceof LevelChunk levelChunk){
 
-             Tag lightTag = levelChunk.getAuxLightManager(chunkpos).serializeNBT(p_63455_.registryAccess());
-             if (lightTag != null) compoundtag.put(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY, lightTag);
+            Tag lightTag = levelChunk.getAuxLightManager(chunkpos).serializeNBT(p_63455_.registryAccess());
+            if (lightTag != null) compoundtag.put(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY, lightTag);
         }
 
-        saveTicks(p_63455_, compoundtag, p_63456_.getTicksForSerialization());
+        // Paper start
+        if (asyncsavedata != null) {
+            compoundtag.put(BLOCK_TICKS_TAG, asyncsavedata.blockTickList());
+            compoundtag.put(FLUID_TICKS_TAG, asyncsavedata.fluidTickList());
+        } else {
+            saveTicks(p_63455_, compoundtag, p_63456_.getTicksForSerialization());
+        }
+        // Paper end
         compoundtag.put("PostProcessing", packOffsets(p_63456_.getPostProcessing()));
         CompoundTag compoundtag2 = new CompoundTag();
 
@@ -404,9 +_,10 @@
         }
         compoundtag.put("Heightmaps", compoundtag2);
         compoundtag.put(
-            "structures",
-            packStructureData(StructurePieceSerializationContext.fromLevel(p_63455_), chunkpos, p_63456_.getAllStarts(), p_63456_.getAllReferences())
+                "structures",
+                packStructureData(StructurePieceSerializationContext.fromLevel(p_63455_), chunkpos, p_63456_.getAllStarts(), p_63456_.getAllReferences())
         );
+        ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.saveLightHook(p_63455_, p_63456_, compoundtag); // Paper - rewrite chunk system
         return compoundtag;
     }
 
@@ -426,7 +_,7 @@
         ListTag listtag1 = getListOfCompoundsOrNull(p_196892_, "block_entities");
         return listtag == null && listtag1 == null ? null : p_196904_ -> {
             if (listtag != null) {
-                p_196891_.addLegacyChunkEntities(EntityType.loadEntitiesRecursive(listtag, p_196891_));
+                p_196891_.addLegacyChunkEntities(EntityType.loadEntitiesRecursive(listtag, p_196891_), p_196904_.getPos()); // Paper - rewrite chunk system
             }
 
             if (listtag1 != null) {
@@ -454,7 +_,7 @@
     }
 
     private static CompoundTag packStructureData(
-        StructurePieceSerializationContext p_188250_, ChunkPos p_188251_, Map<Structure, StructureStart> p_188252_, Map<Structure, LongSet> p_188253_
+            StructurePieceSerializationContext p_188250_, ChunkPos p_188251_, Map<Structure, StructureStart> p_188252_, Map<Structure, LongSet> p_188253_
     ) {
         CompoundTag compoundtag = new CompoundTag();
         CompoundTag compoundtag1 = new CompoundTag();
