--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -70,12 +_,12 @@
         }
     }
 
-    private static ChunkPos readChunkPos(CompoundTag p_156571_) {
+    public  static ChunkPos readChunkPos(CompoundTag p_156571_) { // GoldenForge - public
         int[] aint = p_156571_.getIntArray("Position");
         return new ChunkPos(aint[0], aint[1]);
     }
 
-    private static void writeChunkPos(CompoundTag p_156563_, ChunkPos p_156564_) {
+    public static void writeChunkPos(CompoundTag p_156563_, ChunkPos p_156564_) {
         p_156563_.put("Position", new IntArrayTag(new int[]{p_156564_.x, p_156564_.z}));
     }
 
