--- a/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
@@ -121,6 +_,7 @@
     }
 
     public void emptyAllLivingFromHive(@Nullable Player p_58749_, BlockState p_58750_, BeehiveBlockEntity.BeeReleaseStatus p_58751_) {
+        if (this.level == null) return; // Paper - do not fire from worldgen thread
         List<Entity> list = this.releaseAllOccupants(p_58750_, p_58751_);
         if (p_58749_ != null) {
             for (Entity entity : list) {
