--- a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -26,11 +_,17 @@
          return true;
       });
       list.sort(Comparator.comparingDouble(p_26698_::m_20280_));
-      Optional<ItemEntity> optional = list.stream().filter((p_26706_) -> {
-         return p_26698_.m_7243_(p_26706_.m_32055_());
-      }).filter((p_26701_) -> {
-         return p_26701_.m_19950_(p_26698_, 32.0D);
-      }).filter(p_26698_::m_142582_).findFirst();
-      brain.m_21886_(MemoryModuleType.f_26332_, optional);
+      // Paper start - remove streams in favour of lists
+      ItemEntity nearest = null;
+      for (int i = 0; i < list.size(); i++) {
+         ItemEntity entityItem = list.get(i);
+         if (p_26698_.m_142582_(entityItem)) {
+            // Paper end - remove streams
+            nearest = entityItem;
+            break;
+         }
+      }
+      brain.m_21886_(MemoryModuleType.f_26332_, Optional.ofNullable(nearest));
+      // Paper end
    }
 }
