--- a/net/minecraft/world/level/gameevent/DynamicGameEventListener.java
+++ b/net/minecraft/world/level/gameevent/DynamicGameEventListener.java
@@ -41,7 +_,7 @@
 
     private static void ifChunkExists(LevelReader p_223623_, @Nullable SectionPos p_223624_, Consumer<GameEventListenerRegistry> p_223625_) {
         if (p_223624_ != null) {
-            ChunkAccess chunkaccess = p_223623_.getChunk(p_223624_.x(), p_223624_.z(), ChunkStatus.FULL, false);
+            ChunkAccess chunkaccess = p_223623_.getChunkIfLoadedImmediately(p_223624_.getX(), p_223624_.getZ()); // Paper - Perf: can cause sync loads while completing a chunk, resulting in deadlock
             if (chunkaccess != null) {
                 p_223625_.accept(chunkaccess.getListenerRegistry(p_223624_.y()));
             }
