--- a/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
@@ -36,7 +_,7 @@
     private final EntityPersistentStorage<T> permanentStorage;
     private final EntityLookup<T> visibleEntityStorage;
     final EntitySectionStorage<T> sectionStorage;
-    private final LevelEntityGetter<T> entityGetter;
+    private LevelEntityGetter<T> entityGetter;
     private final Long2ObjectMap<Visibility> chunkVisibility = new Long2ObjectOpenHashMap<>();
     private final Long2ObjectMap<PersistentEntitySectionManager.ChunkLoadStatus> chunkLoadStatuses = new Long2ObjectOpenHashMap<>();
     private final LongSet chunksToUnload = new LongOpenHashSet();
@@ -50,6 +_,12 @@
         this.callbacks = p_157504_;
         this.permanentStorage = p_157505_;
         this.entityGetter = new LevelEntityGetterAdapter<>(this.visibleEntityStorage, this.sectionStorage);
+    }
+
+
+    // Goldenforge: in case mod use it.. can't be in the constructor since moonlight inject here
+    public void setEntityGetter(ca.spottedleaf.moonrise.patches.chunk_system.level.entity.EntityLookup entityLookup) {
+        this.entityGetter = (LevelEntityGetter<T>) entityLookup;
     }
 
     void removeSectionIfEmpty(long p_157510_, EntitySection<T> p_157511_) {
