--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -121,6 +_,8 @@
                int j1 = (j - 1) / -2;
                blockpos$mutableblockpos.m_122190_(blockpos).m_122175_(direction, k + i1).m_122175_(Direction.UP, l + j1);
                BlockState blockstate = this.f_19853_.m_8055_(blockpos$mutableblockpos);
+               if (net.minecraft.world.level.block.Block.m_49863_(this.f_19853_, blockpos$mutableblockpos, this.f_31699_))
+                  continue;
                if (!blockstate.m_60767_().m_76333_() && !DiodeBlock.m_52586_(blockstate)) {
                   return false;
                }
@@ -187,7 +_,7 @@
    public void m_7378_(CompoundTag p_31730_) {
       BlockPos blockpos = new BlockPos(p_31730_.m_128451_("TileX"), p_31730_.m_128451_("TileY"), p_31730_.m_128451_("TileZ"));
       if (!blockpos.m_123314_(this.m_20183_(), 16.0D)) {
-         f_238173_.error("Hanging entity at invalid position: {}", (Object)blockpos);
+         f_238173_.error("Hanging entity at invalid position: {}. Entity  blockpos: {}", (Object)blockpos, this.m_20183_());
       } else {
          this.f_31698_ = blockpos;
       }
