--- a/net/minecraft/world/level/levelgen/Beardifier.java
+++ b/net/minecraft/world/level/levelgen/Beardifier.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level.levelgen;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Iterators;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectList;
 import it.unimi.dsi.fastutil.objects.ObjectListIterator;
@@ -16,6 +_,7 @@
 import net.minecraft.world.level.levelgen.structure.TerrainAdjustment;
 import net.minecraft.world.level.levelgen.structure.pools.JigsawJunction;
 import net.minecraft.world.level.levelgen.structure.pools.StructureTemplatePool;
+import org.spongepowered.asm.mixin.Unique;
 
 public class Beardifier implements DensityFunctions.BeardifierOrMarker {
     public static final int BEARD_KERNEL_RADIUS = 12;
@@ -32,6 +_,17 @@
     protected final ObjectListIterator<Beardifier.Rigid> pieceIterator;
     protected final ObjectListIterator<JigsawJunction> junctionIterator;
 
+    private Beardifier.Rigid[] c2me$pieceArray;
+
+    private JigsawJunction[] c2me$junctionArray;
+
+    private void c2me$initArrays() {
+        this.c2me$pieceArray = Iterators.toArray(this.pieceIterator, Beardifier.Rigid.class);
+        this.pieceIterator.back(Integer.MAX_VALUE);
+        this.c2me$junctionArray = Iterators.toArray(this.junctionIterator, JigsawJunction.class);
+        this.junctionIterator.back(Integer.MAX_VALUE);
+    }
+
     public static Beardifier forStructuresInChunk(StructureManager p_223938_, ChunkPos p_223939_) {
         int i = p_223939_.getMinBlockX();
         int j = p_223939_.getMinBlockZ();
@@ -44,6 +_,11 @@
 
                     for (StructurePiece structurepiece : p_223936_.getPieces()) {
                         if (structurepiece.isCloseToChunk(p_223939_, 12)) {
+                            if (structurepiece instanceof net.neoforged.neoforge.common.world.PieceBeardifierModifier pieceBeardifierModifier) {
+                                if (pieceBeardifierModifier.getTerrainAdjustment() != TerrainAdjustment.NONE) {
+                                    objectlist.add(new Beardifier.Rigid(pieceBeardifierModifier.getBeardifierBox(), pieceBeardifierModifier.getTerrainAdjustment(), pieceBeardifierModifier.getGroundLevelDelta()));
+                                }
+                            } else
                             if (structurepiece instanceof PoolElementStructurePiece) {
                                 PoolElementStructurePiece poolelementstructurepiece = (PoolElementStructurePiece)structurepiece;
                                 StructureTemplatePool.Projection structuretemplatepool$projection = poolelementstructurepiece.getElement().getProjection();
@@ -78,8 +_,59 @@
         this.junctionIterator = p_223918_;
     }
 
-    @Override
-    public double compute(DensityFunction.FunctionContext p_208200_) {
+    public double compute(DensityFunction.FunctionContext pos) {
+        if (this.c2me$pieceArray == null || this.c2me$junctionArray == null) {
+            this.c2me$initArrays();
+        }
+
+        int i = pos.blockX();
+        int j = pos.blockY();
+        int k = pos.blockZ();
+        double d = 0.0;
+
+        for (Beardifier.Rigid piece : this.c2me$pieceArray) {
+            BoundingBox blockBox = piece.box();
+            int l = piece.groundLevelDelta();
+            int m = Math.max(0, Math.max(blockBox.minX() - i, i - blockBox.maxX()));
+            int n = Math.max(0, Math.max(blockBox.minZ() - k, k - blockBox.maxZ()));
+            int o = blockBox.minY() + l;
+            int p = j - o;
+
+//            int q = switch (piece.terrainAdjustment()) {
+//                case NONE -> 0;
+//                case BURY, BEARD_THIN -> p;
+//                case BEARD_BOX -> Math.max(0, Math.max(o - j, j - blockBox.getMaxY()));
+//                case ENCAPSULATE -> Math.max(0, Math.max(blockBox.getMinY() - j, j - blockBox.getMaxY()));
+//            };
+//
+//            d += switch (piece.terrainAdjustment()) {
+//                case NONE -> 0.0;
+//                case BURY -> getMagnitudeWeight(m, (double)q / 2.0, n);
+//                case BEARD_THIN, BEARD_BOX -> getStructureWeight(m, q, n, p) * 0.8;
+//                case ENCAPSULATE -> getMagnitudeWeight((double)m / 2.0, (double)q / 2.0, (double)n / 2.0) * 0.8;
+//            };
+
+            d += switch (piece.terrainAdjustment()) { // 2 switch statement merged
+                case NONE -> 0.0;
+                case BURY -> getBuryContribution(m, (double)p / 2.0, n);
+                case BEARD_THIN -> getBeardContribution(m, p, n, p) * 0.8;
+                case BEARD_BOX -> getBeardContribution(m, Math.max(0, Math.max(o - j, j - blockBox.maxY())), n, p) * 0.8;
+                case ENCAPSULATE -> getBuryContribution((double)m / 2.0, (double)Math.max(0, Math.max(blockBox.minY() - j, j - blockBox.maxY())) / 2.0, (double)n / 2.0) * 0.8;
+            };
+        }
+
+        for (JigsawJunction jigsawJunction : this.c2me$junctionArray) {
+            int r = i - jigsawJunction.getSourceX();
+            int l = j - jigsawJunction.getSourceGroundY();
+            int m = k - jigsawJunction.getSourceZ();
+            d += getBeardContribution(r, l, m, l) * 0.4;
+        }
+
+        return d;
+    }
+
+    //Override
+    public double compute_original(DensityFunction.FunctionContext p_208200_) {
         int i = p_208200_.blockX();
         int j = p_208200_.blockY();
         int k = p_208200_.blockZ();
@@ -133,9 +_,16 @@
         return Double.POSITIVE_INFINITY;
     }
 
-    protected static double getBuryContribution(double p_340947_, double p_340921_, double p_341266_) {
-        double d0 = Mth.length(p_340947_, p_340921_, p_341266_);
-        return Mth.clampedMap(d0, 0.0, 6.0, 1.0, 0.0);
+    private static double getBuryContribution(double x, double y, double z) {
+        // Leaf start - C2ME - Optimize world gen math
+        // Optimize method for beardifier
+        double len = Math.sqrt(x * x + y * y + z * z);
+        if (len > 6.0) {
+            return 0.0;
+        } else {
+            return 1.0 - len / 6.0;
+        }
+        // Leaf end - C2ME - Optimize world gen math
     }
 
     protected static double getBeardContribution(int p_223926_, int p_223927_, int p_223928_, int p_223929_) {
