--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -59,18 +_,26 @@
     private final Holder<NoiseGeneratorSettings> settings;
     private final Supplier<Aquifer.FluidPicker> globalFluidPicker;
 
+    //private int cachedSeaLevel; // Goldenforge: lithium: cache sea level
+
     public NoiseBasedChunkGenerator(BiomeSource p_256415_, Holder<NoiseGeneratorSettings> p_256182_) {
         super(p_256415_);
         this.settings = p_256182_;
         this.globalFluidPicker = Suppliers.memoize(() -> createFluidPicker(p_256182_.value()));
+        //this.cachedSeaLevel = this.settings.value().seaLevel(); // Goldenforge: lithium: cache sea level
     }
 
     private static Aquifer.FluidPicker createFluidPicker(NoiseGeneratorSettings p_249264_) {
         Aquifer.FluidStatus aquifer$fluidstatus = new Aquifer.FluidStatus(-54, Blocks.LAVA.defaultBlockState());
         int i = p_249264_.seaLevel();
         Aquifer.FluidStatus aquifer$fluidstatus1 = new Aquifer.FluidStatus(i, p_249264_.defaultFluid());
-        Aquifer.FluidStatus aquifer$fluidstatus2 = new Aquifer.FluidStatus(DimensionType.MIN_Y * 2, Blocks.AIR.defaultBlockState());
-        return (p_224274_, p_224275_, p_224276_) -> p_224275_ < Math.min(-54, i) ? aquifer$fluidstatus : aquifer$fluidstatus1;
+        // Leaf start - C2ME - Optimize world gen math
+        final int min = Math.min(-54, i);
+        return (x, y, z) -> y < min ? aquifer$fluidstatus : aquifer$fluidstatus1;
+        // Leaf end - C2ME - Optimize world gen math
+
+        //Aquifer.FluidStatus aquifer$fluidstatus2 = new Aquifer.FluidStatus(DimensionType.MIN_Y * 2, Blocks.AIR.defaultBlockState());
+        //return (p_224274_, p_224275_, p_224276_) -> p_224275_ < Math.min(-54, i) ? aquifer$fluidstatus : aquifer$fluidstatus1;
     }
 
     @Override
@@ -78,7 +_,7 @@
         return CompletableFuture.supplyAsync(Util.wrapThreadWithTaskName("init_biomes", () -> {
             this.doCreateBiomes(p_224300_, p_224299_, p_224301_, p_224302_);
             return p_224302_;
-        }), Util.backgroundExecutor());
+        }), Runnable::run); // Paper - rewrite chunk system
     }
 
     private void doCreateBiomes(Blender p_224292_, RandomState p_224293_, StructureManager p_224294_, ChunkAccess p_224295_) {
@@ -347,7 +_,7 @@
             }
 
             return chunkaccess;
-        }), Util.backgroundExecutor());
+        }), Runnable::run); // Paper - rewrite chunk system
     }
 
     private ChunkAccess doFill(Blender p_224285_, StructureManager p_224286_, RandomState p_224287_, ChunkAccess p_224288_, int p_224289_, int p_224290_) {
@@ -437,7 +_,7 @@
 
     @Override
     public int getSeaLevel() {
-        return this.settings.value().seaLevel();
+        return this.settings.value().seaLevel(); // Goldenforge: lithium: cache sea level
     }
 
     @Override
