--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -348,7 +_,17 @@
     }
 
     protected DensityFunction wrap(DensityFunction p_209214_) {
-        return this.wrapped.computeIfAbsent(p_209214_, this::wrapNew);
+        // Leaf start - Reduce worldgen allocations
+        // Avoid lambda allocation
+        DensityFunction func = this.wrapped.get(p_209214_);
+
+        if (func == null) {
+            func = this.wrapNew(p_209214_);
+            this.wrapped.put(p_209214_, func);
+        }
+
+        return func;
+        // Leaf end - Reduce worldgen allocations
     }
 
     private DensityFunction wrapNew(DensityFunction p_209234_) {
