--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -316,8 +_,15 @@
         }
 
         protected void updateY(int p_189577_, int p_189578_, int p_189579_, int p_189580_, int p_189581_, int p_189582_) {
-            this.lastUpdateY++;
-            this.biome = Suppliers.memoize(() -> this.biomeGetter.apply(this.pos.set(p_189580_, p_189581_, p_189582_)));
+            // Leaf start - Reduce worldgen allocations
+            // Reuse supplier object instead of creating new ones every time
+            ++this.lastUpdateY;
+            Supplier<Holder<Biome>> getter = this.biome;
+            if (getter == null) {
+                this.biome = getter = new org.dreeam.leaf.util.biome.PositionalBiomeGetter(this.biomeGetter, this.pos);
+            }
+            ((org.dreeam.leaf.util.biome.PositionalBiomeGetter) getter).update(p_189580_, p_189581_, p_189582_);
+            // Leaf end - Reduce worldgen allocations
             this.blockY = p_189581_;
             this.waterHeight = p_189579_;
             this.stoneDepthBelow = p_189578_;
@@ -574,8 +_,13 @@
         @Nullable
         @Override
         public BlockState tryApply(int p_189694_, int p_189695_, int p_189696_) {
-            for (SurfaceRules.SurfaceRule surfacerules$surfacerule : this.rules) {
-                BlockState blockstate = surfacerules$surfacerule.tryApply(p_189694_, p_189695_, p_189696_);
+            // Leaf start - Reduce worldgen allocations
+            // Avoid iterator allocation
+            int size = this.rules.size();
+            //noinspection ForLoopReplaceableByForEach
+            for (int i = 0; i < size; i++) {
+                BlockState blockstate = this.rules.get(i).tryApply(p_189694_, p_189695_, p_189696_);
+                // Leaf end - Reduce worldgen allocations
                 if (blockstate != null) {
                     return blockstate;
                 }
