--- a/net/minecraft/world/level/chunk/status/ChunkPyramid.java
+++ b/net/minecraft/world/level/chunk/status/ChunkPyramid.java
@@ -54,7 +_,7 @@
         .step(ChunkStatus.CARVERS, p_347448_ -> p_347448_)
         .step(ChunkStatus.FEATURES, p_347446_ -> p_347446_)
         .step(ChunkStatus.INITIALIZE_LIGHT, p_347603_ -> p_347603_.setTask(ChunkStatusTasks::initializeLight))
-        .step(ChunkStatus.LIGHT, p_347463_ -> p_347463_.addRequirement(ChunkStatus.INITIALIZE_LIGHT, 1).setTask(ChunkStatusTasks::light))
+        .step(ChunkStatus.LIGHT, builder -> builder.setTask(ChunkStatusTasks::light)) // Paper - rewrite chunk system - starlight does not need neighbours
         .step(ChunkStatus.SPAWN, p_347516_ -> p_347516_)
         .step(ChunkStatus.FULL, p_347695_ -> p_347695_.setTask(ChunkStatusTasks::full))
         .build();
