--- a/net/minecraft/world/level/CollisionGetter.java
+++ b/net/minecraft/world/level/CollisionGetter.java
@@ -35,6 +_,12 @@
       return this.m_5450_(p_45785_, Shapes.m_83064_(p_45785_.m_20191_()));
    }
 
+   // Paper start - optimise collisions
+   default boolean noCollision(Entity entity, AABB box, boolean loadChunks) {
+      return this.m_45756_(entity, box);
+   }
+   // Paper end - optimise collisions
+
    default boolean m_45772_(AABB p_45773_) {
       return this.m_45756_((Entity)null, p_45773_);
    }
@@ -44,19 +_,21 @@
    }
 
    default boolean m_45756_(@Nullable Entity p_45757_, AABB p_45758_) {
-      for(VoxelShape voxelshape : this.m_186434_(p_45757_, p_45758_)) {
-         if (!voxelshape.m_83281_()) {
-            return false;
+      try { if (p_45757_ != null) p_45757_.collisionLoadChunks = true; // Paper
+         for(VoxelShape voxelShape : this.m_186434_(p_45757_, p_45758_)) {
+            if (!voxelShape.m_83281_()) {
+               return false;
+            }
          }
-      }
+      } finally { if (p_45757_ != null) p_45757_.collisionLoadChunks = false; } // Paper
 
       if (!this.m_183134_(p_45757_, p_45758_).isEmpty()) {
          return false;
       } else if (p_45757_ == null) {
          return true;
       } else {
-         VoxelShape voxelshape1 = this.m_186440_(p_45757_, p_45758_);
-         return voxelshape1 == null || !Shapes.m_83157_(voxelshape1, Shapes.m_83064_(p_45758_), BooleanOp.f_82689_);
+         VoxelShape voxelShape2 = this.m_186440_(p_45757_, p_45758_);
+         return voxelShape2 == null || !Shapes.m_83157_(voxelShape2, Shapes.m_83064_(p_45758_), BooleanOp.f_82689_);
       }
    }
 
@@ -91,6 +_,7 @@
 
       return false;
    }
+
 
    default Optional<Vec3> m_151418_(@Nullable Entity p_151419_, VoxelShape p_151420_, Vec3 p_151421_, double p_151422_, double p_151423_, double p_151424_) {
       if (p_151420_.m_83281_()) {
