--- a/net/minecraft/world/level/LevelReader.java
+++ b/net/minecraft/world/level/LevelReader.java
@@ -26,6 +_,16 @@
    @Deprecated
    boolean m_7232_(int p_46838_, int p_46839_);
 
+
+   // Paper start - rewrite chunk system
+   default ChunkAccess syncLoadNonFull(int chunkX, int chunkZ, ChunkStatus status) {
+      if (status == null || status.m_62427_(ChunkStatus.f_62326_)) {
+         throw new IllegalArgumentException("Status: " + status.toString());
+      }
+      return this.m_6522_(chunkX, chunkZ, status, true);
+   }
+   // Paper end - rewrite chunk system
+
    int m_6924_(Heightmap.Types p_46827_, int p_46828_, int p_46829_);
 
    int m_7445_();
@@ -182,6 +_,10 @@
    }
 
    /** @deprecated */
+   default boolean isAreaLoaded(BlockPos center, int range) {
+      return this.m_46832_(center.m_7918_(-range, -range, -range), center.m_7918_(range, range, range));
+   }
+
    @Deprecated
    default boolean m_46832_(BlockPos p_46833_, BlockPos p_46834_) {
       return this.m_46812_(p_46833_.m_123341_(), p_46833_.m_123342_(), p_46833_.m_123343_(), p_46834_.m_123341_(), p_46834_.m_123342_(), p_46834_.m_123343_());
