--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -10,12 +_,15 @@
 import java.util.function.Consumer;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+
+import net.minecraft.client.multiplayer.ClientChunkCache;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
 import net.minecraft.core.QuartPos;
 import net.minecraft.core.Registry;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.server.level.ServerChunkCache;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.BiomeTags;
 import net.minecraft.tags.BlockTags;
@@ -46,8 +_,10 @@
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.pathfinder.PathComputationType;
 import net.minecraft.world.phys.Vec3;
+import org.goldenforge.config.GoldenForgeConfig;
 import org.slf4j.Logger;
 
+// TODO: ForgeHooks.canEntitySpawn
 public final class NaturalSpawner {
    private static final Logger f_46977_ = LogUtils.getLogger();
    private static final int f_151589_ = 24;
@@ -64,6 +_,12 @@
    }
 
    public static NaturalSpawner.SpawnState m_186524_(int p_186525_, Iterable<Entity> p_186526_, NaturalSpawner.ChunkGetter p_186527_, LocalMobCapCalculator p_186528_) {
+      // Paper start - add countMobs parameter
+      return createState(p_186525_, p_186526_, p_186527_, p_186528_, false);
+   }
+
+   public static NaturalSpawner.SpawnState createState(int p_186525_, Iterable<Entity> p_186526_, NaturalSpawner.ChunkGetter p_186527_, LocalMobCapCalculator p_186528_, boolean countMobs) {
+      // Paper end
       PotentialCalculator potentialcalculator = new PotentialCalculator();
       Object2IntOpenHashMap<MobCategory> object2intopenhashmap = new Object2IntOpenHashMap<>();
       Iterator iterator = p_186526_.iterator();
@@ -84,7 +_,7 @@
             mob = (Mob)entity;
          } while(mob.m_21532_() || mob.m_8023_());
 
-         MobCategory mobcategory = entity.m_6095_().m_20674_();
+         MobCategory mobcategory = entity.getClassification(true);
          if (mobcategory != MobCategory.MISC) {
             Entity entity_f = entity;
             BlockPos blockpos = entity.m_20183_();
@@ -94,11 +_,16 @@
                   potentialcalculator.m_47192_(entity_f.m_20183_(), mobspawnsettings$mobspawncost.m_48400_());
                }
 
-               if (entity_f instanceof Mob) {
+               if (p_186528_ != null && entity_f instanceof Mob) { // Paper
                   p_186528_.m_186512_(p_186541_.m_7697_(), mobcategory);
                }
 
                object2intopenhashmap.addTo(mobcategory, 1);
+               // Paper start
+               if (countMobs) {
+                  ((ServerChunkCache) p_186541_.f_62776_.m_7726_()).f_8325_.updatePlayerMobTypeMap(entity_f);
+               }
+               // Paper end
             });
          }
       }
@@ -112,19 +_,47 @@
       p_47030_.m_46473_().m_6180_("spawner");
 
       for(MobCategory mobcategory : f_46979_) {
-         if ((p_47033_ || !mobcategory.m_21609_()) && (p_47034_ || mobcategory.m_21609_()) && (p_47035_ || !mobcategory.m_21610_()) && p_47032_.m_186548_(mobcategory, p_47031_.m_7697_())) {
-            m_47045_(mobcategory, p_47030_, p_47031_, p_47032_::m_47127_, p_47032_::m_47131_);
+
+         // Paper start - only allow spawns upto the limit per chunk and update count afterwards
+         int limit = mobcategory.m_21608_();
+         int currEntityCount = p_47032_.f_47111_.getInt(mobcategory);
+         int k1 = limit * p_47032_.m_47126_() / NaturalSpawner.f_46978_;
+         int difference = k1 - currEntityCount;
+
+         if (GoldenForgeConfig.Server.perPlayerMobSpawns.get()) {
+            int minDiff = Integer.MAX_VALUE;
+            final com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<net.minecraft.server.level.ServerPlayer> inRange = p_47030_.m_7726_().f_8325_.playerMobDistanceMap.getObjectsInRange(p_47031_.m_7697_());
+            if (inRange != null) {
+               final Object[] backingSet = inRange.getBackingSet();
+               for (int k = 0; k < backingSet.length; k++) {
+                  if (!(backingSet[k] instanceof final net.minecraft.server.level.ServerPlayer player)) {
+                     continue;
+                  }
+                  minDiff = Math.min(limit - p_47030_.m_7726_().f_8325_.getMobCountNear(player, mobcategory), minDiff);
+               }
+            }
+            difference = (minDiff == Integer.MAX_VALUE) ? 0 : minDiff;
+         }
+
+         if ((p_47033_ || !mobcategory.m_21609_()) && (p_47034_ || mobcategory.m_21609_()) && (p_47035_ || !mobcategory.m_21610_()) && difference > 0) {
+            // Paper start
+            int spawnCount = NaturalSpawner.spawnCategoryForChunk(mobcategory, p_47030_, p_47031_, p_47032_::m_47127_, p_47032_::m_47131_);
+            p_47032_.f_47111_.mergeInt(mobcategory, spawnCount, Integer::sum);
+            // Paper end
          }
       }
 
       p_47030_.m_46473_().m_7238_();
    }
 
-   public static void m_47045_(MobCategory p_47046_, ServerLevel p_47047_, LevelChunk p_47048_, NaturalSpawner.SpawnPredicate p_47049_, NaturalSpawner.AfterSpawnCallback p_47050_) {
-      BlockPos blockpos = m_47062_(p_47047_, p_47048_);
-      if (blockpos.m_123342_() >= p_47047_.m_141937_() + 1) {
-         m_47038_(p_47046_, p_47047_, p_47048_, blockpos, p_47049_, p_47050_);
+   public static int spawnCategoryForChunk(MobCategory group, ServerLevel world, LevelChunk chunk, NaturalSpawner.SpawnPredicate checker, NaturalSpawner.AfterSpawnCallback runner) {
+      // Paper end - add parameters and int ret type
+      BlockPos blockposition = NaturalSpawner.m_47062_(world, chunk);
+
+      if (blockposition.m_123342_() >= world.m_141937_() + 1) {
+         NaturalSpawner.m_47038_(group, world, chunk, blockposition, checker, runner); // Paper
       }
+      return 0; // Paper
    }
 
    @VisibleForDebug
@@ -135,14 +_,15 @@
       });
    }
 
+   // Paper start - add maxSpawns parameter and return spawned mobs
    public static void m_47038_(MobCategory p_47039_, ServerLevel p_47040_, ChunkAccess p_47041_, BlockPos p_47042_, NaturalSpawner.SpawnPredicate p_47043_, NaturalSpawner.AfterSpawnCallback p_47044_) {
       StructureManager structuremanager = p_47040_.m_215010_();
       ChunkGenerator chunkgenerator = p_47040_.m_7726_().m_8481_();
       int i = p_47042_.m_123342_();
-      BlockState blockstate = p_47041_.m_8055_(p_47042_);
-      if (!blockstate.m_60796_(p_47041_, p_47042_)) {
+      BlockState blockstate = p_47040_.getBlockStateIfLoadedAndInBounds(p_47042_); // Paper - don't load chunks for mob spawn
+      int j = 0;
+      if (blockstate != null && !blockstate.m_60796_(p_47041_, p_47042_)) { // Paper - don't load chunks for mob spawn
          BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
-         int j = 0;
 
          for(int k = 0; k < 3; ++k) {
             int l = p_47042_.m_123341_();
@@ -159,10 +_,11 @@
                blockpos$mutableblockpos.m_122178_(l, i, i1);
                double d0 = (double)l + 0.5D;
                double d1 = (double)i1 + 0.5D;
-               Player player = p_47040_.m_45924_(d0, (double)i, d1, -1.0D, false);
+               Player player = (p_47041_ instanceof LevelChunk) ? ((LevelChunk)p_47041_).findNearestPlayer(d0, i, d1, 576.0D, net.minecraft.world.entity.EntitySelector.f_20408_) : p_47040_.m_45924_(d0, (double)i, d1, -1.0D, false); // Paper - use chunk's player cache to optimize search in range
+
                if (player != null) {
                   double d2 = player.m_20275_(d0, (double)i, d1);
-                  if (m_47024_(p_47040_, p_47041_, blockpos$mutableblockpos, d2)) {
+                  if (p_47040_.isLoadedAndInBounds(blockpos$mutableblockpos) && m_47024_(p_47040_, p_47041_, blockpos$mutableblockpos, d2)) { // Paper - don't load chunks for mob spawn
                      if (mobspawnsettings$spawnerdata == null) {
                         Optional<MobSpawnSettings.SpawnerData> optional = m_220429_(p_47040_, structuremanager, chunkgenerator, p_47039_, p_47040_.f_46441_, blockpos$mutableblockpos);
                         if (optional.isEmpty()) {
@@ -173,21 +_,34 @@
                         k1 = mobspawnsettings$spawnerdata.f_48405_ + p_47040_.f_46441_.m_188503_(1 + mobspawnsettings$spawnerdata.f_48406_ - mobspawnsettings$spawnerdata.f_48405_);
                      }
 
-                     if (m_220421_(p_47040_, p_47039_, structuremanager, chunkgenerator, mobspawnsettings$spawnerdata, blockpos$mutableblockpos, d2) && p_47043_.m_47106_(mobspawnsettings$spawnerdata.f_48404_, blockpos$mutableblockpos, p_47041_)) {
+                     // Paper start
+                     Boolean doSpawning = m_220421_(p_47040_, p_47039_, structuremanager, chunkgenerator, mobspawnsettings$spawnerdata, blockpos$mutableblockpos, d2);
+                     if (doSpawning == null) {
+                        return; // Paper
+                     }
+                     if (doSpawning && p_47043_.m_47106_(mobspawnsettings$spawnerdata.f_48404_, blockpos$mutableblockpos, p_47041_)) {
+                        // Paper end
                         Mob mob = m_46988_(p_47040_, mobspawnsettings$spawnerdata.f_48404_);
                         if (mob == null) {
-                           return;
+                           return; // Paper
                         }
 
                         mob.m_7678_(d0, (double)i, d1, p_47040_.f_46441_.m_188501_() * 360.0F, 0.0F);
-                        if (m_46991_(p_47040_, mob, d2)) {
+                        int canSpawn = net.minecraftforge.common.ForgeHooks.canEntitySpawn(mob, p_47040_, d0, i, d1, null, MobSpawnType.NATURAL);
+                        if (canSpawn == 1 || (canSpawn == 0 && m_46991_(p_47040_, mob, d2))) {
+                           if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(mob, p_47040_, (float)d0, (float)i, (float)d1, null, MobSpawnType.NATURAL))
                            spawngroupdata = mob.m_6518_(p_47040_, p_47040_.m_6436_(mob.m_20183_()), MobSpawnType.NATURAL, spawngroupdata, (CompoundTag)null);
                            ++j;
                            ++l1;
                            p_47040_.m_47205_(mob);
                            p_47044_.m_47100_(mob, p_47041_);
-                           if (j >= mob.m_5792_()) {
-                              return;
+                           // Paper start
+                           if (GoldenForgeConfig.Server.perPlayerMobSpawns.get()) {
+                              p_47040_.m_7726_().f_8325_.updatePlayerMobTypeMap(mob);
+                           }
+                           // Paper end
+                           if (j >= net.minecraftforge.event.ForgeEventFactory.getMaxSpawnPackSize(mob)) {
+                              return; // Paper
                            }
 
                            if (mob.m_7296_(l1)) {
@@ -266,7 +_,8 @@
    }
 
    private static WeightedRandomList<MobSpawnSettings.SpawnerData> m_220443_(ServerLevel p_220444_, StructureManager p_220445_, ChunkGenerator p_220446_, MobCategory p_220447_, BlockPos p_220448_, @Nullable Holder<Biome> p_220449_) {
-      return m_220455_(p_220448_, p_220444_, p_220447_, p_220445_) ? NetherFortressStructure.f_228517_ : p_220446_.m_223133_(p_220449_ != null ? p_220449_ : p_220444_.m_204166_(p_220448_), p_220445_, p_220447_, p_220448_);
+      // Forge: Add in potential spawns, and replace hardcoded nether fortress mob list
+      return net.minecraftforge.event.ForgeEventFactory.getPotentialSpawns(p_220444_, p_220447_, p_220448_, m_220455_(p_220448_, p_220444_, p_220447_, p_220445_) ? p_220445_.m_220521_().m_175515_(Registry.f_235725_).m_123013_(BuiltinStructures.f_209859_).m_226612_().get(MobCategory.MONSTER).f_210044_() : p_220446_.m_223133_(p_220449_ != null ? p_220449_ : p_220444_.m_204166_(p_220448_), p_220445_, p_220447_, p_220448_));
    }
 
    public static boolean m_220455_(BlockPos p_220456_, ServerLevel p_220457_, MobCategory p_220458_, StructureManager p_220459_) {
@@ -280,10 +_,10 @@
 
    private static BlockPos m_47062_(Level p_47063_, LevelChunk p_47064_) {
       ChunkPos chunkpos = p_47064_.m_7697_();
-      int i = chunkpos.m_45604_() + p_47063_.f_46441_.m_188503_(16);
-      int j = chunkpos.m_45605_() + p_47063_.f_46441_.m_188503_(16);
+      int i = chunkpos.m_45604_() + p_47063_.getThreadUnsafeRandom().m_188503_(16); // Pufferfish - use thread unsafe random
+      int j = chunkpos.m_45605_() + p_47063_.getThreadUnsafeRandom().m_188503_(16); // Pufferfish - use thread unsafe random
       int k = p_47064_.m_5885_(Heightmap.Types.WORLD_SURFACE, i, j) + 1;
-      int l = Mth.m_216287_(p_47063_.f_46441_, p_47063_.m_141937_(), k);
+      int l = Mth.m_216287_(p_47063_.getThreadUnsafeRandom(), p_47063_.m_141937_(), k); // Pufferfish
       return new BlockPos(i, l, j);
    }
 
@@ -305,6 +_,13 @@
       if (p_47052_ == SpawnPlacements.Type.NO_RESTRICTIONS) {
          return true;
       } else if (p_47055_ != null && p_47053_.m_6857_().m_61937_(p_47054_)) {
+         return p_47052_.canSpawnAt(p_47053_, p_47054_, p_47055_);
+      }
+      return false;
+   }
+
+   public static boolean canSpawnAtBody(SpawnPlacements.Type p_47052_, LevelReader p_47053_, BlockPos p_47054_, @Nullable EntityType<?> p_47055_) {
+      {
          BlockState blockstate = p_47053_.m_8055_(p_47054_);
          FluidState fluidstate = p_47053_.m_6425_(p_47054_);
          BlockPos blockpos = p_47054_.m_7494_();
@@ -317,14 +_,12 @@
             case ON_GROUND:
             default:
                BlockState blockstate1 = p_47053_.m_8055_(blockpos1);
-               if (!blockstate1.m_60643_(p_47053_, blockpos1, p_47055_)) {
+            if (!blockstate1.isValidSpawn(p_47053_, blockpos1, p_47052_, p_47055_)) {
                   return false;
                } else {
                   return m_47056_(p_47053_, p_47054_, blockstate, fluidstate, p_47055_) && m_47056_(p_47053_, blockpos, p_47053_.m_8055_(blockpos), p_47053_.m_6425_(blockpos), p_47055_);
                }
          }
-      } else {
-         return false;
       }
    }
 
@@ -370,7 +_,9 @@
                         entity.m_7678_(d0, (double)blockpos.m_123342_(), d1, p_220454_.m_188501_() * 360.0F, 0.0F);
                         if (entity instanceof Mob) {
                            Mob mob = (Mob)entity;
-                           if (mob.m_5545_(p_220451_, MobSpawnType.CHUNK_GENERATION) && mob.m_6914_(p_220451_)) {
+                           int res = net.minecraftforge.common.ForgeHooks.canEntitySpawn(mob, p_220451_, d0, blockpos.m_123342_(), d1, null, MobSpawnType.CHUNK_GENERATION);
+                           if (res == 1 || (res == 0 && mob.m_5545_(p_220451_, MobSpawnType.CHUNK_GENERATION) && mob.m_6914_(p_220451_))) {
+                              if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(mob, p_220451_, (float) d0, blockpos.m_123342_(), (float) d1, null, MobSpawnType.CHUNK_GENERATION))
                               spawngroupdata = mob.m_6518_(p_220451_, p_220451_.m_6436_(mob.m_20183_()), MobSpawnType.CHUNK_GENERATION, spawngroupdata, (CompoundTag)null);
                               p_220451_.m_47205_(mob);
                               flag = true;
@@ -482,7 +_,7 @@
          this.f_47112_.m_47192_(blockpos, d0);
          MobCategory mobcategory = entitytype.m_20674_();
          this.f_47111_.addTo(mobcategory, 1);
-         this.f_186542_.m_186512_(new ChunkPos(blockpos), mobcategory);
+         if (this.f_186542_ != null) this.f_186542_.m_186512_(new ChunkPos(blockpos), mobcategory); // Paper
       }
 
       public int m_47126_() {
