--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -62,7 +_,7 @@
     private static final String DRAGON_PHASE_KEY = "DragonPhase";
     public final double[][] positions = new double[64][3];
     public int posPointer = -1;
-    private final EnderDragonPart[] subEntities;
+    public final EnderDragonPart[] subEntities; // Goldenforge - public
     public final EnderDragonPart head;
     private final EnderDragonPart neck;
     private final EnderDragonPart body;
