--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -57,6 +_,16 @@
 
     @Nullable
     public static MapItemSavedData getSavedData(ItemStack p_42854_, Level p_42855_) {
+        // Neo: Add instance method so that mods can override
+        Item map = p_42854_.getItem();
+        if(map instanceof MapItem) {
+            return ((MapItem)map).getCustomMapData(p_42854_, p_42855_);
+        }
+        return null;
+    }
+
+    @Nullable
+    protected MapItemSavedData getCustomMapData(ItemStack p_42854_, Level p_42855_) {
         MapId mapid = p_42854_.get(DataComponents.MAP_ID);
         return getSavedData(mapid, p_42855_);
     }
@@ -100,8 +_,8 @@
                             int j2 = (j / i + k1 - 64) * i;
                             int k2 = (k / i + l1 - 64) * i;
                             Multiset<MapColor> multiset = LinkedHashMultiset.create();
-                            LevelChunk levelchunk = p_42894_.getChunk(SectionPos.blockToSectionCoord(j2), SectionPos.blockToSectionCoord(k2));
-                            if (!levelchunk.isEmpty()) {
+                            LevelChunk levelchunk = p_42894_.getChunkIfLoaded(SectionPos.blockToSectionCoord(j2), SectionPos.blockToSectionCoord(k2)); // Paper - Maps shouldn't load chunks
+                            if (levelchunk != null && !levelchunk.isEmpty()) { // Paper - Maps shouldn't load chunks
                                 int l2 = 0;
                                 double d1 = 0.0;
                                 if (p_42894_.dimensionType().hasCeiling()) {
@@ -214,7 +_,7 @@
 
                 for (int j1 = 0; j1 < 128; j1++) {
                     for (int k1 = 0; k1 < 128; k1++) {
-                        Holder<Biome> holder = p_42851_.getBiome(blockpos$mutableblockpos.set((l + k1) * i, 0, (i1 + j1) * i));
+                        Holder<Biome> holder = p_42851_.getUncachedNoiseBiome((l + k1) * i, 0, (i1 + j1) * i); // Paper - Perf: Use seed based lookup for treasure maps
                         aboolean[j1 * 128 + k1] = holder.is(BiomeTags.WATER_ON_MAP_OUTLINES);
                     }
                 }
