--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -75,10 +_,10 @@
 
     @Override
     public CompletableFuture<ChunkAccess> createBiomes(RandomState p_224299_, Blender p_224300_, StructureManager p_224301_, ChunkAccess p_224302_) {
-        return CompletableFuture.supplyAsync(Util.wrapThreadWithTaskName("init_biomes", () -> {
+        return CompletableFuture.supplyAsync(() -> {
             this.doCreateBiomes(p_224300_, p_224299_, p_224301_, p_224302_);
             return p_224302_;
-        }), Util.backgroundExecutor());
+        }, Runnable::run); // Paper - rewrite chunk system
     }
 
     private void doCreateBiomes(Blender p_224292_, RandomState p_224293_, StructureManager p_224294_, ChunkAccess p_224295_) {
@@ -326,7 +_,7 @@
         int i = noisesettings.minY();
         int j = Mth.floorDiv(i, noisesettings.getCellHeight());
         int k = Mth.floorDiv(noisesettings.height(), noisesettings.getCellHeight());
-        return k <= 0 ? CompletableFuture.completedFuture(p_224316_) : CompletableFuture.supplyAsync(Util.wrapThreadWithTaskName("wgen_fill_noise", () -> {
+        return k <= 0 ? CompletableFuture.completedFuture(p_224316_) : CompletableFuture.supplyAsync(() -> {
             int l = p_224316_.getSectionIndex(k * noisesettings.getCellHeight() - 1 + i);
             int i1 = p_224316_.getSectionIndex(i);
             Set<LevelChunkSection> set = Sets.newHashSet();
@@ -347,7 +_,7 @@
             }
 
             return chunkaccess;
-        }), Util.backgroundExecutor());
+        }, Runnable::run); // Paper - rewrite chunk system
     }
 
     private ChunkAccess doFill(Blender p_224285_, StructureManager p_224286_, RandomState p_224287_, ChunkAccess p_224288_, int p_224289_, int p_224290_) {
