--- a/net/minecraft/world/entity/ai/village/poi/PoiSection.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiSection.java
@@ -26,7 +_,7 @@
 public class PoiSection {
    private static final Logger f_27260_ = LogUtils.getLogger();
    private final Short2ObjectMap<PoiRecord> f_27261_ = new Short2ObjectOpenHashMap<>();
-   private final Map<Holder<PoiType>, Set<PoiRecord>> f_27262_ = Maps.newHashMap();
+   private final Map<Holder<PoiType>, Set<PoiRecord>> f_27262_ = Maps.newHashMap(); ; public final Map<Holder<PoiType>, Set<PoiRecord>> getData() { return this.f_27262_; } // Paper - public accessor
    private final Runnable f_27263_;
    private boolean f_27264_;
 
