--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -178,7 +_,7 @@
     public static final String ATTRIBUTES_FIELD = "attributes";
     private final AttributeMap attributes;
     private final CombatTracker combatTracker = new CombatTracker(this);
-    private final Map<Holder<MobEffect>, MobEffectInstance> activeEffects = Maps.newHashMap();
+    public final Map<Holder<MobEffect>, MobEffectInstance> activeEffects = Maps.newHashMap(); // Goldenforge - public
     private final NonNullList<ItemStack> lastHandItemStacks = NonNullList.withSize(2, ItemStack.EMPTY);
     private final NonNullList<ItemStack> lastArmorItemStacks = NonNullList.withSize(4, ItemStack.EMPTY);
     private ItemStack lastBodyItemStack = ItemStack.EMPTY;
@@ -214,7 +_,7 @@
     protected float rotOffs;
     protected int deathScore;
     protected float lastHurt;
-    protected boolean jumping;
+    public boolean jumping; // Goldenforge - public
     public float xxa;
     public float yya;
     public float zza;
@@ -2388,6 +_,14 @@
         return false;
     }
 
+    // Paper start - EAR 2
+    @Override
+    public void inactiveTick() {
+        super.inactiveTick();
+        ++this.noActionTime; // Above all the floats
+    }
+    // Paper end - EAR 2
+
     @Override
     public void tick() {
         super.tick();
@@ -2972,9 +_,7 @@
         } else {
             Vec3 vec3 = new Vec3(this.getX(), this.getEyeY(), this.getZ());
             Vec3 vec31 = new Vec3(p_147185_.getX(), p_147185_.getEyeY(), p_147185_.getZ());
-            return vec31.distanceTo(vec3) > 128.0
-                ? false
-                : this.level().clip(new ClipContext(vec3, vec31, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, this)).getType() == HitResult.Type.MISS;
+            return vec31.distanceToSqr(vec3) > 128.0D * 128.0D && this.level().clip(new ClipContext(vec3, vec31, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, this)).getType() == HitResult.Type.MISS; // Paper - Perf: Use distance squared
         }
     }
 
