--- a/net/minecraft/world/entity/PathfinderMob.java
+++ b/net/minecraft/world/entity/PathfinderMob.java
@@ -17,6 +_,8 @@
         super(p_21683_, p_21684_);
     }
 
+    public BlockPos movingTarget; public BlockPos getMovingTarget() { return movingTarget; } // Paper
+
     public float getWalkTargetValue(BlockPos p_21693_) {
         return this.getWalkTargetValue(p_21693_, this.level());
     }
