--- a/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -22,9 +_,11 @@
    private final Map<Attribute, AttributeInstance> f_22139_ = Maps.newHashMap();
    private final Set<AttributeInstance> f_22140_ = Sets.newHashSet();
    private final AttributeSupplier f_22141_;
+   private final java.util.function.Function<Attribute, AttributeInstance> createInstance; // Pufferfish
 
    public AttributeMap(AttributeSupplier p_22144_) {
       this.f_22141_ = p_22144_;
+      this.createInstance = attribute -> this.f_22141_.m_22250_(this::m_22157_, attribute); // Pufferfish
    }
 
    private void m_22157_(AttributeInstance p_22158_) {
@@ -46,9 +_,7 @@
 
    @Nullable
    public AttributeInstance m_22146_(Attribute p_22147_) {
-      return this.f_22139_.computeIfAbsent(p_22147_, (p_22188_) -> {
-         return this.f_22141_.m_22250_(this::m_22157_, p_22188_);
-      });
+      return this.f_22139_.computeIfAbsent(p_22147_, this.createInstance); // Pufferfish - cache lambda, as for some reason java allocates it anyways
    }
 
    public boolean m_22171_(Attribute p_22172_) {
