--- a/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -10,6 +_,9 @@
 import java.util.UUID;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
+
+import it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
 import net.minecraft.Util;
 import net.minecraft.core.Registry;
 import net.minecraft.nbt.CompoundTag;
@@ -19,8 +_,8 @@
 
 public class AttributeMap {
    private static final Logger f_22138_ = LogUtils.getLogger();
-   private final Map<Attribute, AttributeInstance> f_22139_ = Maps.newHashMap();
-   private final Set<AttributeInstance> f_22140_ = Sets.newHashSet();
+   private final Map<Attribute, AttributeInstance> f_22139_ = Maps.newConcurrentMap();
+   private final Set<AttributeInstance> f_22140_ = Sets.newConcurrentHashSet();
    private final AttributeSupplier f_22141_;
 
    public AttributeMap(AttributeSupplier p_22144_) {
@@ -46,8 +_,8 @@
 
    @Nullable
    public AttributeInstance m_22146_(Attribute p_22147_) {
-      return this.f_22139_.computeIfAbsent(p_22147_, (p_22188_) -> {
-         return this.f_22141_.m_22250_(this::m_22157_, p_22188_);
+      return this.f_22139_.computeIfAbsent(p_22147_, (attributex) -> {
+         return this.f_22141_.m_22250_(this::m_22157_, attributex);
       });
    }
 
