--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -41,7 +_,7 @@
    protected long f_26503_;
    protected double f_26504_;
    protected float f_26505_ = 0.5F;
-   protected boolean f_26506_;
+   protected boolean f_26506_; protected final boolean needsPathRecalculation() { return this.f_26506_; } // Paper - public accessor
    protected long f_26507_;
    protected NodeEvaluator f_26508_;
    @Nullable
@@ -51,6 +_,14 @@
    private final PathFinder f_26512_;
    private boolean f_26513_;
 
+
+   // Paper start
+   public boolean isViableForPathRecalculationChecking() {
+      return !this.needsPathRecalculation() &&
+              (this.f_26496_ != null && !this.f_26496_.m_77392_() && this.f_26496_.m_77398_() != 0);
+   }
+   // Paper end
+
    public PathNavigation(Mob p_26515_, Level p_26516_) {
       this.f_26494_ = p_26515_;
       this.f_26495_ = p_26516_;
@@ -157,9 +_,29 @@
       return this.m_26536_(this.m_26524_(p_26520_, p_26521_, p_26522_, 1), p_26523_);
    }
 
+   // Paper start - optimise pathfinding
+   private int lastFailure = 0;
+   private int pathfindFailures = 0;
+   // Paper end
+
    public boolean m_5624_(Entity p_26532_, double p_26533_) {
+      // Paper start - Pathfinding optimizations
+      if (this.pathfindFailures > 10 && this.f_26496_ == null && net.minecraft.server.MinecraftServer.currentTick < this.lastFailure + 40) {
+         return false;
+      }
+      // Paper end
       Path path = this.m_6570_(p_26532_, 1);
-      return path != null && this.m_26536_(path, p_26533_);
+      // Paper start - Pathfinding optimizations
+      if (path != null && this.m_26536_(path, p_26533_)) {
+         this.lastFailure = 0;
+         this.pathfindFailures = 0;
+         return true;
+      } else {
+         this.pathfindFailures++;
+         this.lastFailure = net.minecraft.server.MinecraftServer.currentTick;
+         return false;
+      }
+      // Paper end
    }
 
    public boolean m_26536_(@Nullable Path p_26537_, double p_26538_) {
@@ -227,10 +_,10 @@
       Vec3 vec3 = this.m_7475_();
       this.f_26505_ = this.f_26494_.m_20205_() > 0.75F ? this.f_26494_.m_20205_() / 2.0F : 0.75F - this.f_26494_.m_20205_() / 2.0F;
       Vec3i vec3i = this.f_26496_.m_77400_();
-      double d0 = Math.abs(this.f_26494_.m_20185_() - ((double)vec3i.m_123341_() + 0.5D));
+      double d0 = Math.abs(this.f_26494_.m_20185_() - ((double)vec3i.m_123341_() + (this.f_26494_.m_20205_() + 1) / 2D)); //Forge: Fix MC-94054
       double d1 = Math.abs(this.f_26494_.m_20186_() - (double)vec3i.m_123342_());
-      double d2 = Math.abs(this.f_26494_.m_20189_() - ((double)vec3i.m_123343_() + 0.5D));
-      boolean flag = d0 < (double)this.f_26505_ && d2 < (double)this.f_26505_ && d1 < 1.0D;
+      double d2 = Math.abs(this.f_26494_.m_20189_() - ((double)vec3i.m_123343_() + (this.f_26494_.m_20205_() + 1) / 2D)); //Forge: Fix MC-94054
+      boolean flag = d0 <= (double)this.f_26505_ && d2 <= (double)this.f_26505_ && d1 < 1.0D; //Forge: Fix MC-94054
       if (flag || this.f_26494_.m_21481_(this.f_26496_.m_77401_().f_77282_) && this.m_26559_(vec3)) {
          this.f_26496_.m_77374_();
       }
