--- a/net/minecraft/world/level/gameevent/GameEventDispatcher.java
+++ b/net/minecraft/world/level/gameevent/GameEventDispatcher.java
@@ -39,7 +_,7 @@
 
         for (int l1 = j; l1 <= i1; l1++) {
             for (int i2 = l; i2 <= k1; i2++) {
-                ChunkAccess chunkaccess = this.level.getChunkSource().getChunkNow(l1, i2);
+                ChunkAccess chunkaccess = this.level.getChunkIfLoadedImmediately(l1, i2); // Paper - Use getChunkIfLoadedImmediately
                 if (chunkaccess != null) {
                     for (int j2 = k; j2 <= j1; j2++) {
                         flag |= chunkaccess.getListenerRegistry(j2)
