--- a/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -216,7 +_,7 @@
     }
 
     @Override
-    protected void setDirty(long p_27036_) {
+    public void setDirty(long p_27036_) {
         super.setDirty(p_27036_);
         this.distanceTracker.update(p_27036_, this.distanceTracker.getLevelFromSource(p_27036_), false);
     }
