--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -70,12 +_,12 @@
         }
     }
 
-    private static ChunkPos readChunkPos(CompoundTag p_156571_) {
+    public static ChunkPos readChunkPos(CompoundTag p_156571_) {
         int[] aint = p_156571_.getIntArray("Position");
         return new ChunkPos(aint[0], aint[1]);
     }
 
-    private static void writeChunkPos(CompoundTag p_156563_, ChunkPos p_156564_) {
+    public static void writeChunkPos(CompoundTag p_156563_, ChunkPos p_156564_) {
         p_156563_.put("Position", new IntArrayTag(new int[]{p_156564_.x, p_156564_.z}));
     }
 
@@ -94,8 +_,12 @@
             ListTag listtag = new ListTag();
             p_156559_.getEntities().forEach(p_156567_ -> {
                 CompoundTag compoundtag1 = new CompoundTag();
+                try {
                 if (p_156567_.save(compoundtag1)) {
                     listtag.add(compoundtag1);
+                }
+                } catch (Exception e) {
+                    LOGGER.error("An Entity type {} has thrown an exception trying to write state. It will not persist. Report this to the mod author", p_156567_.getType(), e);
                 }
             });
             CompoundTag compoundtag = NbtUtils.addCurrentDataVersion(new CompoundTag());
