--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -70,7 +_,7 @@
     private final RconConsoleSource rconConsoleSource;
     @Nullable
     private RconThread rconThread;
-    private final DedicatedServerSettings settings;
+    public final DedicatedServerSettings settings;
     @Nullable
     private MinecraftServerGui gui;
     @Nullable
@@ -136,6 +_,12 @@
             this.setLocalIp(dedicatedserverproperties.serverIp);
         }
 
+        // Paper start - initialize global and world-defaults configuration
+        this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
+        this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        // Paper end - initialize global and world-defaults configuration
+        org.spigotmc.WatchdogThread.doStart(60); // Paper - start watchdog thread
+
         this.setPvpAllowed(dedicatedserverproperties.pvp);
         this.setFlightAllowed(dedicatedserverproperties.allowFlight);
         this.setMotd(dedicatedserverproperties.motd);
@@ -174,7 +_,7 @@
         }
 
         if (this.convertOldUsers()) {
-            this.getProfileCache().save();
+            this.getProfileCache().save(false); // Paper
         }
 
         if (!OldUsersConverter.serverReadyAfterUserconversion(this)) {
@@ -209,7 +_,7 @@
                 this.rconThread = RconThread.create(this);
             }
 
-            if (this.getMaxTickLength() > 0L) {
+            if (false && this.getMaxTickLength() > 0L) { // Spigot - disable
                 Thread thread1 = new Thread(new ServerWatchdog(this));
                 thread1.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(LOGGER));
                 thread1.setName("Server Watchdog");
@@ -227,6 +_,7 @@
                 this.dediLanPinger.start();
             }
 
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().pufferfishConfig.enableAsyncMobSpawning) mobSpawnExecutor.start(); // Pufferfish
             net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
             return true;
         }
