--- a/net/minecraft/resources/ResourceLocation.java
+++ b/net/minecraft/resources/ResourceLocation.java
@@ -32,8 +_,10 @@
       this.f_135804_ = StringUtils.isEmpty(p_135814_[0]) ? "minecraft" : p_135814_[0];
       this.f_135805_ = p_135814_[1];
       if (!m_135843_(this.f_135804_)) {
+         new Exception().printStackTrace();
          throw new ResourceLocationException("Non [a-z0-9_.-] character in namespace of location: " + this.f_135804_ + ":" + this.f_135805_);
       } else if (!m_135841_(this.f_135805_)) {
+         new Exception().printStackTrace();
          throw new ResourceLocationException("Non [a-z0-9/._-] character in path of location: " + this.f_135804_ + ":" + this.f_135805_);
       }
    }
@@ -123,6 +_,12 @@
       }
 
       return i;
+   }
+
+   // Normal compare sorts by path first, this compares namespace first.
+   public int compareNamespaced(ResourceLocation o) {
+      int ret = this.f_135804_.compareTo(o.f_135804_);
+      return ret != 0 ? ret : this.f_135805_.compareTo(o.f_135805_);
    }
 
    public String m_179910_() {
