--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -14,6 +_,8 @@
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
+import net.minecraft.client.multiplayer.ClientChunkCache;
+import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Registry;
 import net.minecraft.core.SectionPos;
@@ -21,6 +_,7 @@
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.game.ClientboundLevelChunkPacketData;
 import net.minecraft.server.level.ChunkHolder;
+import net.minecraft.server.level.ServerChunkCache;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.profiling.ProfilerFiller;
 import net.minecraft.world.entity.Entity;
@@ -48,7 +_,7 @@
 import net.minecraft.world.ticks.TickContainerAccess;
 import org.slf4j.Logger;
 
-public class LevelChunk extends ChunkAccess {
+public class LevelChunk extends ChunkAccess implements net.minecraftforge.common.capabilities.ICapabilityProviderImpl<LevelChunk> {
    static final Logger f_62771_ = LogUtils.getLogger();
    private static final TickingBlockEntity f_156361_ = new TickingBlockEntity() {
       public void m_142224_() {
@@ -69,7 +_,7 @@
    private final Map<BlockPos, LevelChunk.RebindableTickingBlockEntityWrapper> f_156362_ = Maps.newHashMap();
    private boolean f_62775_;
    private boolean f_196848_ = false;
-   final Level f_62776_;
+   public final Level f_62776_;
    @Nullable
    private Supplier<ChunkHolder.FullChunkStatus> f_62790_;
    @Nullable
@@ -77,13 +_,377 @@
    private final Int2ObjectMap<GameEventDispatcher> f_156363_;
    private final LevelChunkTicks<Block> f_62784_;
    private final LevelChunkTicks<Fluid> f_187943_;
+   // Paper start - track last save time
+   public long lastSaveTime;
+   public void setLastSaved(long ticks) {
+      this.lastSaveTime = ticks;
+   }
+   // Paper end
+
+   public boolean mustNotSave = false;
+   boolean loadedTicketLevel; public final boolean wasLoadCallbackInvoked() { return this.loadedTicketLevel; } // Paper - public accessor
+
+   // Paper start - optimise checkDespawn
+   private boolean playerGeneralAreaCacheSet;
+   private com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<net.minecraft.server.level.ServerPlayer> playerGeneralAreaCache;
+
+   public com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<net.minecraft.server.level.ServerPlayer> getPlayerGeneralAreaCache() {
+      if (!this.playerGeneralAreaCacheSet) {
+         this.updateGeneralAreaCache();
+      }
+      return this.playerGeneralAreaCache;
+   }
+
+   public void updateGeneralAreaCache() {
+      this.updateGeneralAreaCache(((ServerLevel)this.f_62776_).m_7726_().f_8325_.playerGeneralAreaMap.getObjectsInRange(this.coordinateKey));
+   }
+
+   public void removeGeneralAreaCache() {
+      this.playerGeneralAreaCacheSet = false;
+      this.playerGeneralAreaCache = null;
+   }
+
+   public void updateGeneralAreaCache(com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<net.minecraft.server.level.ServerPlayer> value) {
+      this.playerGeneralAreaCacheSet = true;
+      this.playerGeneralAreaCache = value;
+   }
+
+   public net.minecraft.server.level.ServerPlayer findNearestPlayer(double sourceX, double sourceY, double sourceZ,
+                                                                    double maxRange, java.util.function.Predicate<Entity> predicate) {
+      if (!this.playerGeneralAreaCacheSet) {
+         this.updateGeneralAreaCache();
+      }
+
+      com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<net.minecraft.server.level.ServerPlayer> nearby = this.playerGeneralAreaCache;
+
+      if (nearby == null) {
+         return null;
+      }
+
+      Object[] backingSet = nearby.getBackingSet();
+      double closestDistance = maxRange < 0.0 ? Double.MAX_VALUE : maxRange * maxRange;
+      net.minecraft.server.level.ServerPlayer closest = null;
+      for (int i = 0, len = backingSet.length; i < len; ++i) {
+         Object _player = backingSet[i];
+         if (!(_player instanceof net.minecraft.server.level.ServerPlayer)) {
+            continue;
+         }
+         net.minecraft.server.level.ServerPlayer player = (net.minecraft.server.level.ServerPlayer)_player;
+
+         double distance = player.m_20275_(sourceX, sourceY, sourceZ);
+         if (distance < closestDistance && predicate.test(player)) {
+            closest = player;
+            closestDistance = distance;
+         }
+      }
+
+      return closest;
+   }
+
+   public void getNearestPlayers(double sourceX, double sourceY, double sourceZ, java.util.function.Predicate<Entity> predicate,
+                                 double range, java.util.List<net.minecraft.server.level.ServerPlayer> ret) {
+      if (!this.playerGeneralAreaCacheSet) {
+         this.updateGeneralAreaCache();
+      }
+
+      com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<net.minecraft.server.level.ServerPlayer> nearby = this.playerGeneralAreaCache;
+
+      if (nearby == null) {
+         return;
+      }
+
+      double rangeSquared = range * range;
+
+      Object[] backingSet = nearby.getBackingSet();
+      for (int i = 0, len = backingSet.length; i < len; ++i) {
+         Object _player = backingSet[i];
+         if (!(_player instanceof net.minecraft.server.level.ServerPlayer)) {
+            continue;
+         }
+         net.minecraft.server.level.ServerPlayer player = (net.minecraft.server.level.ServerPlayer)_player;
+
+         if (range >= 0.0) {
+            double distanceSquared = player.m_20275_(sourceX, sourceY, sourceZ);
+            if (distanceSquared > rangeSquared) {
+               continue;
+            }
+         }
+
+         if (predicate == null || predicate.test(player)) {
+            ret.add(player);
+         }
+      }
+   }
+   // Paper end - optimise checkDespawn
 
    public LevelChunk(Level p_187945_, ChunkPos p_187946_) {
       this(p_187945_, p_187946_, UpgradeData.f_63320_, new LevelChunkTicks<>(), new LevelChunkTicks<>(), 0L, (LevelChunkSection[])null, (LevelChunk.PostLoadProcessor)null, (BlendingData)null);
    }
 
+
+   // Paper start - new load callbacks
+   private io.papermc.paper.chunk.system.scheduling.NewChunkHolder chunkHolder;
+   public io.papermc.paper.chunk.system.scheduling.NewChunkHolder getChunkHolder() {
+      return this.chunkHolder;
+   }
+
+   public void setChunkHolder(io.papermc.paper.chunk.system.scheduling.NewChunkHolder chunkHolder) {
+      if (chunkHolder == null) {
+         throw new NullPointerException("Chunkholder cannot be null");
+      }
+      if (this.chunkHolder != null) {
+         throw new IllegalStateException("Already have chunkholder: " + this.chunkHolder + ", cannot replace with " + chunkHolder);
+      }
+      this.chunkHolder = chunkHolder;
+      this.playerChunk = chunkHolder.vanillaChunkHolder;
+   }
+
+   /* Note: We skip the light neighbour chunk loading done for the vanilla full chunk */
+   /* Starlight does not need these chunks for lighting purposes because of edge checks */
+   public void pushChunkIntoLoadedMap() {
+      int chunkX = this.f_187604_.f_45578_;
+      int chunkZ = this.f_187604_.f_45579_;
+      ServerChunkCache chunkProvider = ((ServerChunkCache) this.f_62776_.m_7726_());
+      for (int dx = -NEIGHBOUR_CACHE_RADIUS; dx <= NEIGHBOUR_CACHE_RADIUS; ++dx) {
+         for (int dz = -NEIGHBOUR_CACHE_RADIUS; dz <= NEIGHBOUR_CACHE_RADIUS; ++dz) {
+            LevelChunk neighbour = chunkProvider.getChunkAtIfLoadedMainThreadNoCache(chunkX + dx, chunkZ + dz);
+            if (neighbour != null) {
+               neighbour.setNeighbourLoaded(-dx, -dz, this);
+               // should be in cached already
+               this.setNeighbourLoaded(dx, dz, neighbour);
+            }
+         }
+      }
+      this.setNeighbourLoaded(0, 0, this);
+      ((ServerChunkCache) this.f_62776_.m_7726_()).addLoadedChunk(this);
+   }
+
+   public void onChunkLoad(io.papermc.paper.chunk.system.scheduling.NewChunkHolder chunkHolder) {
+      // figure out how this should interface with:
+      // the entity chunk load event // -> moved to the FULL status
+      // the chunk load event // -> stays here
+      // any entity add to world events // -> in FULL status
+      this.loadCallback();
+      io.papermc.paper.chunk.system.ChunkSystem.onChunkBorder(this, chunkHolder.vanillaChunkHolder);
+   }
+
+   public void onChunkUnload(io.papermc.paper.chunk.system.scheduling.NewChunkHolder chunkHolder) {
+      // figure out how this should interface with:
+      // the entity chunk load event // -> moved to chunk unload to disk (not written yet)
+      // the chunk load event // -> stays here
+      // any entity add to world events // -> goes into the unload logic, it will completely explode
+      // etc later
+      this.unloadCallback();
+      io.papermc.paper.chunk.system.ChunkSystem.onChunkNotBorder(this, chunkHolder.vanillaChunkHolder);
+   }
+
+   public void onChunkTicking(io.papermc.paper.chunk.system.scheduling.NewChunkHolder chunkHolder) {
+      this.m_62812_();
+      ((ServerLevel) this.f_62776_).m_184102_(this);
+      io.papermc.paper.chunk.system.ChunkSystem.onChunkTicking(this, chunkHolder.vanillaChunkHolder);
+   }
+
+   public void onChunkNotTicking(io.papermc.paper.chunk.system.scheduling.NewChunkHolder chunkHolder) {
+      io.papermc.paper.chunk.system.ChunkSystem.onChunkNotTicking(this, chunkHolder.vanillaChunkHolder);
+   }
+
+   public void onChunkEntityTicking(io.papermc.paper.chunk.system.scheduling.NewChunkHolder chunkHolder) {
+      io.papermc.paper.chunk.system.ChunkSystem.onChunkEntityTicking(this, chunkHolder.vanillaChunkHolder);
+   }
+
+   public void onChunkNotEntityTicking(io.papermc.paper.chunk.system.scheduling.NewChunkHolder chunkHolder) {
+      io.papermc.paper.chunk.system.ChunkSystem.onChunkNotEntityTicking(this, chunkHolder.vanillaChunkHolder);
+   }
+   // Paper end - new load callbacks
+
+   public void loadCallback() {
+      if (this.loadedTicketLevel) { f_62771_.error("Double calling chunk load!", new Throwable()); } // Paper
+      // Paper start - neighbour cache
+      int chunkX = this.f_187604_.f_45578_;
+      int chunkZ = this.f_187604_.f_45579_;
+      ServerChunkCache chunkProvider = ((ServerChunkCache) this.f_62776_.m_7726_());
+      for (int dx = -NEIGHBOUR_CACHE_RADIUS; dx <= NEIGHBOUR_CACHE_RADIUS; ++dx) {
+         for (int dz = -NEIGHBOUR_CACHE_RADIUS; dz <= NEIGHBOUR_CACHE_RADIUS; ++dz) {
+            LevelChunk neighbour = chunkProvider.getChunkAtIfLoadedMainThreadNoCache(chunkX + dx, chunkZ + dz);
+            if (neighbour != null) {
+               neighbour.setNeighbourLoaded(-dx, -dz, this);
+               // should be in cached already
+               this.setNeighbourLoaded(dx, dz, neighbour);
+            }
+         }
+      }
+      this.setNeighbourLoaded(0, 0, this);
+      // Paper end - neighbour cache
+      this.loadedTicketLevel = true;
+      ((ServerChunkCache) this.f_62776_.m_7726_()).addLoadedChunk(this); // Paper
+      ((ServerLevel)this.f_62776_).m_7726_().f_8325_.playerChunkManager.onChunkLoad(this.f_187604_.f_45578_, this.f_187604_.f_45579_); // Paper - rewrite player chunk management
+   }
+
+   public void unloadCallback() {
+      if (!this.loadedTicketLevel) { f_62771_.error("Double calling chunk unload!", new Throwable()); } // Paper
+      ((ServerChunkCache) this.f_62776_.m_7726_()).removeLoadedChunk(this); // Paper
+      // Paper start - neighbour cache
+      int chunkX = this.f_187604_.f_45578_;
+      int chunkZ = this.f_187604_.f_45579_;
+      ServerChunkCache chunkProvider = ((ServerChunkCache) this.f_62776_.m_7726_());
+      for (int dx = -NEIGHBOUR_CACHE_RADIUS; dx <= NEIGHBOUR_CACHE_RADIUS; ++dx) {
+         for (int dz = -NEIGHBOUR_CACHE_RADIUS; dz <= NEIGHBOUR_CACHE_RADIUS; ++dz) {
+            LevelChunk neighbour = chunkProvider.getChunkAtIfLoadedMainThreadNoCache(chunkX + dx, chunkZ + dz);
+            if (neighbour != null) {
+               neighbour.setNeighbourUnloaded(-dx, -dz);
+            }
+         }
+      }
+      this.loadedTicketLevel = false;
+      this.resetNeighbours();
+      // Paper end
+   }
+
+    // Paper start
+   public @Nullable ChunkHolder playerChunk;
+
+   static final int NEIGHBOUR_CACHE_RADIUS = 3;
+   public static int getNeighbourCacheRadius() {
+      return NEIGHBOUR_CACHE_RADIUS;
+   }
+
+   private long neighbourChunksLoadedBitset;
+   private final LevelChunk[] loadedNeighbourChunks = new LevelChunk[(NEIGHBOUR_CACHE_RADIUS * 2 + 1) * (NEIGHBOUR_CACHE_RADIUS * 2 + 1)];
+
+   private static int getNeighbourIndex(final int relativeX, final int relativeZ) {
+      // index = (relativeX + NEIGHBOUR_CACHE_RADIUS) + (relativeZ + NEIGHBOUR_CACHE_RADIUS) * (NEIGHBOUR_CACHE_RADIUS * 2 + 1)
+      // optimised variant of the above by moving some of the ops to compile time
+      return relativeX + (relativeZ * (NEIGHBOUR_CACHE_RADIUS * 2 + 1)) + (NEIGHBOUR_CACHE_RADIUS + NEIGHBOUR_CACHE_RADIUS * ((NEIGHBOUR_CACHE_RADIUS * 2 + 1)));
+   }
+
+   public final LevelChunk getRelativeNeighbourIfLoaded(final int relativeX, final int relativeZ) {
+      return this.loadedNeighbourChunks[getNeighbourIndex(relativeX, relativeZ)];
+   }
+
+   public final boolean isNeighbourLoaded(final int relativeX, final int relativeZ) {
+      return (this.neighbourChunksLoadedBitset & (1L << getNeighbourIndex(relativeX, relativeZ))) != 0;
+   }
+
+   public final void setNeighbourLoaded(final int relativeX, final int relativeZ, final LevelChunk chunk) {
+      if (chunk == null) {
+         throw new IllegalArgumentException("Chunk must be non-null, neighbour: (" + relativeX + "," + relativeZ + "), chunk: " + this.f_187604_);
+      }
+      final long before = this.neighbourChunksLoadedBitset;
+      final int index = getNeighbourIndex(relativeX, relativeZ);
+      this.loadedNeighbourChunks[index] = chunk;
+      this.neighbourChunksLoadedBitset |= (1L << index);
+      this.onNeighbourChange(before, this.neighbourChunksLoadedBitset);
+   }
+
+   public final void setNeighbourUnloaded(final int relativeX, final int relativeZ) {
+      final long before = this.neighbourChunksLoadedBitset;
+      final int index = getNeighbourIndex(relativeX, relativeZ);
+      this.loadedNeighbourChunks[index] = null;
+      this.neighbourChunksLoadedBitset &= ~(1L << index);
+      this.onNeighbourChange(before, this.neighbourChunksLoadedBitset);
+   }
+
+   public final void resetNeighbours() {
+      final long before = this.neighbourChunksLoadedBitset;
+      this.neighbourChunksLoadedBitset = 0L;
+      java.util.Arrays.fill(this.loadedNeighbourChunks, null);
+      this.onNeighbourChange(before, 0L);
+   }
+
+   protected void onNeighbourChange(final long bitsetBefore, final long bitsetAfter) {
+
+      // Paper start - no-tick view distance
+      ServerChunkCache chunkProviderServer = ((ServerLevel)this.f_62776_).m_7726_();
+      net.minecraft.server.level.ChunkMap chunkMap = chunkProviderServer.f_8325_;
+      // this code handles the addition of ticking tickets - the distance map handles the removal
+      if (!areNeighboursLoaded(bitsetBefore, 2) && areNeighboursLoaded(bitsetAfter, 2)) {
+         if (chunkMap.playerChunkManager.tickMap.getObjectsInRange(this.coordinateKey) != null) { // Paper - replace old player chunk loading system
+            // now we're ready for entity ticking
+            chunkProviderServer.f_8332_.execute(() -> {
+               // double check that this condition still holds.
+               if (LevelChunk.this.areNeighboursLoaded(2) && chunkMap.playerChunkManager.tickMap.getObjectsInRange(LevelChunk.this.coordinateKey) != null) { // Paper - replace old player chunk loading system
+                  chunkMap.playerChunkManager.onChunkPlayerTickReady(this.f_187604_.f_45578_, this.f_187604_.f_45579_); // Paper - replace old player chunk
+                  chunkProviderServer.addTicketAtLevel(net.minecraft.server.level.TicketType.f_9444_, LevelChunk.this.f_187604_, 31, LevelChunk.this.f_187604_); // 31 -> entity ticking, TODO check on update
+               }
+            });
+         }
+      }
+
+      // this code handles the chunk sending
+      if (!areNeighboursLoaded(bitsetBefore, 1) && areNeighboursLoaded(bitsetAfter, 1)) {
+         // Paper start - replace old player chunk loading system
+         if (chunkMap.playerChunkManager.isChunkNearPlayers(this.f_187604_.f_45578_, this.f_187604_.f_45579_)) {
+            // the post processing is expensive, so we don't want to run it unless we're actually near
+            // a player.
+            chunkProviderServer.f_8332_.execute(() -> {
+               if (!LevelChunk.this.areNeighboursLoaded(1)) {
+                  return;
+               }
+               LevelChunk.this.m_62812_();
+               if (!LevelChunk.this.areNeighboursLoaded(1)) {
+                  return;
+               }
+               chunkMap.playerChunkManager.onChunkSendReady(this.f_187604_.f_45578_, this.f_187604_.f_45579_);
+            });
+         }
+         // Paper end - replace old player chunk loading system
+      }
+      // Paper end - no-tick view distance
+   }
+
+   public final boolean isAnyNeighborsLoaded() {
+      return neighbourChunksLoadedBitset != 0;
+   }
+   public final boolean areNeighboursLoaded(final int radius) {
+      return LevelChunk.areNeighboursLoaded(this.neighbourChunksLoadedBitset, radius);
+   }
+
+   public static boolean areNeighboursLoaded(final long bitset, final int radius) {
+      // index = relativeX + (relativeZ * (NEIGHBOUR_CACHE_RADIUS * 2 + 1)) + (NEIGHBOUR_CACHE_RADIUS + NEIGHBOUR_CACHE_RADIUS * ((NEIGHBOUR_CACHE_RADIUS * 2 + 1)))
+      switch (radius) {
+         case 0: {
+            return (bitset & (1L << getNeighbourIndex(0, 0))) != 0;
+         }
+         case 1: {
+            long mask = 0L;
+            for (int dx = -1; dx <= 1; ++dx) {
+               for (int dz = -1; dz <= 1; ++dz) {
+                  mask |= (1L << getNeighbourIndex(dx, dz));
+               }
+            }
+            return (bitset & mask) == mask;
+         }
+         case 2: {
+            long mask = 0L;
+            for (int dx = -2; dx <= 2; ++dx) {
+               for (int dz = -2; dz <= 2; ++dz) {
+                  mask |= (1L << getNeighbourIndex(dx, dz));
+               }
+            }
+            return (bitset & mask) == mask;
+         }
+         case 3: {
+            long mask = 0L;
+            for (int dx = -3; dx <= 3; ++dx) {
+               for (int dz = -3; dz <= 3; ++dz) {
+                  mask |= (1L << getNeighbourIndex(dx, dz));
+               }
+            }
+            return (bitset & mask) == mask;
+         }
+
+         default:
+            throw new IllegalArgumentException("Radius not recognized: " + radius);
+      }
+   }
+   // Paper end
+
    public LevelChunk(Level p_196854_, ChunkPos p_196855_, UpgradeData p_196856_, LevelChunkTicks<Block> p_196857_, LevelChunkTicks<Fluid> p_196858_, long p_196859_, @Nullable LevelChunkSection[] p_196860_, @Nullable LevelChunk.PostLoadProcessor p_196861_, @Nullable BlendingData p_196862_) {
       super(p_196855_, p_196856_, p_196854_, p_196854_.m_5962_().m_175515_(Registry.f_122885_), p_196859_, p_196860_, p_196862_);
+      // Paper start - rewrite light engine
+      this.setBlockNibbles(ca.spottedleaf.starlight.common.light.StarLightEngine.getFilledEmptyLight(p_196854_));
+      this.setSkyNibbles(ca.spottedleaf.starlight.common.light.StarLightEngine.getFilledEmptyLight(p_196854_));
+      // Paper end - rewrite light engine
       this.f_62776_ = p_196854_;
       this.f_156363_ = new Int2ObjectOpenHashMap<>();
 
@@ -96,11 +_,17 @@
       this.f_62791_ = p_196861_;
       this.f_62784_ = p_196857_;
       this.f_187943_ = p_196858_;
+      this.capProvider.initInternal();
    }
 
    public LevelChunk(ServerLevel p_196850_, ProtoChunk p_196851_, @Nullable LevelChunk.PostLoadProcessor p_196852_) {
       this(p_196850_, p_196851_.m_7697_(), p_196851_.m_7387_(), p_196851_.m_188181_(), p_196851_.m_188182_(), p_196851_.m_6319_(), p_196851_.m_7103_(), p_196852_, p_196851_.m_183407_());
-
+      // Paper start - rewrite light engine
+      this.setBlockNibbles(p_196851_.getBlockNibbles());
+      this.setSkyNibbles(p_196851_.getSkyNibbles());
+      this.setSkyEmptinessMap(p_196851_.getSkyEmptinessMap());
+      this.setBlockEmptinessMap(p_196851_.getBlockEmptinessMap());
+      // Paper end - rewrite light engine
       for(BlockEntity blockentity : p_196851_.m_63292_().values()) {
          this.m_142169_(blockentity);
       }
@@ -147,7 +_,29 @@
       }
    }
 
+   // Paper start - Optimize getBlockData to reduce instructions
+
+   @Override
+   public final BlockState getBlockStateIfLoaded(BlockPos blockposition) {
+      return this.m_8055_(blockposition);
+   }
+   @Override
    public BlockState m_8055_(BlockPos p_62923_) {
+      return this.getBlockStateFinal(p_62923_.m_123341_(), p_62923_.m_123342_(), p_62923_.m_123343_());
+   }
+
+   public final BlockState getBlockStateFinal(final int x, final int y, final int z) {
+      // Method body / logic copied from below
+      final int i = this.m_151564_(y);
+      if (i < 0 || i >= this.f_187612_.length || this.f_187612_[i].f_62969_ == 0 || this.f_187612_[i].m_188008_()) {
+         return Blocks.f_50016_.m_49966_();
+      }
+      // Inlined ChunkSection.getType() and DataPaletteBlock.a(int,int,int)
+      return this.f_187612_[i].f_62972_.m_63085_((y & 15) << 8 | (z & 15) << 4 | x & 15);
+
+   }
+
+   public BlockState getBlockState_old(BlockPos p_62923_) {
       int i = p_62923_.m_123341_();
       int j = p_62923_.m_123342_();
       int k = p_62923_.m_123343_();
@@ -237,14 +_,14 @@
             boolean flag2 = blockstate.m_155947_();
             if (!this.f_62776_.f_46443_) {
                blockstate.m_60753_(this.f_62776_, p_62865_, p_62866_, p_62867_);
-            } else if (!blockstate.m_60713_(block) && flag2) {
+         } else if ((!blockstate.m_60713_(block) || !p_62866_.m_155947_()) && flag2) {
                this.m_8114_(p_62865_);
             }
 
             if (!levelchunksection.m_62982_(j, k, l).m_60713_(block)) {
                return null;
             } else {
-               if (!this.f_62776_.f_46443_) {
+            if (!this.f_62776_.f_46443_ && !this.f_62776_.captureBlockSnapshots) {
                   p_62866_.m_60696_(this.f_62776_, p_62865_, blockstate, p_62867_);
                }
 
@@ -287,6 +_,10 @@
    @Nullable
    public BlockEntity m_5685_(BlockPos p_62868_, LevelChunk.EntityCreationType p_62869_) {
       BlockEntity blockentity = this.f_187610_.get(p_62868_);
+      if (blockentity != null && blockentity.m_58901_()) {
+         f_187610_.remove(p_62868_);
+         blockentity = null;
+      }
       if (blockentity == null) {
          CompoundTag compoundtag = this.f_187609_.remove(p_62868_);
          if (compoundtag != null) {
@@ -304,9 +_,6 @@
                this.m_142170_(blockentity);
             }
          }
-      } else if (blockentity.m_58901_()) {
-         this.f_187610_.remove(p_62868_);
-         return null;
       }
 
       return blockentity;
@@ -322,6 +_,7 @@
          }
 
          this.m_156406_(p_156391_);
+         p_156391_.onLoad();
       }
 
    }
@@ -361,9 +_,14 @@
    public CompoundTag m_8051_(BlockPos p_62932_) {
       BlockEntity blockentity = this.m_7702_(p_62932_);
       if (blockentity != null && !blockentity.m_58901_()) {
+         try {
          CompoundTag compoundtag1 = blockentity.m_187480_();
          compoundtag1.m_128379_("keepPacked", false);
          return compoundtag1;
+         } catch (Exception e) {
+            f_62771_.error("A BlockEntity type {} has thrown an exception trying to write state. It will not persist, Report this to the mod author", blockentity.getClass().getName(), e);
+            return null;
+         }
       } else {
          CompoundTag compoundtag = this.f_187609_.get(p_62932_);
          if (compoundtag != null) {
@@ -378,6 +_,13 @@
    public void m_8114_(BlockPos p_62919_) {
       if (this.m_156370_()) {
          BlockEntity blockentity = this.f_187610_.remove(p_62919_);
+
+         // CraftBukkit start - SPIGOT-5561: Also remove from pending map
+         if (!f_187609_.isEmpty()) {
+            f_187609_.remove(p_62919_);
+         }
+         // CraftBukkit end
+
          if (blockentity != null) {
             Level level = this.f_62776_;
             if (level instanceof ServerLevel) {
@@ -424,6 +_,31 @@
 
    }
 
+
+   // Paper start - add dirty system to tick lists
+   @Override
+   public void m_8092_(boolean needsSaving) {
+      if (!needsSaving) {
+         this.f_62784_.clearDirty();
+         this.f_187943_.clearDirty();
+      }
+      super.m_8092_(needsSaving);
+   }
+   // Paper end - add dirty system to tick lists
+
+   @Override
+   public boolean m_6344_() {
+      // Paper start - add dirty system to tick lists
+      long gameTime = this.f_62776_.m_6106_().m_6793_();
+      if (this.f_62784_.isDirty(gameTime) || this.f_187943_.isDirty(gameTime)) {
+         return true;
+      }
+      // Paper end - add dirty system to tick lists
+      return super.m_6344_(); // Paper - rewrite chunk system - do NOT clobber the dirty flag
+   }
+   // CraftBukkit end
+
+
    public boolean m_6430_() {
       return false;
    }
@@ -445,7 +_,7 @@
       p_187974_.accept((p_187968_, p_187969_, p_187970_) -> {
          BlockEntity blockentity = this.m_5685_(p_187968_, LevelChunk.EntityCreationType.IMMEDIATE);
          if (blockentity != null && p_187970_ != null && blockentity.m_58903_() == p_187969_) {
-            blockentity.m_142466_(p_187970_);
+            blockentity.handleUpdateTag(p_187970_);
          }
 
       });
@@ -465,11 +_,13 @@
 
    public Stream<BlockPos> m_6267_() {
       return StreamSupport.stream(BlockPos.m_121976_(this.f_187604_.m_45604_(), this.m_141937_(), this.f_187604_.m_45605_(), this.f_187604_.m_45608_(), this.m_151558_() - 1, this.f_187604_.m_45609_()).spliterator(), false).filter((p_187990_) -> {
-         return this.m_8055_(p_187990_).m_60791_() != 0;
+         return this.m_8055_(p_187990_).getLightEmission(m_62953_(), p_187990_) != 0;
       });
    }
 
+   public boolean isPostProcessingDone; // Paper - replace chunk loader system
    public void m_62812_() {
+      try { // Paper - replace chunk loader system
       ChunkPos chunkpos = this.m_7697_();
 
       for(int i = 0; i < this.f_187602_.length; ++i) {
@@ -498,6 +_,11 @@
 
       this.f_187609_.clear();
       this.f_187606_.m_63341_(this);
+      } finally { // Paper start - replace chunk loader system
+         this.isPostProcessingDone = true;
+         ((ServerChunkCache) this.f_62776_.m_7726_()).f_8325_.playerChunkManager.onChunkPostProcessing(this.f_187604_.f_45578_, this.f_187604_.f_45579_);
+      }
+      // Paper end - replace chunk loader system
    }
 
    @Nullable
@@ -545,7 +_,7 @@
    }
 
    public ChunkHolder.FullChunkStatus m_6708_() {
-      return this.f_62790_ == null ? ChunkHolder.FullChunkStatus.BORDER : this.f_62790_.get();
+      return this.chunkHolder == null ? ChunkHolder.FullChunkStatus.INACCESSIBLE : this.chunkHolder.getChunkStatus(); // Paper - rewrite chunk system
    }
 
    public void m_62879_(Supplier<ChunkHolder.FullChunkStatus> p_62880_) {
@@ -553,6 +_,7 @@
    }
 
    public void m_187957_() {
+      this.f_187610_.values().forEach(BlockEntity::onChunkUnloaded);
       this.f_187610_.values().forEach(BlockEntity::m_7651_);
       this.f_187610_.clear();
       this.f_156362_.values().forEach((p_187966_) -> {
@@ -562,6 +_,7 @@
    }
 
    public void m_156369_() {
+      this.f_62776_.addFreshBlockEntities(this.f_187610_.values());
       this.f_187610_.values().forEach((p_187988_) -> {
          Level level = this.f_62776_;
          if (level instanceof ServerLevel serverlevel) {
@@ -619,6 +_,41 @@
       this.f_196848_ = p_196865_;
    }
 
+   // FORGE START
+   private final net.minecraftforge.common.capabilities.CapabilityProvider.AsField<LevelChunk> capProvider = new net.minecraftforge.common.capabilities.CapabilityProvider.AsField<>(LevelChunk.class, this);
+
+   @org.jetbrains.annotations.NotNull
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(@org.jetbrains.annotations.NotNull net.minecraftforge.common.capabilities.Capability<T> cap, @org.jetbrains.annotations.Nullable net.minecraft.core.Direction side)
+   {
+      return capProvider.getCapability(cap, side);
+   }
+
+   @Override
+   public boolean areCapsCompatible(net.minecraftforge.common.capabilities.CapabilityProvider<LevelChunk> other)
+   {
+      return capProvider.areCapsCompatible(other);
+   }
+
+   @Override
+   public boolean areCapsCompatible(@org.jetbrains.annotations.Nullable net.minecraftforge.common.capabilities.CapabilityDispatcher other)
+   {
+      return capProvider.areCapsCompatible(other);
+   }
+
+   @Override
+   public void invalidateCaps()
+   {
+      capProvider.invalidateCaps();
+   }
+
+   @Override
+   public void reviveCaps()
+   {
+      capProvider.reviveCaps();
+   }
+   // FORGE END
+
    class BoundTickingBlockEntity<T extends BlockEntity> implements TickingBlockEntity {
       private final T f_156428_;
       private final BlockEntityTicker<T> f_156429_;
@@ -635,6 +_,7 @@
             if (LevelChunk.this.m_156410_(blockpos)) {
                try {
                   ProfilerFiller profilerfiller = LevelChunk.this.f_62776_.m_46473_();
+                  net.minecraftforge.server.timings.TimeTracker.BLOCK_ENTITY_UPDATE.trackStart(f_156428_);
                   profilerfiller.m_6521_(this::m_142280_);
                   BlockState blockstate = LevelChunk.this.m_8055_(blockpos);
                   if (this.f_156428_.m_58903_().m_155262_(blockstate)) {
@@ -650,6 +_,12 @@
                   CrashReport crashreport = CrashReport.m_127521_(throwable, "Ticking block entity");
                   CrashReportCategory crashreportcategory = crashreport.m_127514_("Block entity being ticked");
                   this.f_156428_.m_58886_(crashreportcategory);
+
+                  if (net.minecraftforge.common.ForgeConfig.SERVER.removeErroringBlockEntities.get()) {
+                     f_62771_.error("{}", crashreport.m_127526_());
+                     f_156428_.m_7651_();
+                     LevelChunk.this.m_8114_(f_156428_.m_58899_());
+                  } else
                   throw new ReportedException(crashreport);
                }
             }
@@ -678,6 +_,34 @@
       IMMEDIATE,
       QUEUED,
       CHECK;
+   }
+
+
+   /**
+    * <strong>FOR INTERNAL USE ONLY</strong>
+    * <p>
+    * Only public for use in {@link net.minecraft.world.level.chunk.storage.ChunkSerializer}.
+    */
+   @java.lang.Deprecated
+   @org.jetbrains.annotations.Nullable
+   public final CompoundTag writeCapsToNBT() {
+      return capProvider.serializeInternal();
+   }
+
+   /**
+    * <strong>FOR INTERNAL USE ONLY</strong>
+    * <p>
+    * Only public for use in {@link net.minecraft.world.level.chunk.storage.ChunkSerializer}.
+    *
+    */
+   @java.lang.Deprecated
+   public final void readCapsFromNBT(CompoundTag tag) {
+      capProvider.deserializeInternal(tag);
+   }
+
+   @Override
+   public Level getWorldForge() {
+      return m_62953_();
    }
 
    @FunctionalInterface
