--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -23,7 +_,7 @@
    private short f_62969_;
    private short f_62970_;
    private short f_62971_;
-   private final PalettedContainer<BlockState> f_62972_;
+   public final PalettedContainer<BlockState> f_62972_;
    private PalettedContainerRO<Holder<Biome>> f_187995_;
 
    public LevelChunkSection(int p_238255_, PalettedContainer<BlockState> p_238256_, PalettedContainerRO<Holder<Biome>> p_238257_) {
