--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -49,12 +_,21 @@
    private final ProtoChunkTicks<Block> f_63163_;
    private final ProtoChunkTicks<Fluid> f_188165_;
 
+   public CompoundTag poiData;
+   public java.util.ArrayDeque<Runnable> tasks;
+
    public ProtoChunk(ChunkPos p_188167_, UpgradeData p_188168_, LevelHeightAccessor p_188169_, Registry<Biome> p_188170_, @Nullable BlendingData p_188171_) {
       this(p_188167_, p_188168_, (LevelChunkSection[])null, new ProtoChunkTicks<>(), new ProtoChunkTicks<>(), p_188169_, p_188170_, p_188171_);
    }
 
    public ProtoChunk(ChunkPos p_188173_, UpgradeData p_188174_, @Nullable LevelChunkSection[] p_188175_, ProtoChunkTicks<Block> p_188176_, ProtoChunkTicks<Fluid> p_188177_, LevelHeightAccessor p_188178_, Registry<Biome> p_188179_, @Nullable BlendingData p_188180_) {
       super(p_188173_, p_188174_, p_188178_, p_188179_, 0L, p_188175_, p_188180_);
+      // Paper start - rewrite light engine
+      if (!(this instanceof ImposterProtoChunk)) {
+         this.setBlockNibbles(ca.spottedleaf.starlight.common.light.StarLightEngine.getFilledEmptyLight(p_188178_));
+         this.setSkyNibbles(ca.spottedleaf.starlight.common.light.StarLightEngine.getFilledEmptyLight(p_188178_));
+      }
+      // Paper end - rewrite light engine
       this.f_63163_ = p_188176_;
       this.f_188165_ = p_188177_;
    }
@@ -71,6 +_,11 @@
       return new ChunkAccess.TicksToSave(this.f_63163_, this.f_188165_);
    }
 
+   @Override
+   public final BlockState getBlockStateIfLoaded(BlockPos blockposition) {
+      return this.m_8055_(blockposition);
+   }
+
    public BlockState m_8055_(BlockPos p_63264_) {
       int i = p_63264_.m_123342_();
       if (this.m_151562_(i)) {
@@ -123,13 +_,13 @@
          if (this.f_187612_[l].m_188008_() && p_63218_.m_60713_(Blocks.f_50016_)) {
             return p_63218_;
          } else {
-            if (p_63218_.m_60791_() > 0) {
+            if (p_63218_.getLightEmission(this, p_63217_) > 0) {
                this.f_63158_.add(new BlockPos((i & 15) + this.m_7697_().m_45604_(), j, (k & 15) + this.m_7697_().m_45605_()));
             }
 
             LevelChunkSection levelchunksection = this.m_183278_(l);
             BlockState blockstate = levelchunksection.m_62986_(i & 15, j & 15, k & 15, p_63218_);
-            if (this.f_63153_.m_62427_(ChunkStatus.f_62322_) && p_63218_ != blockstate && (p_63218_.m_60739_(this, p_63217_) != blockstate.m_60739_(this, p_63217_) || p_63218_.m_60791_() != blockstate.m_60791_() || p_63218_.m_60787_() || blockstate.m_60787_())) {
+            if (this.f_63153_.m_62427_(ChunkStatus.f_62322_) && p_63218_ != blockstate && (p_63218_.m_60739_(this, p_63217_) != blockstate.m_60739_(this, p_63217_) || p_63218_.getLightEmission(this, p_63217_) != blockstate.getLightEmission(this, p_63217_) || p_63218_.m_60787_() || blockstate.m_60787_())) {
                this.f_63151_.m_7174_(p_63217_);
             }
 
