--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -53,6 +_,12 @@
 
    public ProtoChunk(ChunkPos p_188173_, UpgradeData p_188174_, @Nullable LevelChunkSection[] p_188175_, ProtoChunkTicks<Block> p_188176_, ProtoChunkTicks<Fluid> p_188177_, LevelHeightAccessor p_188178_, Registry<Biome> p_188179_, @Nullable BlendingData p_188180_) {
       super(p_188173_, p_188174_, p_188178_, p_188179_, 0L, p_188175_, p_188180_);
+      // Paper start - rewrite light engine
+      if (!(this instanceof ImposterProtoChunk)) {
+         this.setBlockNibbles(ca.spottedleaf.starlight.common.light.StarLightEngine.getFilledEmptyLight(p_188178_));
+         this.setSkyNibbles(ca.spottedleaf.starlight.common.light.StarLightEngine.getFilledEmptyLight(p_188178_));
+      }
+      // Paper end - rewrite light engine
       this.blockTicks = p_188176_;
       this.fluidTicks = p_188177_;
    }
