--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -62,11 +_,26 @@
 
 public class Main {
     private static final Logger LOGGER = LogUtils.getLogger();
+    public static final java.time.Instant BOOT_TIME = java.time.Instant.now(); // Paper - track initial start time
 
     @DontObfuscate
     public static void main(String[] p_129699_) throws Exception {
         SharedConstants.tryDetectVersion();
-        OptionParser optionparser = new OptionParser();
+        org.goldenforge.GoldenForgeEntryPoint.init();
+        OptionParser optionparser = new OptionParser() {
+            {
+                acceptsAll(java.util.Arrays.asList("paper-dir", "paper-settings-directory"), "Directory for Paper settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File(io.papermc.paper.configuration.PaperConfigurations.CONFIG_DIR))
+                        .describedAs("Config directory");
+                acceptsAll(java.util.Arrays.asList("paper", "paper-settings"), "File for Paper settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("paper.yml"))
+                        .describedAs("Yml file");
+            }
+        };
         OptionSpec<Void> optionspec = optionparser.accepts("nogui");
         OptionSpec<Void> optionspec1 = optionparser.accepts("initSettings", "Initializes 'server.properties' and 'eula.txt', then quits");
         OptionSpec<Void> optionspec2 = optionparser.accepts("demo");
@@ -131,7 +_,7 @@
             }
 
             File file1 = new File(optionset.valueOf(optionspec9));
-            Services services = Services.create(new YggdrasilAuthenticationService(Proxy.NO_PROXY), file1);
+            Services services = Services.create(new YggdrasilAuthenticationService(Proxy.NO_PROXY), file1, new java.io.File("."), optionset);
             String s = Optional.ofNullable(optionset.valueOf(optionspec10)).orElse(dedicatedserversettings.getProperties().levelName);
             if (s == null || s.isEmpty() || new File(file1, s).getAbsolutePath().equals(new File(s).getAbsolutePath())) {
                 LOGGER.error("Invalid world directory specified, must not be null, empty or the same directory as your universe! " + s);
@@ -297,15 +_,6 @@
                     return dedicatedserver1;
                 }
             );
-            Thread thread = new Thread("Server Shutdown Thread") {
-                @Override
-                public void run() {
-                    dedicatedserver.halt(true);
-                    org.apache.logging.log4j.LogManager.shutdown(); // we're manually managing the logging shutdown on the server. Make sure we do it here at the end.
-                }
-            };
-            thread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
-            Runtime.getRuntime().addShutdownHook(thread);
         } catch (Exception exception1) {
             LOGGER.error(LogUtils.FATAL_MARKER, "Failed to start the minecraft server", (Throwable)exception1);
         }
