--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -3,6 +_,8 @@
 import com.google.common.collect.Sets;
 import java.util.Set;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.nbt.CompoundTag;
@@ -522,6 +_,13 @@
         protected boolean isValidTarget(LevelReader p_33963_, BlockPos p_33964_) {
             return p_33963_.getBlockState(p_33964_).is(Blocks.LAVA) && p_33963_.getBlockState(p_33964_.above()).isPathfindable(PathComputationType.LAND);
         }
+
+        // Leaf start - Async search block
+        @Override
+        protected TypeToCheck typeToCheck() {
+            return TypeToCheck.Strider;
+        }
+        // Leaf end - Async search block
     }
 
     static class StriderPathNavigation extends GroundPathNavigation {
@@ -529,10 +_,26 @@
             super(p_33969_, p_33970_);
         }
 
+        // Kaiiju start - petal - async path processing
+        private static final org.dreeam.leaf.async.path.NodeEvaluatorGenerator nodeEvaluatorGenerator = (org.dreeam.leaf.async.path.NodeEvaluatorFeatures nodeEvaluatorFeatures) -> {
+            WalkNodeEvaluator nodeEvaluator = new WalkNodeEvaluator();
+            nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+            nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+            nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+            nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+            return nodeEvaluator;
+        };
+        // Kaiiju end
+
         @Override
         protected PathFinder createPathFinder(int p_33972_) {
             this.nodeEvaluator = new WalkNodeEvaluator();
             this.nodeEvaluator.setCanPassDoors(true);
+            // Kaiiju start - async path processing
+            if (GlobalConfiguration.get().asyncPathFinding.enabled) {
+                return new PathFinder(this.nodeEvaluator, p_33972_, nodeEvaluatorGenerator);
+            }
+            // Kaiiju end
             return new PathFinder(this.nodeEvaluator, p_33972_);
         }
 
