--- a/net/minecraft/world/entity/AgeableMob.java
+++ b/net/minecraft/world/entity/AgeableMob.java
@@ -23,6 +_,31 @@
       super(p_146738_, p_146739_);
    }
 
+   // Spigot start
+   @Override
+   public void inactiveTick()
+   {
+      super.inactiveTick();
+      if ( this.f_19853_.f_46443_ )
+      {
+         this.m_6210_();
+      } else
+      {
+         int i = this.m_146764_();
+
+         if ( i < 0 )
+         {
+            ++i;
+            this.m_146762_( i );
+         } else if ( i > 0 )
+         {
+            --i;
+            this.m_146762_( i );
+         }
+      }
+   }
+   // Spigot end
+
    public SpawnGroupData m_6518_(ServerLevelAccessor p_146746_, DifficultyInstance p_146747_, MobSpawnType p_146748_, @Nullable SpawnGroupData p_146749_, @Nullable CompoundTag p_146750_) {
       if (p_146749_ == null) {
          p_146749_ = new AgeableMob.AgeableMobGroupData(true);
