--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -148,6 +_,18 @@
       this.f_19687_ = p_19735_;
    }
 
+   // Spigot start - copied from below
+   @Override
+   public void inactiveTick() {
+      super.inactiveTick();
+
+      if (this.f_19797_ >= this.f_19688_ + this.f_19687_) {
+         this.m_146870_();
+         return;
+      }
+   }
+   // Spigot end
+
    public void m_8119_() {
       super.m_8119_();
       boolean flag = this.m_19747_();
