--- a/net/minecraft/Util.java
+++ b/net/minecraft/Util.java
@@ -89,9 +_,25 @@
     private static final int DEFAULT_MAX_THREADS = 255;
     private static final int DEFAULT_SAFE_FILE_OPERATION_RETRIES = 10;
     private static final String MAX_THREADS_SYSTEM_PROPERTY = "max.bg.threads";
-    private static final ExecutorService BACKGROUND_EXECUTOR = makeExecutor("Main");
+    private static final ExecutorService BACKGROUND_EXECUTOR = makeExecutor("Main", -1); // Paper - Perf: add priority
     private static final ExecutorService IO_POOL = makeIoExecutor("IO-Worker-", false);
     private static final ExecutorService DOWNLOAD_POOL = makeIoExecutor("Download-", true);
+    // Paper start - don't submit BLOCKING PROFILE LOOKUPS to the world gen thread
+    public static final ExecutorService PROFILE_EXECUTOR = Executors.newFixedThreadPool(2, new java.util.concurrent.ThreadFactory() {
+
+        private final AtomicInteger count = new AtomicInteger();
+
+        @Override
+        public Thread newThread(Runnable run) {
+            Thread ret = new Thread(run);
+            ret.setName("Profile Lookup Executor #" + this.count.getAndIncrement());
+            ret.setUncaughtExceptionHandler((Thread thread, Throwable throwable) -> {
+                LOGGER.error("Uncaught exception in thread " + thread.getName(), throwable);
+            });
+            return ret;
+        }
+    });
+    // Paper end - don't submit BLOCKING PROFILE LOOKUPS to the world gen thread
     private static final DateTimeFormatter FILENAME_DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd_HH.mm.ss", Locale.ROOT);
     public static final int LINEAR_LOOKUP_THRESHOLD = 8;
     private static final Set<String> ALLOWED_UNTRUSTED_LINK_PROTOCOLS = Set.of("http", "https");
@@ -135,7 +_,7 @@
     }
 
     public static long getNanos() {
-        return timeSource.getAsLong();
+        return System.nanoTime(); // Paper
     }
 
     public static long getEpochMillis() {
@@ -146,32 +_,52 @@
         return FILENAME_DATE_TIME_FORMATTER.format(ZonedDateTime.now());
     }
 
-    private static ExecutorService makeExecutor(String p_137478_) {
-        int i = Mth.clamp(Runtime.getRuntime().availableProcessors() - 1, 1, getMaxThreads());
-        ExecutorService executorservice;
+    private static ExecutorService makeExecutor(String name, final int priorityModifier) { // Paper - Perf: add priority
+        int i = maxAllowedExecutorThreads();
+        // Paper start - Perf: use simpler thread pool that allows 1 thread and reduce worldgen thread worker count for low core count CPUs
+        final ExecutorService directExecutorService;
         if (i <= 0) {
-            executorservice = MoreExecutors.newDirectExecutorService();
+            directExecutorService = MoreExecutors.newDirectExecutorService();
         } else {
-            AtomicInteger atomicinteger = new AtomicInteger(1);
-            executorservice = new ForkJoinPool(i, p_314383_ -> {
-                ForkJoinWorkerThread forkjoinworkerthread = new ForkJoinWorkerThread(p_314383_) {
+            AtomicInteger atomicInteger = new AtomicInteger(1);
+            directExecutorService = new ForkJoinPool(i, forkJoinPool -> {
+                final String string = "Worker-" + name + "-" + atomicInteger.getAndIncrement();
+                ForkJoinWorkerThread forkJoinWorkerThread = new ForkJoinWorkerThread(forkJoinPool) {
                     @Override
-                    protected void onTermination(Throwable p_211561_) {
-                        if (p_211561_ != null) {
-                            Util.LOGGER.warn("{} died", this.getName(), p_211561_);
+                    protected void onTermination(Throwable throwOnTermination) {
+                        if (throwOnTermination != null) {
+                            Util.LOGGER.warn("{} died", this.getName(), throwOnTermination);
                         } else {
                             Util.LOGGER.debug("{} shutdown", this.getName());
                         }
 
-                        super.onTermination(p_211561_);
+                        super.onTermination(throwOnTermination);
                     }
                 };
-                forkjoinworkerthread.setName("Worker-" + p_137478_ + "-" + atomicinteger.getAndIncrement());
-                return forkjoinworkerthread;
-            }, Util::onThreadException, true);
-        }
-
-        return executorservice;
+                forkJoinWorkerThread.setPriority(Thread.NORM_PRIORITY + priorityModifier); // Paper - Deprioritize over main
+                forkJoinWorkerThread.setName(string);
+                return forkJoinWorkerThread;
+            }, Util::onThreadException, true, 0, Integer.MAX_VALUE, 1, null, 365, TimeUnit.DAYS); // Paper - do not expire threads
+        }
+
+        return directExecutorService;
+    }
+
+    public static int maxAllowedExecutorThreads() {
+        // Paper start - Perf: use simpler thread pool that allows 1 thread and reduce worldgen thread worker count for low core count CPUs
+        final int cpus = Runtime.getRuntime().availableProcessors() / 2;
+        int maxExecutorThreads;
+        if (cpus <= 4) {
+            maxExecutorThreads = cpus <= 2 ? 1 : 2;
+        } else if (cpus <= 8) {
+            // [5, 8]
+            maxExecutorThreads = Math.max(3, cpus - 2);
+        } else {
+            maxExecutorThreads = cpus * 2 / 3;
+        }
+        maxExecutorThreads = Math.min(8, maxExecutorThreads);
+        return Integer.getInteger("Paper.WorkerThreadCount", maxExecutorThreads);
+        // Paper end - Perf: use simpler thread pool that allows 1 thread and reduce worldgen thread worker count for low core count CPUs
     }
 
     private static int getMaxThreads() {
@@ -239,7 +_,7 @@
         throw p_137560_ instanceof RuntimeException ? (RuntimeException)p_137560_ : new RuntimeException(p_137560_);
     }
 
-    private static void onThreadException(Thread p_137496_, Throwable p_137497_) {
+    public static void onThreadException(Thread p_137496_, Throwable p_137497_) {
         pauseInIde(p_137497_);
         if (p_137497_ instanceof CompletionException) {
             p_137497_ = p_137497_.getCause();
@@ -267,8 +_,8 @@
                 .getSchema(DataFixUtils.makeKey(SharedConstants.getCurrentVersion().getDataVersion().getVersion()))
                 .getChoiceType(p_137552_, p_137553_);
         } catch (IllegalArgumentException illegalargumentexception) {
-            LOGGER.error("No data fixer registered for {}", p_137553_);
-            if (SharedConstants.IS_RUNNING_IN_IDE) {
+            LOGGER.debug("No data fixer registered for {}", p_137553_);
+            if (SharedConstants.IS_RUNNING_IN_IDE && false) {
                 throw illegalargumentexception;
             }
         }
@@ -473,6 +_,28 @@
         }
     }
 
+    //Goldenforge: from leaf
+    public static <V> CompletableFuture<List<V>> faster_sequence(List<? extends CompletableFuture<V>> p_137568_) {
+        if (p_137568_.isEmpty()) {
+            return CompletableFuture.completedFuture(List.of());
+        } else if (p_137568_.size() == 1) {
+            return p_137568_.get(0).thenApply(List::of);
+        } else {
+            CompletableFuture<Void> completablefuture = CompletableFuture.allOf(p_137568_.toArray(new CompletableFuture[0]));
+            return completablefuture.thenCompose(void_ ->
+                    CompletableFuture.supplyAsync(() -> {
+                                List<V> list = new java.util.ArrayList<>();
+
+                                for (CompletableFuture<V> future : p_137568_) {
+                                    list.add(future.join());
+                                }
+
+                                return list;
+                            }
+                    ));
+        }
+    }
+
     public static <V> CompletableFuture<List<V>> sequenceFailFast(List<? extends CompletableFuture<? extends V>> p_143841_) {
         CompletableFuture<List<V>> completablefuture = new CompletableFuture<>();
         return fallibleSequence(p_143841_, completablefuture::completeExceptionally).applyToEither(completablefuture, Function.identity());
@@ -526,20 +_,20 @@
 
     public static void logAndPauseIfInIde(String p_143786_) {
         LOGGER.error(p_143786_);
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             doPause(p_143786_);
         }
     }
 
     public static void logAndPauseIfInIde(String p_200891_, Throwable p_200892_) {
         LOGGER.error(p_200891_, p_200892_);
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             doPause(p_200891_);
         }
     }
 
     public static <T extends Throwable> T pauseInIde(T p_137571_) {
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             LOGGER.error("Trying to throw a fatal exception, pausing in IDE", p_137571_);
             doPause(p_137571_.getMessage());
         }
