--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -28,6 +_,7 @@
 import net.minecraft.network.protocol.game.ClientboundUpdateAttributesPacket;
 import net.minecraft.network.protocol.game.VecDeltaCodec;
 import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.server.network.ServerPlayerConnection;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EquipmentSlot;
@@ -62,13 +_,17 @@
     private Vec3 lastSentMovement;
     private int tickCount;
     private int teleportDelay;
-    private List<Entity> lastPassengers = Collections.emptyList();
+    private List<Entity> lastPassengers = com.google.common.collect.ImmutableList.of(); // Paper - optimize passenger checks
     private boolean wasRiding;
     private boolean wasOnGround;
     @Nullable
     private List<SynchedEntityData.DataValue<?>> trackedDataValues;
+    // Goldenforge start
+    private final Set<ServerPlayerConnection> trackedPlayers;
 
-    public ServerEntity(ServerLevel p_8528_, Entity p_8529_, int p_8530_, boolean p_8531_, Consumer<Packet<?>> p_8532_) {
+    public ServerEntity(ServerLevel p_8528_, Entity p_8529_, int p_8530_, boolean p_8531_, Consumer<Packet<?>> p_8532_, Set<ServerPlayerConnection> trackedPlayers) {
+        this.trackedPlayers = trackedPlayers;
+        // Goldenforge end
         this.level = p_8528_;
         this.broadcast = p_8532_;
         this.entity = p_8529_;
@@ -83,7 +_,19 @@
         this.trackedDataValues = p_8529_.getEntityData().getNonDefaultValues();
     }
 
+    // Paper start - fix desync when a player is added to the tracker
+    private boolean forceStateResync;
+    public void onPlayerAdd() {
+        this.forceStateResync = true;
+    }
+    // Paper end - fix desync when a player is added to the tracker
+
     public void sendChanges() {
+        // Paper start - optimise collisions
+        if (((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this.entity).moonrise$isHardColliding()) {
+            this.teleportDelay = 9999;
+        }
+        // Paper end - optimise collisions
         List<Entity> list = this.entity.getPassengers();
         if (!list.equals(this.lastPassengers)) {
             this.broadcast.accept(new ClientboundSetPassengersPacket(this.entity));
@@ -99,26 +_,29 @@
             this.lastPassengers = list;
         }
 
-        if (this.entity instanceof ItemFrame itemframe && this.tickCount % 10 == 0) {
-            ItemStack itemstack = itemframe.getItem();
-            if (itemstack.getItem() instanceof MapItem) {
-                MapId mapid = itemstack.get(DataComponents.MAP_ID);
-                MapItemSavedData mapitemsaveddata = MapItem.getSavedData(mapid, this.level);
-                if (mapitemsaveddata != null) {
-                    for (ServerPlayer serverplayer : this.level.players()) {
-                        mapitemsaveddata.tickCarriedBy(serverplayer, itemstack);
-                        Packet<?> packet = mapitemsaveddata.getUpdatePacket(mapid, serverplayer);
-                        if (packet != null) {
-                            serverplayer.connection.send(packet);
+        if (!this.trackedPlayers.isEmpty() && entity instanceof ItemFrame entityitemframe) { // Paper - Perf: Only tick item frames if players can see it
+            if (true && this.tickCount % 10 == 0) {
+                ItemStack itemstack = entityitemframe.getItem();
+                if (true) {
+                    MapId mapid = entityitemframe.cachedMapId; // Paper - Perf: Cache map ids on item frames
+                    MapItemSavedData mapitemsaveddata = MapItem.getSavedData(itemstack, this.level);
+                    if (mapitemsaveddata != null) {
+                        for (ServerPlayerConnection connection : this.trackedPlayers) { // Paper
+                            final ServerPlayer serverplayer = connection.getPlayer(); // Paper
+                            mapitemsaveddata.tickCarriedBy(serverplayer, itemstack);
+                            Packet<?> packet = mapitemsaveddata.getUpdatePacket(mapid, serverplayer);
+                            if (packet != null) {
+                                serverplayer.connection.send(packet);
+                            }
                         }
                     }
                 }
+
+                this.sendDirtyEntityData();
             }
-
-            this.sendDirtyEntityData();
         }
 
-        if (this.tickCount % this.updateInterval == 0 || this.entity.hasImpulse || this.entity.getEntityData().isDirty()) {
+        if (this.forceStateResync || this.tickCount % this.updateInterval == 0 || this.entity.hasImpulse || this.entity.getEntityData().isDirty()) { // Paper - fix desync when a player is added to the tracker
             if (this.entity.isPassenger()) {
                 int i1 = Mth.floor(this.entity.getYRot() * 256.0F / 360.0F);
                 int l1 = Mth.floor(this.entity.getXRot() * 256.0F / 360.0F);
@@ -147,7 +_,7 @@
                 long j = this.positionCodec.encodeY(vec31);
                 long k = this.positionCodec.encodeZ(vec31);
                 boolean flag2 = i < -32768L || i > 32767L || j < -32768L || j > 32767L || k < -32768L || k > 32767L;
-                if (flag2 || this.teleportDelay > 400 || this.wasRiding || this.wasOnGround != this.entity.onGround()) {
+                if (!this.forceStateResync && flag2 || this.teleportDelay > 400 || this.wasRiding || this.wasOnGround != this.entity.onGround()) { // Paper - fix desync when a player is added to the tracker
                     this.wasOnGround = this.entity.onGround();
                     this.teleportDelay = 0;
                     packet1 = new ClientboundTeleportEntityPacket(this.entity);
@@ -217,6 +_,7 @@
             }
 
             this.entity.hasImpulse = false;
+            this.forceStateResync = false; // Paper - fix desync when a player is added to the tracker
         }
 
         this.tickCount++;
@@ -233,22 +_,25 @@
     public void removePairing(ServerPlayer p_8535_) {
         this.entity.stopSeenByPlayer(p_8535_);
         p_8535_.connection.send(new ClientboundRemoveEntitiesPacket(this.entity.getId()));
+        net.neoforged.neoforge.event.EventHooks.onStopEntityTracking(this.entity, p_8535_);
     }
 
     public void addPairing(ServerPlayer p_8542_) {
         List<Packet<? super ClientGamePacketListener>> list = new ArrayList<>();
-        this.sendPairingData(p_8542_, list::add);
+        this.sendPairingData(p_8542_, new net.neoforged.neoforge.network.bundle.PacketAndPayloadAcceptor<>(list::add));
         p_8542_.connection.send(new ClientboundBundlePacket(list));
         this.entity.startSeenByPlayer(p_8542_);
+        net.neoforged.neoforge.event.EventHooks.onStartEntityTracking(this.entity, p_8542_);
     }
 
-    public void sendPairingData(ServerPlayer p_289562_, Consumer<Packet<ClientGamePacketListener>> p_289563_) {
+    public void sendPairingData(ServerPlayer p_289562_, net.neoforged.neoforge.network.bundle.PacketAndPayloadAcceptor<net.minecraft.network.protocol.game.ClientGamePacketListener> p_289563_) {
         if (this.entity.isRemoved()) {
             LOGGER.warn("Fetching packet for removed entity {}", this.entity);
         }
 
         Packet<ClientGamePacketListener> packet = this.entity.getAddEntityPacket(this);
         p_289563_.accept(packet);
+        this.entity.sendPairingData(p_289562_, p_289563_::accept);
         if (this.trackedDataValues != null) {
             p_289563_.accept(new ClientboundSetEntityDataPacket(this.entity.getId(), this.trackedDataValues));
         }
