--- a/net/minecraft/world/entity/animal/Bee.java
+++ b/net/minecraft/world/entity/animal/Bee.java
@@ -486,6 +_,7 @@
         } else if (this.isTooFarAway(this.hivePos)) {
             return false;
         } else {
+            if (this.level().getChunkIfLoadedImmediately(this.hivePos.getX() >> 4, this.hivePos.getZ() >> 4) == null) return true; // Paper - just assume the hive is still there, no need to load the chunk(s)
             BlockEntity blockentity = this.level().getBlockEntity(this.hivePos);
             return blockentity instanceof BeehiveBlockEntity;
         }
