--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -48,6 +_,7 @@
     private static final float HEIGHT = 0.75F;
     private float dropChance = 1.0F;
     private boolean fixed;
+    public @Nullable MapId cachedMapId; // Paper - Perf: Cache map ids on item frames
 
     public ItemFrame(EntityType<? extends ItemFrame> p_31761_, Level p_31762_) {
         super(p_31761_, p_31762_);
@@ -272,6 +_,7 @@
     }
 
     private void onItemChanged(ItemStack p_218866_) {
+        this.cachedMapId = p_218866_.getComponents().get(net.minecraft.core.component.DataComponents.MAP_ID); // Paper - Perf: Cache map ids on item frames
         if (!p_218866_.isEmpty() && p_218866_.getFrame() != this) {
             p_218866_.setEntityRepresentation(this);
         }
