--- a/net/minecraft/world/phys/shapes/BitSetDiscreteVoxelShape.java
+++ b/net/minecraft/world/phys/shapes/BitSetDiscreteVoxelShape.java
@@ -4,13 +_,13 @@
 import net.minecraft.core.Direction;
 
 public final class BitSetDiscreteVoxelShape extends DiscreteVoxelShape {
-    private final BitSet storage;
-    private int xMin;
-    private int yMin;
-    private int zMin;
-    private int xMax;
-    private int yMax;
-    private int zMax;
+    public final BitSet storage;
+    public int xMin;
+    public int yMin;
+    public int zMin;
+    public int xMax;
+    public int yMax;
+    public int zMax;
 
     public BitSetDiscreteVoxelShape(int p_82588_, int p_82589_, int p_82590_) {
         super(p_82588_, p_82589_, p_82590_);
