--- a/net/minecraft/world/phys/shapes/CubeVoxelShape.java
+++ b/net/minecraft/world/phys/shapes/CubeVoxelShape.java
@@ -7,6 +_,7 @@
 public final class CubeVoxelShape extends VoxelShape {
     protected CubeVoxelShape(DiscreteVoxelShape p_82765_) {
         super(p_82765_);
+        ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)this).moonrise$initCache(); // Paper - optimise collisions
     }
 
     @Override
