--- a/net/minecraft/world/phys/shapes/Shapes.java
+++ b/net/minecraft/world/phys/shapes/Shapes.java
@@ -16,9 +_,15 @@
     public static final double EPSILON = 1.0E-7;
     public static final double BIG_EPSILON = 1.0E-6;
     private static final VoxelShape BLOCK = Util.make(() -> {
-        DiscreteVoxelShape discretevoxelshape = new BitSetDiscreteVoxelShape(1, 1, 1);
-        discretevoxelshape.fill(0, 0, 0);
-        return new CubeVoxelShape(discretevoxelshape);
+        // Paper start - optimise collisions
+        final DiscreteVoxelShape shape = new BitSetDiscreteVoxelShape(1, 1, 1);
+        shape.fill(0, 0, 0);
+
+        return new ArrayVoxelShape(
+                shape,
+                ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.ZERO_ONE, ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.ZERO_ONE, ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.ZERO_ONE
+        );
+        // Paper end - optimise collisions
     });
     public static final VoxelShape INFINITY = box(
         Double.NEGATIVE_INFINITY,
@@ -43,6 +_,30 @@
         return BLOCK;
     }
 
+    // Paper start - optimise collisions
+    private static final DoubleArrayList[] PARTS_BY_BITS = new DoubleArrayList[] {
+            DoubleArrayList.wrap(generateCubeParts(1 << 0)),
+            DoubleArrayList.wrap(generateCubeParts(1 << 1)),
+            DoubleArrayList.wrap(generateCubeParts(1 << 2)),
+            DoubleArrayList.wrap(generateCubeParts(1 << 3))
+    };
+
+    private static double[] generateCubeParts(final int parts) {
+        // note: parts is a power of two, so we do not need to worry about loss of precision here
+        // note: parts is from [2^0, 2^3]
+        final double inc = 1.0 / (double)parts;
+
+        final double[] ret = new double[parts + 1];
+        double val = 0.0;
+        for (int i = 0; i <= parts; ++i) {
+            ret[i] = val;
+            val += inc;
+        }
+
+        return ret;
+    }
+    // Paper end - optimise collisions
+
     public static VoxelShape box(double p_83049_, double p_83050_, double p_83051_, double p_83052_, double p_83053_, double p_83054_) {
         if (!(p_83049_ > p_83052_) && !(p_83050_ > p_83053_) && !(p_83051_ > p_83054_)) {
             return create(p_83049_, p_83050_, p_83051_, p_83052_, p_83053_, p_83054_);
@@ -51,40 +_,43 @@
         }
     }
 
-    public static VoxelShape create(double p_166050_, double p_166051_, double p_166052_, double p_166053_, double p_166054_, double p_166055_) {
-        if (!(p_166053_ - p_166050_ < 1.0E-7) && !(p_166054_ - p_166051_ < 1.0E-7) && !(p_166055_ - p_166052_ < 1.0E-7)) {
-            int i = findBits(p_166050_, p_166053_);
-            int j = findBits(p_166051_, p_166054_);
-            int k = findBits(p_166052_, p_166055_);
-            if (i < 0 || j < 0 || k < 0) {
+    public static VoxelShape create(double minX, double minY, double minZ, double maxX, double maxY, double maxZ) {
+        // Paper start - optimise collisions
+        if (!(maxX - minX < 1.0E-7) && !(maxY - minY < 1.0E-7) && !(maxZ - minZ < 1.0E-7)) {
+            final int bitsX = findBits(minX, maxX);
+            final int bitsY = findBits(minY, maxY);
+            final int bitsZ = findBits(minZ, maxZ);
+            if (bitsX >= 0 && bitsY >= 0 && bitsZ >= 0) {
+                if (bitsX == 0 && bitsY == 0 && bitsZ == 0) {
+                    return BLOCK;
+                } else {
+                    final int sizeX = 1 << bitsX;
+                    final int sizeY = 1 << bitsY;
+                    final int sizeZ = 1 << bitsZ;
+                    final BitSetDiscreteVoxelShape shape = BitSetDiscreteVoxelShape.withFilledBounds(
+                            sizeX, sizeY, sizeZ,
+                            (int)Math.round(minX * (double)sizeX), (int)Math.round(minY * (double)sizeY), (int)Math.round(minZ * (double)sizeZ),
+                            (int)Math.round(maxX * (double)sizeX), (int)Math.round(maxY * (double)sizeY), (int)Math.round(maxZ * (double)sizeZ)
+                    );
+                    return new ArrayVoxelShape(
+                            shape,
+                            PARTS_BY_BITS[bitsX],
+                            PARTS_BY_BITS[bitsY],
+                            PARTS_BY_BITS[bitsZ]
+                    );
+                }
+            } else {
                 return new ArrayVoxelShape(
-                    BLOCK.shape,
-                    DoubleArrayList.wrap(new double[]{p_166050_, p_166053_}),
-                    DoubleArrayList.wrap(new double[]{p_166051_, p_166054_}),
-                    DoubleArrayList.wrap(new double[]{p_166052_, p_166055_})
-                );
-            } else if (i == 0 && j == 0 && k == 0) {
-                return block();
-            } else {
-                int l = 1 << i;
-                int i1 = 1 << j;
-                int j1 = 1 << k;
-                BitSetDiscreteVoxelShape bitsetdiscretevoxelshape = BitSetDiscreteVoxelShape.withFilledBounds(
-                    l,
-                    i1,
-                    j1,
-                    (int)Math.round(p_166050_ * (double)l),
-                    (int)Math.round(p_166051_ * (double)i1),
-                    (int)Math.round(p_166052_ * (double)j1),
-                    (int)Math.round(p_166053_ * (double)l),
-                    (int)Math.round(p_166054_ * (double)i1),
-                    (int)Math.round(p_166055_ * (double)j1)
-                );
-                return new CubeVoxelShape(bitsetdiscretevoxelshape);
+                        BLOCK.shape,
+                        minX == 0.0 && maxX == 1.0 ? ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.ZERO_ONE : DoubleArrayList.wrap(new double[] { minX, maxX }),
+                        minY == 0.0 && maxY == 1.0 ? ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.ZERO_ONE : DoubleArrayList.wrap(new double[] { minY, maxY }),
+                        minZ == 0.0 && maxZ == 1.0 ? ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.ZERO_ONE : DoubleArrayList.wrap(new double[] { minZ, maxZ })
+                );
             }
         } else {
-            return empty();
+            return EMPTY;
         }
+        // Paper end - optimise collisions
     }
 
     public static VoxelShape create(AABB p_83065_) {
@@ -119,8 +_,42 @@
         return join(p_83111_, p_83112_, BooleanOp.OR);
     }
 
-    public static VoxelShape or(VoxelShape p_83125_, VoxelShape... p_83126_) {
-        return Arrays.stream(p_83126_).reduce(p_83125_, Shapes::or);
+    // Paper start - optimise collisions
+    public static VoxelShape or(VoxelShape shape1, VoxelShape... others) {
+        int size = others.length;
+        if (size == 0) {
+            return shape1;
+        }
+
+        // reduce complexity of joins by splitting the merges
+
+        // add extra slot for first shape
+        ++size;
+        final VoxelShape[] tmp = Arrays.copyOf(others, size);
+        // insert first shape
+        tmp[size - 1] = shape1;
+
+        while (size > 1) {
+            int newSize = 0;
+            for (int i = 0; i < size; i += 2) {
+                final int next = i + 1;
+                if (next >= size) {
+                    // nothing to merge with, so leave it for next iteration
+                    tmp[newSize++] = tmp[i];
+                    break;
+                } else {
+                    // merge with adjacent
+                    final VoxelShape first = tmp[i];
+                    final VoxelShape second = tmp[next];
+
+                    tmp[newSize++] = Shapes.joinUnoptimized(first, second, BooleanOp.OR);
+                }
+            }
+            size = newSize;
+        }
+
+        return tmp[0].optimize();
+        // Paper end - optimise collisions
     }
 
     public static VoxelShape join(VoxelShape p_83114_, VoxelShape p_83115_, BooleanOp p_83116_) {
@@ -142,23 +_,23 @@
             } else {
                 IndexMerger indexmerger = createIndexMerger(1, p_83149_.getCoords(Direction.Axis.X), p_83150_.getCoords(Direction.Axis.X), flag, flag1);
                 IndexMerger indexmerger1 = createIndexMerger(
-                    indexmerger.size() - 1, p_83149_.getCoords(Direction.Axis.Y), p_83150_.getCoords(Direction.Axis.Y), flag, flag1
+                        indexmerger.size() - 1, p_83149_.getCoords(Direction.Axis.Y), p_83150_.getCoords(Direction.Axis.Y), flag, flag1
                 );
                 IndexMerger indexmerger2 = createIndexMerger(
-                    (indexmerger.size() - 1) * (indexmerger1.size() - 1),
-                    p_83149_.getCoords(Direction.Axis.Z),
-                    p_83150_.getCoords(Direction.Axis.Z),
-                    flag,
-                    flag1
+                        (indexmerger.size() - 1) * (indexmerger1.size() - 1),
+                        p_83149_.getCoords(Direction.Axis.Z),
+                        p_83150_.getCoords(Direction.Axis.Z),
+                        flag,
+                        flag1
                 );
                 BitSetDiscreteVoxelShape bitsetdiscretevoxelshape = BitSetDiscreteVoxelShape.join(
-                    p_83149_.shape, p_83150_.shape, indexmerger, indexmerger1, indexmerger2, p_83151_
+                        p_83149_.shape, p_83150_.shape, indexmerger, indexmerger1, indexmerger2, p_83151_
                 );
                 return (VoxelShape)(indexmerger instanceof DiscreteCubeMerger
                         && indexmerger1 instanceof DiscreteCubeMerger
                         && indexmerger2 instanceof DiscreteCubeMerger
-                    ? new CubeVoxelShape(bitsetdiscretevoxelshape)
-                    : new ArrayVoxelShape(bitsetdiscretevoxelshape, indexmerger.getList(), indexmerger1.getList(), indexmerger2.getList()));
+                        ? new CubeVoxelShape(bitsetdiscretevoxelshape)
+                        : new ArrayVoxelShape(bitsetdiscretevoxelshape, indexmerger.getList(), indexmerger1.getList(), indexmerger2.getList()));
             }
         }
     }
@@ -188,14 +_,14 @@
 
                     IndexMerger indexmerger = createIndexMerger(1, p_83158_.getCoords(Direction.Axis.X), p_83159_.getCoords(Direction.Axis.X), flag2, flag3);
                     IndexMerger indexmerger1 = createIndexMerger(
-                        indexmerger.size() - 1, p_83158_.getCoords(Direction.Axis.Y), p_83159_.getCoords(Direction.Axis.Y), flag2, flag3
+                            indexmerger.size() - 1, p_83158_.getCoords(Direction.Axis.Y), p_83159_.getCoords(Direction.Axis.Y), flag2, flag3
                     );
                     IndexMerger indexmerger2 = createIndexMerger(
-                        (indexmerger.size() - 1) * (indexmerger1.size() - 1),
-                        p_83158_.getCoords(Direction.Axis.Z),
-                        p_83159_.getCoords(Direction.Axis.Z),
-                        flag2,
-                        flag3
+                            (indexmerger.size() - 1) * (indexmerger1.size() - 1),
+                            p_83158_.getCoords(Direction.Axis.Z),
+                            p_83159_.getCoords(Direction.Axis.Z),
+                            flag2,
+                            flag3
                     );
                     return joinIsNotEmpty(indexmerger, indexmerger1, indexmerger2, p_83158_.shape, p_83159_.shape, p_83160_);
                 }
@@ -245,8 +_,8 @@
             return DoubleMath.fuzzyEquals(voxelshape.max(direction$axis), 1.0, 1.0E-7)
                 && DoubleMath.fuzzyEquals(voxelshape1.min(direction$axis), 0.0, 1.0E-7)
                 && !joinIsNotEmpty(
-                    new SliceShape(voxelshape, direction$axis, voxelshape.shape.getSize(direction$axis) - 1),
-                    new SliceShape(voxelshape1, direction$axis, 0),
+                    new SliceShape(voxelshape, direction$axis, voxelshape.shape.getSize(direction$axis) - 1, true), // Goldenforge
+                    new SliceShape(voxelshape1, direction$axis, 0, true), // Goldenforge
                     booleanop
                 );
         }
@@ -267,7 +_,7 @@
                 i = 0;
             }
 
-            return (VoxelShape)(!flag ? empty() : new SliceShape(p_83122_, direction$axis, i));
+            return (VoxelShape)(!flag ? empty() : new SliceShape(p_83122_, direction$axis, i, true)); // Goldenforge
         }
     }
 
@@ -288,8 +_,8 @@
             return !joinIsNotEmpty(
                 block(),
                 joinUnoptimized(
-                    new SliceShape(voxelshape, direction$axis, voxelshape.shape.getSize(direction$axis) - 1),
-                    new SliceShape(voxelshape1, direction$axis, 0),
+                    new SliceShape(voxelshape, direction$axis, voxelshape.shape.getSize(direction$axis) - 1, true), // Goldenforge
+                    new SliceShape(voxelshape1, direction$axis, 0, true), // Goldenforge
                     BooleanOp.OR
                 ),
                 BooleanOp.ONLY_FIRST
