--- a/net/minecraft/world/phys/shapes/SliceShape.java
+++ b/net/minecraft/world/phys/shapes/SliceShape.java
@@ -12,6 +_,7 @@
         super(makeSlice(p_83173_.shape, p_83174_, p_83175_));
         this.delegate = p_83173_;
         this.axis = p_83174_;
+        ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)this).moonrise$initCache(); // Paper - optimise collisions
     }
 
     private static DiscreteVoxelShape makeSlice(DiscreteVoxelShape p_83177_, Direction.Axis p_83178_, int p_83179_) {
